<!DOCTYPE html>
<html lang="en" class="js csstransforms3d">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="generator" content="Hugo 0.81.0" />
    <meta name="description" content="Analysis of bulk and single-cell ATAC-seq datasets">
<meta name="author" content="Jacques Serizay">

    <link rel="icon" href="/scATACseq-workshop/images/favicon.png" type="image/png">

    <title>Homework 2 :: Analysis of bulk and single-cell ATAC-seq datasets</title>

    
    <link href="/scATACseq-workshop/css/nucleus.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/fontawesome-all.min.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/hybrid.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/featherlight.min.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/perfect-scrollbar.min.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/auto-complete.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/atom-one-dark-reasonable.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/theme.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/hugo-theme.css?1630582900" rel="stylesheet">
    
    <link href="/scATACseq-workshop/css/theme-red.css?1630582900" rel="stylesheet">
    
    <link href="/scATACseq-workshop/css/custom_css.css?1630582900" rel="stylesheet">

    <script src="/scATACseq-workshop/js/jquery-3.3.1.min.js?1630582900"></script>

    <style>
      :root #header + #content > #left > #rlblock_left{
          display:none !important;
      }
      
        :not(pre) > code + span.copy-to-clipboard {
            display: none;
        }
      
    </style>
    
  </head>
  <body class="" data-url="/scATACseq-workshop/homeworks/homework_2/">
    <nav id="sidebar" class="">



  <div id="header-wrapper">
    <div id="header">
      <a id="logo">



</a>

    </div>
    
  </div>
  

    <div class="highlightable">
    <ul class="topics">

        
          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/introduction/" title=" Bulk and single-cell ATAC-seq analysis" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/introduction/">
           Bulk and single-cell ATAC-seq analysis
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/introduction/01_program/" title="1. Program" class="dd-item ">
        <a href="/scATACseq-workshop/introduction/01_program/">
        1. Program
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/introduction/02_prerequisites/" title="2. Prerequisites and required machine configuration" class="dd-item ">
        <a href="/scATACseq-workshop/introduction/02_prerequisites/">
        2. Prerequisites and required machine configuration
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/hands_on_sessions/" title="1. Hands-on sessions" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/hands_on_sessions/">
          1. Hands-on sessions
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/hands_on_sessions/day1/" title="Day 01" class="dd-item ">
        <a href="/scATACseq-workshop/hands_on_sessions/day1/">
        Day 01
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/hands_on_sessions/day2/" title="Day 02" class="dd-item ">
        <a href="/scATACseq-workshop/hands_on_sessions/day2/">
        Day 02
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/hands_on_sessions/day3/" title="Day 03" class="dd-item ">
        <a href="/scATACseq-workshop/hands_on_sessions/day3/">
        Day 03
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/homeworks/" title="2. Homeworks" class="dd-item
        parent
        
        
        ">
      <a href="/scATACseq-workshop/homeworks/">
          2. Homeworks
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/homeworks/homework_1/" title="Homework 1" class="dd-item ">
        <a href="/scATACseq-workshop/homeworks/homework_1/">
        Homework 1
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/homeworks/homework_2/" title="Homework 2" class="dd-item active">
        <a href="/scATACseq-workshop/homeworks/homework_2/">
        Homework 2
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/homeworks/homework_3/" title="Homework 3" class="dd-item ">
        <a href="/scATACseq-workshop/homeworks/homework_3/">
        Homework 3
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/extra_resources/" title="Extra resources" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/extra_resources/">
          Extra resources
          
      </a>
      
      
    </li>
  
 

          
        
    </ul>

    
    

    
    <section id="footer">
      <center>

    <a href="https://github.com/js2264/scATACseq-workshop">Check this website source on Github</a>
    <p>Created by J. Serizay</p>
    <p>Copyright 2021</p>

</center>

<script async defer src="https://buttons.github.io/buttons.js"></script>

    </section>
  </div>
</nav>




        <section id="body">
        <div id="overlay"></div>
        <div class="padding highlightable">
        <div id="head-tags">
        
        </div>
        
        <div id="body-inner">
          
            <h1>
              
              Homework 2
            </h1>
          

        



<script src="/scATACseq-workshop/rmarkdown-libs/header-attrs/header-attrs.js"></script>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<div id="goals" class="section level2">
<h2>Goals</h2>
<ol style="list-style-type: decimal">
<li>Import counts from mouse E18 brain scATACseq (5K) (provided by 10X Genomics) in R and process them</li>
</ol>
<blockquote>
<p>Processed data for mouse E18 brain scATACseq can be obtained directly from <a href="https://support.10xgenomics.com/single-cell-atac/datasets/1.2.0/atac_v1_E18_brain_fresh_5k">10X Genomics</a></p>
</blockquote>
<ol start="2" style="list-style-type: decimal">
<li>Transfer annotations from mouse E18 brain scRNAseq (5K) to mouse E18 brain scATACseq</li>
</ol>
<blockquote>
<p>Processed data for mouse E18 brain scRNAseq can be obtained directly from <a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K">10X Genomics</a></p>
</blockquote>
<ol start="3" style="list-style-type: decimal">
<li>Find genes associated to DA peaks and run GO over-enrichment / gene set enrichment analysis</li>
</ol>
</div>
<div id="process-mouse-e18-brain-5k-scatacseq-data" class="section level2">
<h2>1. Process mouse E18 brain 5K scATACseq data</h2>
<div id="download-data" class="section level3">
<h3>Download data</h3>
<p>scATACseq for mouse E18 5K brain can be downloaded from
<a href="https://support.10xgenomics.com/single-cell-atac/datasets/1.2.0/atac_v1_E18_brain_fresh_5k">10X Genomics</a>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="fu">mkdir</span> -p data/Homework_2/scATAC</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_analysis.tar.gz -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_analysis.tar.gz</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_filtered_peak_bc_matrix.h5 -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_filtered_peak_bc_matrix.h5</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_filtered_tf_bc_matrix.h5 -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_filtered_tf_bc_matrix.h5</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_fragments.tsv.gz -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_fragments.tsv.gz</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_fragments.tsv.gz.tbi -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_fragments.tsv.gz.tbi</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_peak_annotation.tsv -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_peak_annotation.tsv</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_peaks.bed -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_peaks.bed</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true"></a><span class="ex">curl</span> https://cg.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_possorted_bam.bam -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_possorted_bam.bam</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_possorted_bam.bam.bai -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_possorted_bam.bam.bai</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_singlecell.csv -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_singlecell.csv</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_web_summary.html -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_web_summary.html</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_E18_brain_fresh_5k/atac_v1_E18_brain_fresh_5k_cloupe.cloupe -o data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_cloupe.cloupe</span></code></pre></div>
</div>
<div id="import-data-in-r" class="section level3">
<h3>Import data in R</h3>
<p>First, the count matrix has to be imported. Use <code>Seurat</code> to do import the filtered <em>peak</em> count matrix.
How many peaks were found?</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="kw">library</span>(Signac)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true"></a><span class="kw">library</span>(Seurat)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true"></a><span class="kw">library</span>(BiocParallel)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true"></a><span class="kw">library</span>(plyranges)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true"></a><span class="co">## -- Read brain ATAC-seq counts</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true"></a>brain_assay &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">CreateChromatinAssay</span>(</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true"></a>    <span class="dt">counts =</span> Seurat<span class="op">::</span><span class="kw">Read10X_h5</span>(<span class="st">&quot;data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_filtered_peak_bc_matrix.h5&quot;</span>),</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true"></a>    <span class="dt">fragments =</span> <span class="st">&#39;data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_fragments.tsv.gz&#39;</span>,</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true"></a>    <span class="dt">sep =</span> <span class="kw">c</span>(<span class="st">&quot;:&quot;</span>, <span class="st">&quot;-&quot;</span>),</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true"></a>    <span class="dt">genome =</span> <span class="st">&quot;mm10&quot;</span>,</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true"></a>    <span class="dt">min.cells =</span> <span class="dv">1</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true"></a>)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true"></a>brain_assay</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true"></a><span class="kw">dim</span>(brain_assay)</span></code></pre></div>
<p>Cell metadata also needs to be read. Metadata is stored in a regular <code>csv</code> file.
You can import it in R using <code>read.csv()</code> from <code>base R</code>, but <code>vroom</code> package wil take
care of finding the best and fastest way to read it. It is especially useful for reading
large rectangular data in R.</p>
<p>How many cell barcode are there in total, in the <code>*_singlecell.csv</code> file? Should
we take this into consideration and filter some out?</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a><span class="co">## -- Read metadata</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a>metadata &lt;-<span class="st"> </span>vroom<span class="op">::</span><span class="kw">vroom</span>(</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a>    <span class="dt">file =</span> <span class="st">&quot;data/Homework_2/scATAC/atac_v1_E18_brain_fresh_5k_singlecell.csv&quot;</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true"></a>    <span class="dt">col_names =</span> <span class="ot">TRUE</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true"></a>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(barcode <span class="op">!=</span><span class="st"> &#39;NO_BARCODE&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">cell =</span> barcode) <span class="op">%&gt;%</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(cell <span class="op">%in%</span><span class="st"> </span><span class="kw">colnames</span>(brain_assay)) <span class="op">%&gt;%</span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true"></a><span class="st">    </span><span class="kw">column_to_rownames</span>(<span class="st">&#39;barcode&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true"></a><span class="st">    </span><span class="kw">as.data.frame</span>() </span></code></pre></div>
<p>Now that cells and counts over peaks / cell are created, one can bind them together
into a <code>Seurat</code> object. Try and create a standard <code>Seurat</code> object containing
the <code>chromatinAssay</code> object and corresponding cell metadata.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a><span class="co">## -- Create Seurat object</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">CreateSeuratObject</span>(</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true"></a>    <span class="dt">counts =</span> brain_assay,</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true"></a>    <span class="dt">assay =</span> <span class="st">&#39;peaks&#39;</span>,</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true"></a>    <span class="dt">project =</span> <span class="st">&#39;ATAC&#39;</span>,</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true"></a>    <span class="dt">meta.data =</span> metadata</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true"></a>)</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true"></a></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true"></a><span class="co">## -- Add genome annotations</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true"></a>annotations &lt;-<span class="st"> </span><span class="kw">GetGRangesFromEnsDb</span>(<span class="dt">ensdb =</span> EnsDb.Mmusculus.v79<span class="op">::</span>EnsDb.Mmusculus.v79)</span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true"></a><span class="kw">seqlevelsStyle</span>(annotations) &lt;-<span class="st"> &#39;UCSC&#39;</span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true"></a><span class="kw">genome</span>(annotations) &lt;-<span class="st"> &quot;mm10&quot;</span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true"></a><span class="kw">Annotation</span>(brain) &lt;-<span class="st"> </span>annotations</span></code></pre></div>
</div>
<div id="qc-data" class="section level3">
<h3>QC data</h3>
<blockquote>
<p>Nucleosome signal</p>
</blockquote>
<p>Few QC controls can be done. The <code>nucleosome signal</code> metric estimates the
ratio of nucleosome-spanning fragments vs mononucleosomal fragments, for each cell.
A low ratio (e.g. <code>&lt; 4</code>) indicates enrichment of fragments over an open
chromatin locus.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a><span class="co">## -- Nucleosome signal </span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">NucleosomeSignal</span>(<span class="dt">object =</span> brain)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true"></a>brain<span class="op">$</span>nucleosome_group &lt;-<span class="st"> </span><span class="kw">ifelse</span>(brain<span class="op">$</span>nucleosome_signal <span class="op">&gt;</span><span class="st"> </span><span class="dv">4</span>, <span class="st">&#39;NS &gt; 4&#39;</span>, <span class="st">&#39;NS &lt; 4&#39;</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true"></a><span class="kw">range</span>(brain<span class="op">$</span>nucleosome_signal)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true"></a><span class="kw">table</span>(brain<span class="op">$</span>nucleosome_group)</span></code></pre></div>
<p>Check the number of fragments / cell overlapping a known DHS locus (check <code>brain</code>
pre-computed metrics to find the corresponding metric), for cells with varying
<code>nucleosome_signal</code> scores.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a>p&lt;-<span class="st"> </span><span class="kw">tibble</span>(<span class="dt">dnase =</span> brain<span class="op">$</span>DNase_sensitive_region_fragments, <span class="dt">nuc_signal =</span> brain<span class="op">$</span>nucleosome_signal) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a>        <span class="dt">nuc_group =</span> <span class="kw">round</span>(nuc_signal <span class="op">/</span><span class="st"> </span><span class="dv">5</span>)<span class="op">*</span><span class="dv">5</span> <span class="op">+</span><span class="st"> </span><span class="dv">1</span>, </span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true"></a>        <span class="dt">nuc_group =</span> <span class="kw">factor</span>(nuc_group, <span class="kw">seq</span>(<span class="dv">1</span>, <span class="kw">max</span>(nuc_group)))</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true"></a>    ) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> nuc_group, <span class="dt">y =</span> dnase)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true"></a><span class="st">    </span><span class="kw">geom_boxplot</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">x =</span> <span class="st">&#39;Av. nucleosome signal&#39;</span>)</span></code></pre></div>
<blockquote>
<p>Fragment size distribution</p>
</blockquote>
<p>We have covered this already in previous exercises focusing on bulk ATAC.
It is a metric that one can also evaluate in scATACseq.
For this, one can import the fragments (from <code>*_fragments.tsv.gz</code> file) in R.
Here again, we highly recommend the use of <code>vroom</code> package, as the fragment file
can be several Gb big!</p>
<p>Try plotting the fragment width for the first 10M fragments.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a><span class="co">## -- Frag. size distribution</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true"></a><span class="kw">GetFragmentData</span>(<span class="kw">Fragments</span>(brain)[[<span class="dv">1</span>]])</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true"></a>frags &lt;-<span class="st"> </span>vroom<span class="op">::</span><span class="kw">vroom</span>(</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true"></a>        <span class="kw">GetFragmentData</span>(<span class="kw">Fragments</span>(brain)[[<span class="dv">1</span>]]), </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true"></a>        <span class="dt">col_names =</span> <span class="ot">FALSE</span>, </span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true"></a>        <span class="dt">n_max =</span> <span class="dv">10000000</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true"></a>    ) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true"></a><span class="st">    </span><span class="kw">setNames</span>(<span class="kw">c</span>(<span class="st">&#39;seqnames&#39;</span>, <span class="st">&#39;start&#39;</span>, <span class="st">&#39;end&#39;</span>, <span class="st">&#39;cell&#39;</span>, <span class="st">&#39;cluster&#39;</span>))</span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true"></a>p &lt;-frags <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">width =</span> end <span class="op">-</span><span class="st"> </span>start) <span class="op">%&gt;%</span></span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> width)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true"></a><span class="st">    </span><span class="kw">geom_bar</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true"></a><span class="st">    </span><span class="kw">xlim</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">800</span>))</span></code></pre></div>
<blockquote>
<p>TSSES</p>
</blockquote>
<p>This metric has also already been previously mentionned. It essentially represents
the signal-to-noise ratio at TSSs.
It can be estimated from a <code>Seurat</code> object using the <code>Signac::TSSEnrichment()</code> function.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true"></a><span class="co">## -- TSS enrichment</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">TSSEnrichment</span>(brain, <span class="dt">fast =</span> <span class="ot">TRUE</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true"></a><span class="kw">quantile</span>(brain<span class="op">$</span>TSS.enrichment, <span class="dt">probs =</span> <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.1</span>))</span></code></pre></div>
<p>We can also plot the average coverage at TSSs. Let’s plot it for the first
10M fragments</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true"></a>p&lt;-<span class="st"> </span><span class="kw">as_granges</span>(frags) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true"></a><span class="st">    </span><span class="kw">resize</span>(<span class="dv">1</span>, <span class="dt">fix =</span> <span class="st">&#39;start&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true"></a><span class="st">    </span><span class="kw">coverage</span>() <span class="op">%&gt;%</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true"></a><span class="st">    &#39;[&#39;</span>(target) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true"></a><span class="st">    </span><span class="kw">data.matrix</span>() <span class="op">%&gt;%</span></span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true"></a><span class="st">    </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true"></a><span class="st">    </span><span class="kw">setNames</span>(<span class="kw">seq</span>(<span class="op">-</span><span class="fl">999.5</span>, <span class="fl">999.5</span>, <span class="dt">length.out =</span> <span class="dv">2000</span>)) <span class="op">%&gt;%</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">peak =</span> <span class="kw">seq</span>(<span class="dv">1</span>, <span class="kw">nrow</span>(.)), <span class="dt">strand =</span> <span class="kw">as.character</span>(<span class="kw">strand</span>(target))) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true"></a><span class="st">    </span><span class="kw">pivot_longer</span>(<span class="op">-</span><span class="kw">c</span>(peak, strand), <span class="dt">names_to =</span> <span class="st">&#39;distance&#39;</span>, <span class="dt">values_to =</span> <span class="st">&#39;cnt&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">distance =</span> <span class="kw">as.numeric</span>(distance)) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">distance =</span> <span class="kw">ifelse</span>(strand <span class="op">==</span><span class="st"> &#39;-&#39;</span>, <span class="op">-</span>distance, distance)) <span class="op">%&gt;%</span></span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true"></a><span class="st">    </span><span class="kw">select</span>(<span class="op">-</span>peak, <span class="op">-</span>strand) <span class="op">%&gt;%</span></span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true"></a><span class="st">    </span><span class="kw">group_by</span>(distance) <span class="op">%&gt;%</span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true"></a><span class="st">    </span><span class="kw">tally</span>(cnt) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true"></a><span class="st">    </span><span class="kw">ggplot</span>(<span class="kw">aes</span>(<span class="dt">x =</span> distance, <span class="dt">y =</span> n)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb9-16"><a href="#cb9-16" aria-hidden="true"></a><span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span></span>
<span id="cb9-17"><a href="#cb9-17" aria-hidden="true"></a><span class="st">    </span><span class="kw">geom_smooth</span>(<span class="dt">method =</span> <span class="st">&#39;loess&#39;</span>, <span class="dt">span =</span> <span class="fl">0.05</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb9-18"><a href="#cb9-18" aria-hidden="true"></a><span class="st">    </span><span class="kw">theme_minimal</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb9-19"><a href="#cb9-19" aria-hidden="true"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&#39;none&#39;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb9-20"><a href="#cb9-20" aria-hidden="true"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&#39;Fragment &quot;cut&quot; sites&#39;</span>, <span class="dt">x =</span> <span class="st">&#39;Distance from TSS&#39;</span>, <span class="dt">y =</span> <span class="st">&#39;# of cut sites&#39;</span>)</span></code></pre></div>
</div>
<div id="subsetting-data" class="section level3">
<h3>Subsetting data</h3>
<p>Check the distribution of values for important metrics related to scATAC
quality. Notably, the # and % of fragments overlapping identified peaks or
blacklist regions.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a><span class="kw">quantile</span>(brain<span class="op">$</span>peak_region_fragments, <span class="dt">probs =</span> <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.1</span>))</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true"></a><span class="kw">quantile</span>(brain<span class="op">$</span>blacklist_region_fragments, <span class="dt">probs =</span> <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.1</span>))</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true"></a>brain<span class="op">$</span>FRiP &lt;-<span class="st"> </span>brain<span class="op">$</span>peak_region_fragments <span class="op">/</span><span class="st"> </span>brain<span class="op">$</span>passed_filters <span class="op">*</span><span class="st"> </span><span class="dv">100</span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true"></a><span class="kw">quantile</span>(brain<span class="op">$</span>FRiP, <span class="dt">probs =</span> <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.1</span>))</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true"></a>brain<span class="op">$</span>FRiBl &lt;-<span class="st"> </span>brain<span class="op">$</span>blacklist_region_fragments <span class="op">/</span><span class="st"> </span>brain<span class="op">$</span>peak_region_fragments</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true"></a><span class="kw">quantile</span>(brain<span class="op">$</span>FRiBl, <span class="dt">probs =</span> <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.1</span>))</span></code></pre></div>
<p>Set sensible thresholds to remove cells which do not seem to be high-quality,
then subset the <code>brain</code> Seurat object using these thresholds.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">subset</span>(</span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true"></a>    brain,</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true"></a>    peak_region_fragments <span class="op">&gt;</span><span class="st"> </span><span class="dv">2000</span> <span class="op">&amp;</span><span class="st"> </span><span class="co"># Keep cells with high number of fragments in peaks</span></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true"></a><span class="st">    </span>peak_region_fragments <span class="op">&lt;</span><span class="st"> </span><span class="dv">100000</span> <span class="op">&amp;</span><span class="st"> </span><span class="co"># Remove cells with number of fragments in peaks too high</span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true"></a><span class="st">    </span>FRiP <span class="op">&gt;</span><span class="st"> </span><span class="dv">50</span> <span class="op">&amp;</span><span class="st"> </span><span class="co"># Keep cells with high FRiP </span></span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true"></a><span class="st">    </span>FRiBl <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span> <span class="op">&amp;</span><span class="st"> </span><span class="co"># Keep cells with low FRiBl</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true"></a><span class="st">    </span>nucleosome_signal <span class="op">&lt;</span><span class="st"> </span><span class="dv">4</span> <span class="op">&amp;</span><span class="st"> </span><span class="co"># Remove cells with high nucleosome signal</span></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true"></a><span class="st">    </span>TSS.enrichment <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span> <span class="co"># Keep cells with good TSSES</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true"></a>)</span></code></pre></div>
<p>Don’t forget to also subset some peaks (which may have been identified from the
low-quality cells).</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>brain[</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true"></a>    <span class="kw">rowSums</span>(<span class="kw">GetAssayData</span>(brain, <span class="dt">slot =</span> <span class="st">&quot;counts&quot;</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>, <span class="co"># Remove peaks detected in less than 10 cells</span></span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true"></a>]</span></code></pre></div>
</div>
<div id="latent-semantic-indexing-normalizing-and-reduce-dimensionality-data-with-tf-idf-and-svd" class="section level3">
<h3>Latent semantic indexing: normalizing and reduce dimensionality data with ‘TF-IDF’ and ‘SVD’</h3>
<p>Normalization is done using a TF-IDF approach, rather than traditional log-normalization,
due to excessive sparsity of the scATAC data.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true"></a><span class="kw">GetAssayData</span>(brain, <span class="st">&#39;counts&#39;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true"></a><span class="kw">GetAssayData</span>(brain, <span class="st">&#39;data&#39;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">RunTFIDF</span>(brain)</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true"></a><span class="kw">GetAssayData</span>(brain, <span class="st">&#39;counts&#39;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true"></a><span class="kw">GetAssayData</span>(brain, <span class="st">&#39;data&#39;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span></code></pre></div>
<p>Dimensionality reduction can then be performed using the normalized data.
By using a <code>singular value decomposition</code> approach, one effectively recapitulates the
<a href="https://en.wikipedia.org/wiki/Latent_semantic_analysis"><code>latent semantic indexing</code></a> approach.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">FindTopFeatures</span>(brain, <span class="dt">min.cutoff =</span> <span class="st">&#39;q50&#39;</span>) <span class="co"># Find variable features</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">RunSVD</span>(<span class="dt">object =</span> brain) </span></code></pre></div>
</div>
<div id="clustering" class="section level3">
<h3>Clustering</h3>
<p>Now that cells are embedded in a lower dimensional space, they can be graph-clustered
just like “regular” cells from scRNAseq.</p>
<p>Try to perform graph-based clustering on the <code>Seurat</code> object, using the <code>lsi</code> embedded space
(skipping the first dimension).</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(<span class="dt">object =</span> brain, <span class="dt">reduction =</span> <span class="st">&#39;lsi&#39;</span>, <span class="dt">dims =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">30</span>)</span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(<span class="dt">object =</span> brain, <span class="dt">algorithm =</span> <span class="dv">3</span>, <span class="dt">resolution =</span> <span class="fl">1.2</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span></code></pre></div>
<p>Why was the first dimension skipped here? It appears that in <code>LSI</code>, the first dimension
is frequently (anti-)correlated with depth, so we exclude it in further analysis.</p>
</div>
<div id="visualization" class="section level3">
<h3>Visualization</h3>
<p>Now that the data has been normalized and dimensionality is linearly reduced (through <code>SVD</code>),
we can attempt to visualize the cluster locations in 2D.</p>
<p>Try computing tSNE and UMAP embeddings of the normalized <code>LSI</code> data, using <code>Seurat</code>-based
functions.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">RunTSNE</span>(brain, <span class="dt">reduction =</span> <span class="st">&#39;lsi&#39;</span>, <span class="dt">dims =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">30</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>(brain, <span class="dt">reduction =</span> <span class="st">&#39;lsi&#39;</span>, <span class="dt">dims =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">30</span>)</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true"></a>p1 &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">DimPlot</span>(brain, <span class="dt">reduction =</span> <span class="st">&#39;tsne&#39;</span>)</span></code></pre></div>
<p>One can also leverage <code>Signac</code>-based functions for visualization. Notably, the
<code>CoveragePlot()</code> function is quite useful: it can plot in R the coverage signal
over a genomic region, aggregated per cell cluster. Try and plot the aggregated
signal over individual markers known to be specific to certain brain cell types.</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true"></a>p2 &lt;-<span class="st"> </span><span class="kw">CoveragePlot</span>(</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true"></a>    <span class="dt">object =</span> brain,</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true"></a>    <span class="dt">region =</span> <span class="st">&quot;Tjp1&quot;</span>, <span class="co"># Or Gad2 (interneurons), Cq1a (Microglia), Gfap (astrocytes), Tubb3 (neurons)</span></span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true"></a>    <span class="dt">annotation =</span> <span class="ot">TRUE</span>,</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true"></a>    <span class="dt">peaks =</span> <span class="ot">FALSE</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true"></a>)</span></code></pre></div>
</div>
</div>
<div id="integrate-scatacseq-and-scrnaseq" class="section level2">
<h2>2. Integrate scATACseq and scRNAseq</h2>
<div id="download-data-1" class="section level3">
<h3>Download data</h3>
<p>scRNAseq for mouse E18 5K brain can be downloaded from
<a href="https://support.10xgenomics.com/single-cell-gene-expression/datasets/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K">10X Genomics</a>.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true"></a><span class="fu">mkdir</span> data/Homework_2/scRNA</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_web_summary.html -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_web_summary.html</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_metrics_summary.csv -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_metrics_summary.csv</span>
<span id="cb18-4"><a href="#cb18-4" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_cloupe.cloupe -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_cloupe.cloupe</span>
<span id="cb18-5"><a href="#cb18-5" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_sample_alignments.bam -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_sample_alignments.bam</span>
<span id="cb18-6"><a href="#cb18-6" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_sample_alignments.bam.bai -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_sample_alignments.bam.bai</span>
<span id="cb18-7"><a href="#cb18-7" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_sample_barcodes.csv -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_sample_barcodes.csv</span>
<span id="cb18-8"><a href="#cb18-8" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_sample_feature_bc_matrix.h5 -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_sample_feature_bc_matrix.h5</span>
<span id="cb18-9"><a href="#cb18-9" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_sample_feature_bc_matrix.tar.gz -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_sample_feature_bc_matrix.tar.gz</span>
<span id="cb18-10"><a href="#cb18-10" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_sample_molecule_info.h5 -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_sample_molecule_info.h5</span>
<span id="cb18-11"><a href="#cb18-11" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-exp/6.0.0/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K/SC3_v3_NextGem_DI_Neurons_5K_SC3_v3_NextGem_DI_Neurons_5K_count_analysis.tar.gz -o data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_analysis.tar.gz</span></code></pre></div>
</div>
<div id="process-data" class="section level3">
<h3>Process data</h3>
<p>Process the scRNAseq data with <code>Seurat</code> using a pipeline approach to limit the
generation of intermediate objects. Why is such approach sometimes useful in R?</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true"></a><span class="kw">library</span>(Seurat)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true"></a>brain_rna &lt;-<span class="st"> </span><span class="kw">Read10X_h5</span>(<span class="st">&quot;data/Homework_2/scRNA/SC3_v3_Neurons_5K_count_sample_feature_bc_matrix.h5&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true"></a><span class="st">    </span><span class="kw">CreateSeuratObject</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true"></a><span class="st">    </span><span class="kw">SCTransform</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true"></a><span class="st">    </span><span class="kw">RunPCA</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true"></a><span class="st">    </span><span class="kw">FindNeighbors</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true"></a><span class="st">    </span><span class="kw">FindClusters</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true"></a><span class="st">    </span><span class="kw">RunTSNE</span>(<span class="dt">reduction =</span> <span class="st">&#39;pca&#39;</span>, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">50</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb19-10"><a href="#cb19-10" aria-hidden="true"></a><span class="st">    </span><span class="kw">RunUMAP</span>(<span class="dt">reduction =</span> <span class="st">&#39;pca&#39;</span>, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">50</span>)</span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true"></a>p &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(brain_rna, <span class="dt">reduction =</span> <span class="st">&quot;umap&quot;</span>)</span></code></pre></div>
</div>
<div id="annotate-scrnaseq-data-with-public-databases" class="section level3">
<h3>Annotate scRNAseq data with public databases</h3>
<p>Leverage the <code>scRNAseq</code> package to recover pre-processed and annotated scRNAseq
data from Zeisel et al., 2018. We will use this to annotate cell identity of the
mouse E18 5K brain scRNAseq unlabelled data from 10X Genomics.</p>
<p>Which class is the newly created object? How different from a <code>Seurat</code> object this one is?</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true"></a>ref &lt;-<span class="st"> </span>scRNAseq<span class="op">::</span><span class="kw">ZeiselBrainData</span>()</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true"></a><span class="kw">class</span>(ref)</span></code></pre></div>
<p>Don’t forget to normalize the new counts in <code>R</code>! This can be achieved using
the <code>scuttle</code> package. One could also coerce the <code>ref</code> object as a <code>Seurat</code> object
to apply <code>SCT</code> normalization!</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true"></a>ref &lt;-<span class="st"> </span>scuttle<span class="op">::</span><span class="kw">logNormCounts</span>(ref)</span></code></pre></div>
<p>With the <code>SingleR</code> package, we can annotate cells using (1) the normalized
<code>count</code> <code>SparseMatrix</code> from the mouse E18 5K brain scRNAseq and (2) the
<code>SingleCellExperiment</code> <code>ref</code> object and its annotations.
Check <code>SingleR</code> documentation to see how this can be done.</p>
<p>Are the scRNAseq new annotations corroborating with the clusters found in the scRNAseq data?</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true"></a>ref<span class="op">$</span>level1class[ref<span class="op">$</span>level1class <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;pyramidal CA1&#39;</span>, <span class="st">&#39;pyramidal SS&#39;</span>)] &lt;-<span class="st"> &#39;pyramidal&#39;</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true"></a>pred &lt;-<span class="st"> </span>SingleR<span class="op">::</span><span class="kw">SingleR</span>(</span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true"></a>    <span class="dt">test =</span> <span class="kw">GetAssayData</span>(brain_rna, <span class="st">&#39;data&#39;</span>), </span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true"></a>    <span class="dt">ref =</span> ref, </span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true"></a>    <span class="dt">labels =</span> ref<span class="op">$</span>level1class, </span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true"></a>    <span class="dt">BPPARAM =</span> BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>(<span class="dt">workers =</span> <span class="dv">12</span>)</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true"></a>)</span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true"></a>brain_rna<span class="op">$</span>annotation &lt;-<span class="st"> </span>pred<span class="op">$</span>pruned.labels</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true"></a>p1 &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(brain_rna, <span class="dt">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="dt">group.by =</span> <span class="st">&#39;annotation&#39;</span>)</span>
<span id="cb22-10"><a href="#cb22-10" aria-hidden="true"></a>p2 &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(brain_rna, <span class="dt">reduction =</span> <span class="st">&quot;umap&quot;</span>, <span class="dt">group.by =</span> <span class="st">&#39;annotation&#39;</span>)</span></code></pre></div>
</div>
<div id="transfer-annotations-from-scrnaseq-to-scatacseq-datasets" class="section level3">
<h3>Transfer annotations from scRNAseq to scATACseq datasets</h3>
<p>A commonly used trick to transfer annotation from scRNAseq to scATACseq is to
process the scATACseq data so that it recapitulates “gene activity”.
With <code>Signac</code>, this can be done with the <code>GeneActivity()</code> function. Read its documentation
to understand what it does!</p>
<p>Using this function, we can create a second assay, containing “gene activity” metrics
for eeach gene in each cell. Now, the features are not accessible peaks anymore, but
“gene activity”.</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" aria-hidden="true"></a><span class="co">## - Compute an &quot;activity score&quot; for each gene in each cell (based on ATAC counts) and store it as a new assay</span></span>
<span id="cb23-2"><a href="#cb23-2" aria-hidden="true"></a>gene_activities &lt;-<span class="st"> </span><span class="kw">GeneActivity</span>(brain, <span class="dt">features =</span> <span class="kw">VariableFeatures</span>(brain_rna))</span>
<span id="cb23-3"><a href="#cb23-3" aria-hidden="true"></a>brain<span class="op">@</span>assays[<span class="st">&quot;gene_activity&quot;</span>] &lt;-<span class="st"> </span><span class="kw">CreateAssayObject</span>(<span class="dt">counts =</span> gene_activities)</span></code></pre></div>
<p>Don’t forget that these “gene activity” scores are obtained from raw scATACseq
counts. One needs to normalize them! In scRNAseq processed in <code>Seurat</code>, the <code>SCT</code> method is
typically used to normalize data.</p>
<p>Once you have ran <code>SCTransform()</code>, check the number of assays for the <code>brain</code>
object. How many are they? What does each assay contain exactly? Which one is the
default one now?</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true"></a><span class="co">## - Normalize this new assay with SCTransform</span></span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true"></a><span class="kw">GetAssayData</span>(brain, <span class="st">&#39;counts&#39;</span>)[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">SCTransform</span>(brain, <span class="dt">assay =</span> <span class="st">&#39;gene_activity&#39;</span>)</span></code></pre></div>
<p>Using <code>Seurat</code>-based <code>FindTransferAnchors()</code> and <code>TransferData()</code> functions,
find anchors between scRNAseq and scATACseq data, then transfer the annotations from
scRNAseq to scATACseq.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true"></a><span class="co">## - Find anchors between scRNAseq and &quot;gene activity&quot; from scATACseq. </span></span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true"></a>anchors &lt;-<span class="st"> </span><span class="kw">FindTransferAnchors</span>(</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true"></a>    <span class="dt">reference =</span> brain_rna, </span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true"></a>    <span class="dt">query =</span> brain, </span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true"></a>    <span class="dt">features =</span> <span class="kw">VariableFeatures</span>(<span class="dt">object =</span> brain_rna),</span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true"></a>    <span class="dt">reference.assay =</span> <span class="st">&quot;SCT&quot;</span>, </span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true"></a>    <span class="dt">query.assay =</span> <span class="st">&quot;SCT&quot;</span>, </span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true"></a>    <span class="dt">reduction =</span> <span class="st">&quot;cca&quot;</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true"></a>)</span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true"></a></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true"></a><span class="co">## - Transfer annotations from scRNA to scATAC</span></span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true"></a>pred_labels &lt;-<span class="st"> </span><span class="kw">TransferData</span>(</span>
<span id="cb25-13"><a href="#cb25-13" aria-hidden="true"></a>    <span class="dt">anchorset =</span> anchors, </span>
<span id="cb25-14"><a href="#cb25-14" aria-hidden="true"></a>    <span class="dt">refdata =</span> brain_rna<span class="op">$</span>annotation,</span>
<span id="cb25-15"><a href="#cb25-15" aria-hidden="true"></a>    <span class="dt">weight.reduction =</span> brain[[<span class="st">&#39;lsi&#39;</span>]], </span>
<span id="cb25-16"><a href="#cb25-16" aria-hidden="true"></a>    <span class="dt">dims =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">30</span></span>
<span id="cb25-17"><a href="#cb25-17" aria-hidden="true"></a>)</span>
<span id="cb25-18"><a href="#cb25-18" aria-hidden="true"></a></span>
<span id="cb25-19"><a href="#cb25-19" aria-hidden="true"></a><span class="co">## - Save transferred annotations</span></span>
<span id="cb25-20"><a href="#cb25-20" aria-hidden="true"></a>brain<span class="op">$</span>transferred_annotation &lt;-<span class="st"> </span>pred_labels<span class="op">$</span>predicted.id</span>
<span id="cb25-21"><a href="#cb25-21" aria-hidden="true"></a>brain<span class="op">$</span>transferred_annotation[pred_labels<span class="op">$</span>prediction.score.max <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.8</span>] &lt;-<span class="st"> </span><span class="ot">NA</span></span></code></pre></div>
<p>Check the transferred annotations: do they overlap what you would have thought from
earlier exploration of known markers?</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" aria-hidden="true"></a>p &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">DimPlot</span>(brain, <span class="dt">reduction =</span> <span class="st">&#39;tsne&#39;</span>, <span class="dt">group.by =</span> <span class="st">&#39;transferred_annotation&#39;</span>)</span></code></pre></div>
<p>BONUS QUESTION: have we already used a non-<code>Seurat</code> approach to transfer annotations from
one dataset to another? Try and apply it to transfer annotations from scRNAseq to scATACseq
data. Does this work?</p>
</div>
</div>
<div id="functional-analysis-of-da-peaks" class="section level2">
<h2>3. Functional analysis of DA peaks</h2>
<div id="extract-peaks-enriched-in-microglia-vs.-others" class="section level3">
<h3>Extract peaks enriched in microglia vs. others</h3>
<p><code>Seurat</code> implements differential abundance statistical tests for stored assays.
Try to use <code>FindMarkers()</code> to find peaks that are over-accessible in microglia
cells compared to the rest of the cells.</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true"></a><span class="co">## - Find microglia-specific peaks </span></span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true"></a><span class="kw">Idents</span>(brain) &lt;-<span class="st"> </span>brain<span class="op">$</span>transferred_annotation</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true"></a>microglia_peaks &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">FindMarkers</span>(</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true"></a>    <span class="dt">object =</span> brain,</span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true"></a>    <span class="dt">assay =</span> <span class="st">&#39;peaks&#39;</span>, </span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true"></a>    <span class="dt">ident.1 =</span> <span class="st">&#39;microglia&#39;</span>,</span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true"></a>    <span class="dt">ident.2 =</span> <span class="kw">unique</span>(<span class="kw">Idents</span>(brain)[<span class="kw">Idents</span>(brain) <span class="op">!=</span><span class="st"> &#39;microglia&#39;</span>]),</span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true"></a>    <span class="dt">min.pct =</span> <span class="fl">0.15</span>, </span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true"></a>    <span class="dt">only.pos =</span> <span class="ot">TRUE</span></span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true"></a>)</span></code></pre></div>
</div>
<div id="find-nearest-gene-for-each-microglia-specific-peak" class="section level3">
<h3>Find nearest gene for each microglia-specific peak</h3>
<p>A key step in scATACseq analysis is to link each peak to the gene it regulates.
The most basic way to do this is to associate a peak to the nearest gene.
Although this is a rather crude way of doing, it works reasonably well as a
“quick-and-dirty” first step to identify genes which are potentially undergoing
differential regulation between cell clusters.</p>
<div class="sourceCode" id="cb28"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb28-1"><a href="#cb28-1" aria-hidden="true"></a><span class="co">## - Link microglia peaks to nearest genes</span></span>
<span id="cb28-2"><a href="#cb28-2" aria-hidden="true"></a>mm_genes &lt;-<span class="st"> </span>ensembldb<span class="op">::</span><span class="kw">genes</span>(EnsDb.Mmusculus.v79<span class="op">::</span>EnsDb.Mmusculus.v79) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(gene_biotype <span class="op">==</span><span class="st"> &#39;protein_coding&#39;</span>)</span>
<span id="cb28-3"><a href="#cb28-3" aria-hidden="true"></a><span class="kw">seqlevelsStyle</span>(mm_genes) &lt;-<span class="st"> &#39;UCSC&#39;</span></span>
<span id="cb28-4"><a href="#cb28-4" aria-hidden="true"></a>microglia_genes &lt;-<span class="st"> </span>microglia_peaks <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb28-5"><a href="#cb28-5" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(p_val_adj <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>, avg_log2FC <span class="op">&gt;</span><span class="st"> </span><span class="dv">1</span>, pct<span class="fl">.1</span> <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.2</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb28-6"><a href="#cb28-6" aria-hidden="true"></a><span class="st">    </span><span class="kw">rownames</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb28-7"><a href="#cb28-7" aria-hidden="true"></a><span class="st">    </span><span class="kw">str_replace</span>(<span class="st">&#39;-&#39;</span>, <span class="st">&#39;:&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb28-8"><a href="#cb28-8" aria-hidden="true"></a><span class="st">    </span><span class="kw">GRanges</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb28-9"><a href="#cb28-9" aria-hidden="true"></a><span class="st">    </span><span class="kw">join_nearest</span>(mm_genes) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb28-10"><a href="#cb28-10" aria-hidden="true"></a><span class="st">    </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(symbol) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb28-11"><a href="#cb28-11" aria-hidden="true"></a><span class="st">    </span><span class="kw">unique</span>()</span></code></pre></div>
</div>
<div id="perform-go-over-representation-analysis" class="section level3">
<h3>Perform GO over-representation analysis</h3>
<p>The easiest way to perform GO over-representation analysis in <code>R</code> is to use
<code>gprofiler2</code> package. It sends the query to the <a href="https://biit.cs.ut.ee/gprofiler/gost">web-based application</a>,
which is well maintained, frequently updated and supports a large number of species.</p>
<p>Check whether there are enriched GO annotations associated with the genes undergoing microglia-specific regulation.</p>
<div class="sourceCode" id="cb29"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true"></a><span class="co">## - Run GO analysis</span></span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true"></a>goa &lt;-<span class="st"> </span>gprofiler2<span class="op">::</span><span class="kw">gost</span>(microglia_genes, <span class="dt">organism =</span> <span class="st">&#39;mmusculus&#39;</span>)</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true"></a>goa<span class="op">$</span>result <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(p_value <span class="op">&lt;=</span><span class="st"> </span><span class="fl">0.05</span>, <span class="kw">grepl</span>(<span class="st">&#39;GO:&#39;</span>, source)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(p_value) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">print</span>(<span class="dt">n =</span> <span class="dv">40</span>)</span></code></pre></div>
</div>
</div>


<footer class="footline">
	
</footer>

        
        </div>
        

      </div>

    <div id="navigation">
        
        

        
            
            
                
                    
                    
                
                

                    
                    
                        
                    
                    

                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        

        


	 
	 
		
			<a class="nav nav-prev" href="/scATACseq-workshop/homeworks/homework_1/" title="Homework 1"> <i class="fa fa-chevron-left"></i></a>
		
		
			<a class="nav nav-next" href="/scATACseq-workshop/homeworks/homework_3/" title="Homework 3" style="margin-right: 0px;"><i class="fa fa-chevron-right"></i></a>
		
	
    </div>

    </section>

    <div style="left: -1000px; overflow: scroll; position: absolute; top: -1000px; border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;">
      <div style="border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;"></div>
    </div>
    <script src="/scATACseq-workshop/js/clipboard.min.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/perfect-scrollbar.min.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/perfect-scrollbar.jquery.min.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/jquery.sticky.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/featherlight.min.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/highlight.pack.js?1630582900"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script src="/scATACseq-workshop/js/modernizr.custom-3.6.0.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/learn.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/hugo-learn.js?1630582900"></script>
    
        
            <script src="/scATACseq-workshop/mermaid/mermaid.js?1630582900"></script>
        
        <script>
            mermaid.initialize({ startOnLoad: true });
        </script>
    
    <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-105947713-1', 'auto');
  ga('send', 'pageview');

</script>

  </body>
</html>

