<!DOCTYPE html>
<html lang="en" class="js csstransforms3d">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="generator" content="Hugo 0.81.0" />
    <meta name="description" content="Analysis of bulk and single-cell ATAC-seq datasets">
<meta name="author" content="Jacques Serizay">

    <link rel="icon" href="/scATACseq-workshop/images/favicon.png" type="image/png">

    <title>Homework 3 :: Analysis of bulk and single-cell ATAC-seq datasets</title>

    
    <link href="/scATACseq-workshop/css/nucleus.css?1628787916" rel="stylesheet">
    <link href="/scATACseq-workshop/css/fontawesome-all.min.css?1628787916" rel="stylesheet">
    <link href="/scATACseq-workshop/css/hybrid.css?1628787916" rel="stylesheet">
    <link href="/scATACseq-workshop/css/featherlight.min.css?1628787916" rel="stylesheet">
    <link href="/scATACseq-workshop/css/perfect-scrollbar.min.css?1628787916" rel="stylesheet">
    <link href="/scATACseq-workshop/css/auto-complete.css?1628787916" rel="stylesheet">
    <link href="/scATACseq-workshop/css/atom-one-dark-reasonable.css?1628787916" rel="stylesheet">
    <link href="/scATACseq-workshop/css/theme.css?1628787916" rel="stylesheet">
    <link href="/scATACseq-workshop/css/hugo-theme.css?1628787916" rel="stylesheet">
    
    <link href="/scATACseq-workshop/css/theme-red.css?1628787916" rel="stylesheet">
    
    <link href="/scATACseq-workshop/css/custom_css.css?1628787916" rel="stylesheet">

    <script src="/scATACseq-workshop/js/jquery-3.3.1.min.js?1628787916"></script>

    <style>
      :root #header + #content > #left > #rlblock_left{
          display:none !important;
      }
      
        :not(pre) > code + span.copy-to-clipboard {
            display: none;
        }
      
    </style>
    
  </head>
  <body class="" data-url="/scATACseq-workshop/homeworks/homework_3/">
    <nav id="sidebar" class="">



  <div id="header-wrapper">
    <div id="header">
      <a id="logo">



</a>

    </div>
    
  </div>
  

    <div class="highlightable">
    <ul class="topics">

        
          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/introduction/" title=" Bulk and single-cell ATAC-seq analysis" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/introduction/">
           Bulk and single-cell ATAC-seq analysis
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/introduction/01_program/" title="1. Program" class="dd-item ">
        <a href="/scATACseq-workshop/introduction/01_program/">
        1. Program
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/introduction/02_prerequisites/" title="2. Prerequisites and required machine configuration" class="dd-item ">
        <a href="/scATACseq-workshop/introduction/02_prerequisites/">
        2. Prerequisites and required machine configuration
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/hands_on_sessions/" title="1. Hands-on sessions" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/hands_on_sessions/">
          1. Hands-on sessions
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/hands_on_sessions/day1/" title="Day 01" class="dd-item ">
        <a href="/scATACseq-workshop/hands_on_sessions/day1/">
        Day 01
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/hands_on_sessions/day2/" title="Day 02" class="dd-item ">
        <a href="/scATACseq-workshop/hands_on_sessions/day2/">
        Day 02
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/hands_on_sessions/day3/" title="Day 03" class="dd-item ">
        <a href="/scATACseq-workshop/hands_on_sessions/day3/">
        Day 03
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/homeworks/" title="2. Homeworks" class="dd-item
        parent
        
        
        ">
      <a href="/scATACseq-workshop/homeworks/">
          2. Homeworks
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/homeworks/homework_1/" title="Homework 1" class="dd-item ">
        <a href="/scATACseq-workshop/homeworks/homework_1/">
        Homework 1
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/homeworks/homework_2/" title="Homework 2" class="dd-item ">
        <a href="/scATACseq-workshop/homeworks/homework_2/">
        Homework 2
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/homeworks/homework_3/" title="Homework 3" class="dd-item active">
        <a href="/scATACseq-workshop/homeworks/homework_3/">
        Homework 3
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/extra_resources/" title="Extra resources" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/extra_resources/">
          Extra resources
          
      </a>
      
      
    </li>
  
 

          
        
    </ul>

    
    

    
    <section id="footer">
      <center>

    <a href="https://github.com/js2264/scATACseq-workshop">Check this website source on Github</a>
    <p>Created by J. Serizay</p>
    <p>Copyright 2021</p>

</center>

<script async defer src="https://buttons.github.io/buttons.js"></script>

    </section>
  </div>
</nav>




        <section id="body">
        <div id="overlay"></div>
        <div class="padding highlightable">
        <div id="head-tags">
        
        </div>
        
        <div id="body-inner">
          
            <h1>
              
              Homework 3
            </h1>
          

        



<script src="/scATACseq-workshop/rmarkdown-libs/header-attrs/header-attrs.js"></script>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<div id="goals" class="section level2">
<h2>Goals</h2>
<ol style="list-style-type: decimal">
<li><p>Import, filter, normalize and plot cells from human hematopoiesis scATACseq dataset</p></li>
<li><p>Compute gene activity scores and check known markers</p></li>
<li><p>Run chromVAR to identify TF activity in each cell cluster</p></li>
</ol>
</div>
<div id="process-human-hematopoiesis-dataset" class="section level2">
<h2>1. Process human hematopoiesis dataset</h2>
<div id="download-data" class="section level3">
<h3>Download data</h3>
<p>The data comes from Satpathy et al., Nat. Biotechnol. 2019 (<a href="https://doi.org/10.1038/s41587-019-0206-z">DOI: 10.1038/s41587-019-0206-z</a>).
Counts data is available from GEO (GSE: <code>GSE129785</code>).</p>
<p>Download the files related to scATACseq of all human hematopoiesis cells.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="fu">mkdir</span> -p data/Homework_3/scATAC/</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a><span class="ex">curl</span> https://ftp.ncbi.nlm.nih.gov/geo/series/GSE129nnn/GSE129785/suppl/GSE129785_scATAC-Hematopoiesis-All.cell_barcodes.txt.gz -o data/Homework_3/scATAC/barcodes.tsv.gz</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a><span class="ex">curl</span> https://ftp.ncbi.nlm.nih.gov/geo/series/GSE129nnn/GSE129785/suppl/GSE129785_scATAC-Hematopoiesis-All.mtx.gz -o data/Homework_3/scATAC/matrix.mtx.gz</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a><span class="ex">curl</span> https://ftp.ncbi.nlm.nih.gov/geo/series/GSE129nnn/GSE129785/suppl/GSE129785_scATAC-Hematopoiesis-All.peaks.txt.gz -o data/Homework_3/scATAC/features.tsv.gz</span></code></pre></div>
</div>
<div id="process-data-wih-signac" class="section level3">
<h3>Process data wih Signac</h3>
<div id="import-data" class="section level4">
<h4>Import data</h4>
<p>Notice how the count matrix is in a different format than previously seen.
You will frequently encounter this type of situation. Analysis packages often provide
import functions for the different input formats. Check <code>Seurat</code> documentation if needed,
to identify which function can be used to import a <code>mtx</code> format.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a>cnts &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">Read10X</span>(<span class="st">&#39;data/Homework_3/scATAC/&#39;</span>)</span></code></pre></div>
<p>Does it work? Why? The most likely reason is that the count matrix is not
structured exactly like the one which comes straight out of a <code>cellranger count</code>
pipeline. It is a very frequent issue occurring when one attempts to analyse
public data.</p>
<p>The alternative, longer route is to import each file separately,
make sure they are compatible with each other, and merge them
manually into a <code>SeuratChromatinAssay</code>.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a><span class="co">## - Load required libraries </span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a><span class="kw">library</span>(plyranges)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true"></a><span class="kw">library</span>(Seurat)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true"></a><span class="kw">library</span>(Signac)</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true"></a></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true"></a><span class="co">## - Import counts, barcodes and features separately</span></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true"></a>barcode &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="st">&#39;data/Homework_3/scATAC/&#39;</span>, <span class="st">&quot;barcodes.tsv.gz&quot;</span>)</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true"></a>features &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="st">&#39;data/Homework_3/scATAC/&#39;</span>, <span class="st">&quot;features.tsv.gz&quot;</span>)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true"></a>matrix &lt;-<span class="st"> </span><span class="kw">file.path</span>(<span class="st">&#39;data/Homework_3/scATAC/&#39;</span>, <span class="st">&quot;matrix.mtx.gz&quot;</span>)</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true"></a>cnts &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">readMM</span>(<span class="st">&#39;data/Homework_3/scATAC/matrix.mtx.gz&#39;</span>)</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true"></a>barcodes &lt;-<span class="st"> </span>vroom<span class="op">::</span><span class="kw">vroom</span>(<span class="st">&#39;data/Homework_3/scATAC/barcodes.tsv.gz&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.data.frame</span>()</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true"></a>features &lt;-<span class="st"> </span><span class="kw">readLines</span>(<span class="st">&#39;data/Homework_3/scATAC/features.tsv.gz&#39;</span>)[<span class="op">-</span><span class="dv">1</span>] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true"></a><span class="st">    </span><span class="kw">str_replace</span>(<span class="st">&#39;_&#39;</span>, <span class="st">&#39;:&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true"></a><span class="st">    </span><span class="kw">str_replace</span>(<span class="st">&#39;_&#39;</span>, <span class="st">&#39;-&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true"></a><span class="st">    </span><span class="kw">GRanges</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">peak =</span> <span class="kw">as.character</span>(.) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">str_replace</span>(<span class="st">&#39;:&#39;</span>, <span class="st">&#39;-&#39;</span>))</span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true"></a></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true"></a><span class="co">## - Make sure that the row/colnames for the count matrix match those of </span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true"></a><span class="co">#       barcodes and features</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true"></a><span class="kw">colnames</span>(cnts) &lt;-<span class="st"> </span><span class="kw">rownames</span>(barcodes) &lt;-<span class="st"> </span>barcodes<span class="op">$</span>Group_Barcode</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true"></a><span class="kw">rownames</span>(cnts) &lt;-<span class="st"> </span><span class="kw">names</span>(features) &lt;-<span class="st"> </span>features<span class="op">$</span>peak</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true"></a></span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true"></a><span class="co">## - Get human gene annotations (hg19/GRCh37) to feed it into the future `SeuratChromatinAssay`</span></span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true"></a>annotations &lt;-<span class="st"> </span><span class="kw">GetGRangesFromEnsDb</span>(<span class="dt">ensdb =</span> EnsDb.Hsapiens.v75<span class="op">::</span>EnsDb.Hsapiens.v75)</span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true"></a><span class="kw">seqlevelsStyle</span>(annotations) &lt;-<span class="st"> &#39;UCSC&#39;</span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true"></a><span class="kw">genome</span>(annotations) &lt;-<span class="st"> &quot;hg19&quot;</span></span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true"></a></span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true"></a><span class="co">## - Create Chromatin Assay</span></span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true"></a>assay &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">CreateChromatinAssay</span>(</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true"></a>    <span class="dt">counts =</span> cnts,</span>
<span id="cb3-32"><a href="#cb3-32" aria-hidden="true"></a>    <span class="dt">ranges =</span> features,</span>
<span id="cb3-33"><a href="#cb3-33" aria-hidden="true"></a>    <span class="dt">genome =</span> <span class="st">&quot;hg19&quot;</span>,</span>
<span id="cb3-34"><a href="#cb3-34" aria-hidden="true"></a>    <span class="dt">min.cells =</span> <span class="dv">10</span>, </span>
<span id="cb3-35"><a href="#cb3-35" aria-hidden="true"></a>    <span class="dt">min.features =</span> <span class="dv">10</span>, </span>
<span id="cb3-36"><a href="#cb3-36" aria-hidden="true"></a>    <span class="dt">annotation =</span> annotations</span>
<span id="cb3-37"><a href="#cb3-37" aria-hidden="true"></a>)</span>
<span id="cb3-38"><a href="#cb3-38" aria-hidden="true"></a>hemato &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">CreateSeuratObject</span>(</span>
<span id="cb3-39"><a href="#cb3-39" aria-hidden="true"></a>    <span class="dt">counts =</span> assay,</span>
<span id="cb3-40"><a href="#cb3-40" aria-hidden="true"></a>    <span class="dt">assay =</span> <span class="st">&#39;ATAC&#39;</span>,</span>
<span id="cb3-41"><a href="#cb3-41" aria-hidden="true"></a>    <span class="dt">meta.data =</span> barcodes</span>
<span id="cb3-42"><a href="#cb3-42" aria-hidden="true"></a>)</span></code></pre></div>
<p>You will find that <code>CreateChromatinAssay()</code> function takes a long time to run.
This is because it internally computes metrics from the count matrix. You can
check the cell metadata to see which ones were calculated (on top of the ones
which come from the <code>barcodes</code> data frame).</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a><span class="kw">head</span>(hemato[[]])</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true"></a><span class="kw">colnames</span>(hemato[[]])[<span class="op">!</span><span class="kw">colnames</span>(hemato[[]]) <span class="op">%in%</span><span class="st"> </span><span class="kw">colnames</span>(barcodes)]</span></code></pre></div>
</div>
<div id="process-data" class="section level4">
<h4>Process data</h4>
<p>Is the <code>fragments</code> file (which should exist for any scATACseq experiment) available
for this experiment? Check the related GEO webpage.</p>
<p>For which analysis are the <code>fragments</code> essential, exactly? Can we still
perform normalization/clustering/annotation without them? And motif enrichment analysis?</p>
<p>Since we don’t have the <code>fragments</code> file at hand, most of the QC steps are not
available (e.g. <code>TSSEnrichment</code>, <code>NucleosomeSignal</code> or fragment size distribution).</p>
<p>(Note that the <code>fragments</code> file is actually available as raw data from GEO.
Feel free to download it and perform further QC).</p>
<p>Still, we do have other pre-computed metrics (e.g. <code>FRIP</code> and <code>depth</code>).
Check these metrics and filter the <code>Seurat</code> object (cells and features)
as deemed appropriate.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a><span class="co">## - Filter data </span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true"></a><span class="kw">quantile</span>(hemato<span class="op">$</span>depth, <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.1</span>))</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true"></a><span class="kw">quantile</span>(hemato<span class="op">$</span>nFeature_ATAC, <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.1</span>))</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true"></a><span class="kw">quantile</span>(hemato<span class="op">$</span>nCount_ATAC, <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.1</span>))</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true"></a><span class="kw">quantile</span>(hemato<span class="op">$</span>FRIP, <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.1</span>))</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true"></a>hemato &lt;-<span class="st"> </span><span class="kw">subset</span>(hemato, <span class="dt">subset =</span> depth <span class="op">&gt;</span><span class="st"> </span><span class="dv">5000</span> <span class="op">&amp;</span><span class="st"> </span>depth <span class="op">&lt;</span><span class="st"> </span><span class="dv">200000</span>) </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true"></a>hemato &lt;-<span class="st"> </span><span class="kw">subset</span>(hemato, <span class="dt">subset =</span> nCount_ATAC <span class="op">&gt;</span><span class="st"> </span><span class="dv">3000</span> <span class="op">&amp;</span><span class="st"> </span>nCount_ATAC <span class="op">&lt;</span><span class="st"> </span><span class="dv">100000</span>) </span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true"></a>hemato &lt;-<span class="st"> </span><span class="kw">subset</span>(hemato, <span class="dt">subset =</span> nFeature_ATAC <span class="op">&gt;</span><span class="st"> </span><span class="dv">1000</span> <span class="op">&amp;</span><span class="st"> </span>nFeature_ATAC <span class="op">&lt;</span><span class="st"> </span><span class="dv">20000</span>) </span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true"></a>hemato &lt;-<span class="st"> </span><span class="kw">subset</span>(hemato, <span class="dt">subset =</span> FRIP <span class="op">&gt;</span><span class="st"> </span><span class="fl">0.30</span> <span class="op">&amp;</span><span class="st"> </span>FRIP <span class="op">&lt;</span><span class="st"> </span><span class="dv">20000</span>) </span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true"></a></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true"></a><span class="co">## - Remove peaks with low coverage</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true"></a>hemato &lt;-<span class="st"> </span>hemato[<span class="kw">rowSums</span>(<span class="kw">GetAssayData</span>(hemato, <span class="dt">slot =</span> <span class="st">&quot;counts&quot;</span>)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>, ]</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true"></a>hemato &lt;-<span class="st"> </span>hemato[<span class="kw">rowSums</span>(<span class="kw">GetAssayData</span>(hemato, <span class="dt">slot =</span> <span class="st">&quot;counts&quot;</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>, ]</span></code></pre></div>
<p>Now that the dataset is filtered, one can normalize (by using TF-IDF
approach) then further reduce the dimensionality for visualization purposes.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a><span class="co">## - Normalize data </span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a>hemato &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">RunTFIDF</span>(hemato) </span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true"></a><span class="co">## - Reduce dimensionality</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true"></a>hemato &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">FindTopFeatures</span>(hemato, <span class="dt">min.cutoff =</span> <span class="st">&#39;q50&#39;</span>) </span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true"></a>hemato &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">RunSVD</span>(hemato) </span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true"></a></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true"></a><span class="co">## - Label clusters according to the original publication</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true"></a>clusters &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true"></a>    <span class="st">&#39;Cluster1&#39;</span> =<span class="st"> &#39;HSC&#39;</span>,</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true"></a>    <span class="st">&#39;Cluster2&#39;</span> =<span class="st"> &#39;MEP&#39;</span>,</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true"></a>    <span class="st">&#39;Cluster3&#39;</span> =<span class="st"> &#39;CMP&#39;</span>,</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true"></a>    <span class="st">&#39;Cluster4&#39;</span> =<span class="st"> &#39;LMPP&#39;</span>,</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true"></a>    <span class="st">&#39;Cluster5&#39;</span> =<span class="st"> &#39;CLP&#39;</span>,</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true"></a>    <span class="st">&#39;Cluster6&#39;</span> =<span class="st"> &#39;Pro-B&#39;</span>,</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true"></a>    <span class="st">&#39;Cluster7&#39;</span> =<span class="st"> &#39;Pre-B&#39;</span>,</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true"></a>    <span class="st">&#39;Cluster8&#39;</span> =<span class="st"> &#39;GMP&#39;</span>,</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true"></a>    <span class="st">&#39;Cluster9&#39;</span> =<span class="st"> &#39;MDP&#39;</span>,</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true"></a>    <span class="st">&#39;Cluster10&#39;</span> =<span class="st"> &#39;pDC&#39;</span>,</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true"></a>    <span class="st">&#39;Cluster11&#39;</span> =<span class="st"> &#39;cDC&#39;</span>,</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true"></a>    <span class="st">&#39;Cluster12&#39;</span> =<span class="st"> &#39;Mono.&#39;</span>,</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true"></a>    <span class="st">&#39;Cluster13&#39;</span> =<span class="st"> &#39;Mono.&#39;</span>,</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true"></a>    <span class="st">&#39;Cluster14&#39;</span> =<span class="st"> &#39;Nai. B&#39;</span>,</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true"></a>    <span class="st">&#39;Cluster15&#39;</span> =<span class="st"> &#39;Mem. B&#39;</span>,</span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true"></a>    <span class="st">&#39;Cluster16&#39;</span> =<span class="st"> &#39;Plasma&#39;</span>,</span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true"></a>    <span class="st">&#39;Cluster17&#39;</span> =<span class="st"> &#39;Baso.&#39;</span>,</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true"></a>    <span class="st">&#39;Cluster18&#39;</span> =<span class="st"> &#39;Imm. NK&#39;</span>,</span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true"></a>    <span class="st">&#39;Cluster19&#39;</span> =<span class="st"> &#39;Mat. NK&#39;</span>,</span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true"></a>    <span class="st">&#39;Cluster20&#39;</span> =<span class="st"> &#39;Mat. NK&#39;</span>,</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true"></a>    <span class="st">&#39;Cluster21&#39;</span> =<span class="st"> &#39;Nai. CD4&#39;</span>,</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true"></a>    <span class="st">&#39;Cluster22&#39;</span> =<span class="st"> &#39;Nai. CD4&#39;</span>,</span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true"></a>    <span class="st">&#39;Cluster23&#39;</span> =<span class="st"> &#39;Nai. Treg&#39;</span>,</span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true"></a>    <span class="st">&#39;Cluster24&#39;</span> =<span class="st"> &#39;Mem. CD4&#39;</span>,</span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true"></a>    <span class="st">&#39;Cluster25&#39;</span> =<span class="st"> &#39;Treg&#39;</span>,</span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true"></a>    <span class="st">&#39;Cluster26&#39;</span> =<span class="st"> &#39;Nai. CD8&#39;</span>,</span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true"></a>    <span class="st">&#39;Cluster27&#39;</span> =<span class="st"> &#39;Nai. CD8&#39;</span>,</span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true"></a>    <span class="st">&#39;Cluster28&#39;</span> =<span class="st"> &#39;Nai. CD8&#39;</span>,</span>
<span id="cb6-38"><a href="#cb6-38" aria-hidden="true"></a>    <span class="st">&#39;Cluster29&#39;</span> =<span class="st"> &#39;Mem. CD8&#39;</span>,</span>
<span id="cb6-39"><a href="#cb6-39" aria-hidden="true"></a>    <span class="st">&#39;Cluster30&#39;</span> =<span class="st"> &#39;Mem. CD8&#39;</span>,</span>
<span id="cb6-40"><a href="#cb6-40" aria-hidden="true"></a>    <span class="st">&#39;Cluster31&#39;</span> =<span class="st"> &#39;Gamm. Del. T&#39;</span></span>
<span id="cb6-41"><a href="#cb6-41" aria-hidden="true"></a>)</span>
<span id="cb6-42"><a href="#cb6-42" aria-hidden="true"></a>hemato<span class="op">$</span>renamed_clusters &lt;-<span class="st"> </span>clusters[hemato<span class="op">$</span>Clusters]</span>
<span id="cb6-43"><a href="#cb6-43" aria-hidden="true"></a></span>
<span id="cb6-44"><a href="#cb6-44" aria-hidden="true"></a><span class="co">## - Visualize data </span></span>
<span id="cb6-45"><a href="#cb6-45" aria-hidden="true"></a>hemato &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">RunUMAP</span>(hemato, <span class="dt">reduction =</span> <span class="st">&#39;lsi&#39;</span>, <span class="dt">dims =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">30</span>)</span>
<span id="cb6-46"><a href="#cb6-46" aria-hidden="true"></a>p &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(hemato, <span class="dt">group.by =</span> <span class="st">&#39;renamed_clusters&#39;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">coord_fixed</span>(<span class="dt">ratio =</span> <span class="dv">1</span>)</span></code></pre></div>
<p>What can you observe in the <code>UMAP</code> projection of the dataset? Comment on the
separation of some cell types into different spatially-resolved clusters.</p>
</div>
</div>
</div>
<div id="compute-gene-activity-scores" class="section level2">
<h2>2. Compute gene activity scores</h2>
<p>Here again, the absence of <code>fragments</code> complicates the computation of gene activity scores.
Indeed, Signac’s <code>GeneActivity()</code> function does require scATACseq fragment information.</p>
<p>Without actual fragments, can you think of a way to still compute a ‘gene activity’-like score,
approximating ~ the sum of scATACseq fragments overlapping a gene?</p>
<p>One way to do this is to:</p>
<ol style="list-style-type: decimal">
<li>Extend gene annotations to cover 2000bp upstream of TSS</li>
<li>For each gene, link it to all the peaks overlapping it to create a gene &lt;-&gt; peak relationship table</li>
<li>Create a joint data table containing [ gene | peak | cell | count ]</li>
<li>Group the data table by <code>cell + gene</code> and sum up the peak counts for each pair</li>
<li>Transform this count table into a SparseMatrix</li>
<li>Create the new <code>Seurat</code> assay containing “gene activity”-like scores and normalize it</li>
</ol>
<ul>
<li>Step 1:</li>
</ul>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a><span class="co">## - Extend all the genes in hg19 so that it covers 1000 bp upstream of the genebody </span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true"></a>hg_genes_extended &lt;-<span class="st"> </span>ensembldb<span class="op">::</span><span class="kw">genes</span>(EnsDb.Hsapiens.v75<span class="op">::</span>EnsDb.Hsapiens.v75) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(gene_biotype <span class="op">==</span><span class="st"> &#39;protein_coding&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true"></a><span class="st">    </span><span class="kw">keepStandardChromosomes</span>(<span class="dt">pruning.mode =</span> <span class="st">&#39;coarse&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(seqnames <span class="op">!=</span><span class="st"> &#39;MT&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true"></a><span class="st">    </span><span class="kw">resize</span>(<span class="dt">width =</span> <span class="kw">width</span>(.) <span class="op">+</span><span class="st"> </span><span class="dv">2000</span>, <span class="dt">fix =</span> <span class="st">&#39;start&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true"></a><span class="st">    </span><span class="kw">shift</span>(<span class="dv">1000</span> <span class="op">*</span><span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">strand</span>(.) <span class="op">==</span><span class="st"> &#39;+&#39;</span>, <span class="dv">-1</span>, <span class="dv">1</span>)) <span class="op">%&gt;%</span></span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">id =</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(.)) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">gene_name =</span> <span class="kw">make.unique</span>(gene_name))</span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true"></a><span class="kw">seqlevelsStyle</span>(hg_genes_extended) &lt;-<span class="st"> &#39;UCSC&#39;</span></span></code></pre></div>
<ul>
<li>Step 2:</li>
</ul>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true"></a><span class="co">## - Find overlaps between all the scATAC peaks and the extended genes</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true"></a>ov &lt;-<span class="st"> </span><span class="kw">findOverlaps</span>(<span class="dt">query =</span> <span class="kw">granges</span>(hemato), <span class="dt">subject =</span> hg_genes_extended)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true"></a></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true"></a><span class="co">## - For each peak-gene pair, add peak name and gene name</span></span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true"></a>ov &lt;-<span class="st"> </span><span class="kw">as_tibble</span>(ov) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw">rename</span>(<span class="dt">peaknb =</span> queryHits, <span class="dt">genenb =</span> subjectHits) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">gene_name =</span> hg_genes_extended<span class="op">$</span>gene_name[genenb]) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">row =</span> <span class="kw">granges</span>(hemato)<span class="op">$</span>peak[peaknb])</span></code></pre></div>
<ul>
<li>Step 3:</li>
</ul>
<p>This step requires to transform the original <code>counts</code> SparseMatrix into a tidy
<code>data.table</code>, to enable fast <code>left_join</code> between (1) peak counts / cell and
(2) gene &lt;-&gt; peak relationship table (<code>ov</code>)</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true"></a><span class="co">## - Extract counts for the gene-associated peaks</span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true"></a>cnts &lt;-<span class="st"> </span><span class="kw">GetAssayData</span>(hemato, <span class="st">&#39;counts&#39;</span>)[<span class="kw">unique</span>(ov<span class="op">$</span>peaknb), ]</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true"></a></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true"></a><span class="co">## - Transform the cnts SparseMatrix into a tidy dense long data.frame</span></span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true"></a>tidy_cnts &lt;-<span class="st"> </span>broom<span class="op">::</span><span class="kw">tidy</span>(cnts) </span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true"></a></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true"></a><span class="co">## - Transform the dense long data.frame into a data.table for faster computation</span></span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true"></a>tidy_cnts &lt;-<span class="st"> </span>dtplyr<span class="op">::</span><span class="kw">lazy_dt</span>(tidy_cnts)</span>
<span id="cb9-9"><a href="#cb9-9" aria-hidden="true"></a></span>
<span id="cb9-10"><a href="#cb9-10" aria-hidden="true"></a><span class="co">## - For each peak/cell count value, associate the corresponding gene</span></span>
<span id="cb9-11"><a href="#cb9-11" aria-hidden="true"></a>joint_data &lt;-<span class="st"> </span><span class="kw">left_join</span>(tidy_cnts, ov, <span class="dt">by =</span> <span class="st">&#39;row&#39;</span>)</span>
<span id="cb9-12"><a href="#cb9-12" aria-hidden="true"></a><span class="kw">class</span>(joint_data)</span>
<span id="cb9-13"><a href="#cb9-13" aria-hidden="true"></a><span class="co">## WATCH OUT: You cannot see what is stored into `joint_data` as it </span></span>
<span id="cb9-14"><a href="#cb9-14" aria-hidden="true"></a><span class="co">## is a deferred data.table, i.e. it will only execute once you pipe it to </span></span>
<span id="cb9-15"><a href="#cb9-15" aria-hidden="true"></a><span class="co">## `as_tibble()`. Once you pipe it, the computation starts and takes a lot of time.</span></span></code></pre></div>
<ul>
<li>Step 4:</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a><span class="co">## - Group the joint_table by gene + cell and sum up all the counts</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true"></a>gene_averaged_cnts &lt;-<span class="st"> </span><span class="kw">group_by</span>(joint_data, column, gene_name) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true"></a><span class="st">    </span><span class="kw">summarize</span>(<span class="dt">n =</span> <span class="kw">sum</span>(value)) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true"></a><span class="st">    </span><span class="kw">as_tibble</span>()</span></code></pre></div>
<ul>
<li>Step 5:</li>
</ul>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true"></a><span class="co">## - To transform the data.table into SparseMatrix, it needs to be numerical</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true"></a>gene_averaged_cnts_num &lt;-<span class="st"> </span><span class="kw">data.matrix</span>(gene_averaged_cnts)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true"></a></span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true"></a><span class="co">## - Transform into sparse matrix </span></span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true"></a>gene_averaged_cnts_sparse &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">sparseMatrix</span>(</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true"></a>    <span class="dt">i =</span> gene_averaged_cnts_num[, <span class="st">&#39;gene_name&#39;</span>],</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true"></a>    <span class="dt">j =</span> gene_averaged_cnts_num[, <span class="st">&#39;column&#39;</span>],</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true"></a>    <span class="dt">x =</span> gene_averaged_cnts_num[, <span class="st">&#39;n&#39;</span>],</span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true"></a>)</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true"></a></span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true"></a><span class="co">## - Add back the actual cell and genes names</span></span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true"></a><span class="kw">colnames</span>(gene_averaged_cnts_sparse) &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">unique</span>(gene_averaged_cnts<span class="op">$</span>column))</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true"></a><span class="kw">rownames</span>(gene_averaged_cnts_sparse) &lt;-<span class="st"> </span><span class="kw">sort</span>(<span class="kw">unique</span>(gene_averaged_cnts<span class="op">$</span>gene_name))</span></code></pre></div>
<ul>
<li>Step 6:</li>
</ul>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true"></a><span class="co">## - Create new assay in hemato</span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true"></a>hemato[[<span class="st">&#39;GENE&#39;</span>]] &lt;-<span class="st"> </span><span class="kw">CreateAssayObject</span>(<span class="dt">counts =</span> gene_averaged_cnts_sparse)</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true"></a></span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true"></a><span class="co">## - Normalize this new assay with SCTransform</span></span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true"></a>hemato &lt;-<span class="st"> </span><span class="kw">SCTransform</span>(hemato, <span class="dt">assay =</span> <span class="st">&#39;GENE&#39;</span>)</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true"></a></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true"></a><span class="co">## - Save the resulting object!</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true"></a><span class="kw">saveRDS</span>(hemato, <span class="st">&#39;data/Homework_3/scATAC/hemato.rds&#39;</span>)</span></code></pre></div>
<p>Now the <code>Seurat</code> object contains 3 assays, the first one corresponding to the original scATACseq counts,
and the second one containing gene activity pseudocounts, and the third one containing <code>SCT</code>-normalized
data. One can now perform “gene differential expression”-like analysis using the <code>SCT</code> assay!</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true"></a><span class="co">## - Reduce dimensionality to visualize cells in 2D</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true"></a>hemato &lt;-<span class="st"> </span><span class="kw">RunPCA</span>(hemato) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">RunUMAP</span>(<span class="dt">reduction =</span> <span class="st">&#39;pca&#39;</span>, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">50</span>)</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true"></a></span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true"></a><span class="co">## - Find gene markers for Basophils</span></span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true"></a><span class="kw">DefaultAssay</span>(hemato) &lt;-<span class="st"> &quot;SCT&quot;</span></span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true"></a>p &lt;-<span class="st"> </span><span class="kw">FeaturePlot</span>(hemato, <span class="dt">features =</span> <span class="st">&#39;CD14&#39;</span>, <span class="dt">reduction =</span> <span class="st">&quot;umap&quot;</span>) <span class="co"># Or NCR1 (NKp46), CD8A, CD14, MS4A1</span></span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true"></a><span class="kw">Idents</span>(hemato) &lt;-<span class="st"> </span>hemato<span class="op">$</span>renamed_clusters</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true"></a>basophils_genes &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">FindMarkers</span>(</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true"></a>    <span class="dt">object =</span> hemato,</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true"></a>    <span class="dt">assay =</span> <span class="st">&#39;GENE&#39;</span>, </span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true"></a>    <span class="dt">ident.1 =</span> <span class="st">&#39;Baso.&#39;</span>,</span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true"></a>    <span class="dt">ident.2 =</span> <span class="kw">unique</span>(<span class="kw">Idents</span>(hemato)[<span class="kw">Idents</span>(hemato) <span class="op">!=</span><span class="st"> &#39;Baso.&#39;</span>]),</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true"></a>    <span class="dt">min.pct =</span> <span class="fl">0.15</span>, </span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true"></a>    <span class="dt">only.pos =</span> <span class="ot">TRUE</span></span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true"></a>)</span></code></pre></div>
</div>
<div id="chromvar-analysis" class="section level2">
<h2>3. chromVAR analysis</h2>
<div id="get-a-summarizedexperiment-of-scatacseq-counts-over-peaks" class="section level3">
<h3>Get a SummarizedExperiment of scATACseq counts over peaks</h3>
<p><code>chromVAR</code> works with raw counts stored as a <code>RangedSummarizedExperiment</code>. By now, you should be able to
easily extract the raw counts from a specific <code>Seurat</code> assay. Store it in a <code>RangedSummarizedExperiment</code>
(~ equivalent to <code>Seurat</code> assay but in <code>Bioconductor</code>).</p>
<p>What are the dimensions of the <code>RangedSummarizedExperiment</code> generated? What are the features and what are the columns?</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true"></a><span class="kw">library</span>(SummarizedExperiment)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true"></a>sumExp &lt;-<span class="st"> </span><span class="kw">SummarizedExperiment</span>(</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true"></a>    <span class="dt">assays =</span> <span class="kw">list</span>(<span class="st">&#39;counts&#39;</span> =<span class="st"> </span><span class="kw">GetAssayData</span>(hemato, <span class="dt">assay =</span> <span class="st">&#39;ATAC&#39;</span>, <span class="dt">slot =</span> <span class="st">&#39;counts&#39;</span>)), </span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true"></a>    <span class="dt">rowRanges =</span> <span class="kw">GetAssayData</span>(hemato, <span class="dt">assay =</span> <span class="st">&#39;ATAC&#39;</span>, <span class="dt">slot =</span> <span class="st">&#39;ranges&#39;</span>), </span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true"></a>    <span class="dt">colData =</span> hemato[[]]</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true"></a>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true"></a><span class="kw">dim</span>(sumExp)</span></code></pre></div>
</div>
<div id="add-gc-bias-to-peaks" class="section level3">
<h3>Add GC bias to peaks</h3>
<p>This step is important to correct GC bias associated with each peak.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true"></a>sumExp &lt;-<span class="st"> </span>chromVAR<span class="op">::</span><span class="kw">addGCBias</span>(sumExp, <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19<span class="op">::</span>BSgenome.Hsapiens.UCSC.hg19)</span></code></pre></div>
</div>
<div id="map-motifs-over-peaks" class="section level3">
<h3>Map motifs over peaks</h3>
<p>Using <code>motifmatchr</code> package, one can map a list of motifs (e.g. from JASPAR database)
over a <code>RangedSummarizedExperiment</code> (or anything which can be coerced to a <code>GRanges</code>, actually).</p>
<p>First, you need to import motifs from a public database in <code>R</code>. This can be done
(among other ways) with the <code>TFBSTools</code> pacakge.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true"></a><span class="co">## - Import motifs from JASPAR database</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true"></a>motifs &lt;-<span class="st"> </span>TFBSTools<span class="op">::</span><span class="kw">getMatrixSet</span>(</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true"></a>    JASPAR2020<span class="op">::</span>JASPAR2020,</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true"></a>    <span class="kw">list</span>(<span class="dt">species =</span> <span class="dv">9606</span>, <span class="dt">all_versions =</span> <span class="ot">FALSE</span>)</span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true"></a>)</span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true"></a><span class="kw">names</span>(motifs) &lt;-<span class="st"> </span>TFBSTools<span class="op">::</span><span class="kw">name</span>(motifs)</span></code></pre></div>
<p>Now map the subset of motifs of interest to the filtered peaks, using <code>motifmatchr::matchMotifs()</code>.</p>
<p>Read <code>matchMotifs()</code> documentation and run it on your <code>sumExp</code>. What is the class of the generated
object? Its dimensions? What are the features and what are the columns?</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true"></a><span class="kw">library</span>(BiocParallel)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true"></a><span class="kw">register</span>(<span class="kw">MulticoreParam</span>(<span class="dt">workers =</span> <span class="dv">16</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>))</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true"></a>motif_mappings &lt;-<span class="st"> </span>motifmatchr<span class="op">::</span><span class="kw">matchMotifs</span>(motifs, sumExp, <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg19<span class="op">::</span>BSgenome.Hsapiens.UCSC.hg19)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true"></a><span class="kw">class</span>(motif_mappingss)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true"></a><span class="kw">dim</span>(motif_mappings)</span></code></pre></div>
<p>Because there are so many motifs and so many peaks, and in interest of time,
we will perform <code>chromVAR</code> analysis on a subset of motifs only. You can manually filter the motifs
list to motifs that may be relevant in our context (e.g. <code>EOMES</code>, or <code>GATA2</code>, …).</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true"></a><span class="co">## - Filter to interesting motifs to keep analysis relatively quick</span></span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true"></a><span class="co">#       Feel free to select any TF of interest!</span></span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true"></a>motif_mappings_sub &lt;-<span class="st"> </span>motif_mappings[, <span class="kw">c</span>(<span class="st">&#39;EOMES&#39;</span>, <span class="st">&#39;GATA2&#39;</span>, <span class="st">&#39;IRF8&#39;</span>, <span class="st">&#39;EBF1&#39;</span>, <span class="st">&#39;SPI1&#39;</span>, <span class="st">&#39;TCF3&#39;</span>, <span class="st">&#39;CEBPA&#39;</span>)]</span></code></pre></div>
</div>
<div id="search-for-motifs-with-high-deviation-of-mapping-compared-to-background" class="section level3">
<h3>Search for motifs with high deviation of mapping compared to background</h3>
<p><code>chromVAR</code>’s <code>computeDeviations()</code> function combines (1) the peak counts / cell (stored here in <code>sumExp</code>) and
(2) the TF mapping over each peak (stored here in <code>motif_mappings</code>).
to assess the mapping deviation for each TF over each cell compared to other cells.</p>
<p>Run <code>computeDeviations()</code> on the set of peaks and motifs. What is the class of the generated
object? Its dimensions? What are the features and what are the columns?</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true"></a><span class="co">## - Find background signal</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true"></a>bg &lt;-<span class="st"> </span>chromVAR<span class="op">::</span><span class="kw">getBackgroundPeaks</span>(<span class="dt">object =</span> sumExp)</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true"></a><span class="co">## - For each motif, compute its mapping deviation over the filtered peaks</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true"></a>motif_deviations &lt;-<span class="st"> </span>chromVAR<span class="op">::</span><span class="kw">computeDeviations</span>(<span class="dt">object =</span> sumExp, <span class="dt">annotations =</span> motif_mappings_sub, <span class="dt">background_peaks =</span> bg)</span></code></pre></div>
</div>
<div id="check-tf-motif-enrichment-in-different-cell-types" class="section level3">
<h3>Check TF motif enrichment in different cell types</h3>
<p>The <code>motif_deviations</code> object can be added as a new assay to the
<code>hemato</code> <code>Seurat</code> object. This way, one can rely on <code>Seurat</code>-based
plotting functions to plot cells in their preferred dimensional space, and color
them using motif deviation scores computed with <code>chromVAR</code>.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true"></a>hemato[[<span class="st">&#39;MOTIF&#39;</span>]] &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">CreateAssayObject</span>(<span class="dt">counts =</span> chromVAR<span class="op">::</span><span class="kw">deviationScores</span>(motif_deviations))</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true"></a><span class="kw">DefaultAssay</span>(hemato) &lt;-<span class="st"> &#39;MOTIF&#39;</span></span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true"></a>list_p &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="kw">rownames</span>(hemato), <span class="cf">function</span>(motif) {</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true"></a>    <span class="kw">FeaturePlot</span>(hemato, <span class="dt">features =</span> motif, <span class="dt">reduction =</span> <span class="st">&quot;umap&quot;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true"></a><span class="st">        </span><span class="kw">scale_colour_gradientn</span>(</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true"></a>            <span class="dt">colors =</span> <span class="kw">c</span>(<span class="st">&#39;#190886&#39;</span>, <span class="st">&#39;#6F07F8&#39;</span>, <span class="st">&#39;#F954A5&#39;</span>, <span class="st">&#39;#FF9D66&#39;</span>, <span class="st">&#39;#edf118&#39;</span>),</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true"></a>            <span class="dt">limits =</span> <span class="kw">c</span>(<span class="op">-</span><span class="dv">5</span>, <span class="dv">5</span>), <span class="dt">oob =</span> scales<span class="op">::</span>squish</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true"></a>        ) <span class="op">+</span><span class="st"> </span></span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true"></a><span class="st">        </span><span class="kw">theme</span>(<span class="dt">aspect.ratio =</span> <span class="dv">1</span>)</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true"></a>})</span>
<span id="cb20-11"><a href="#cb20-11" aria-hidden="true"></a>list_p[[<span class="kw">length</span>(list_p) <span class="op">+</span><span class="st"> </span><span class="dv">1</span>]] &lt;-<span class="st"> </span><span class="kw">DimPlot</span>(hemato, <span class="dt">group.by =</span> <span class="st">&#39;renamed_clusters&#39;</span>, <span class="dt">reduction =</span> <span class="st">&quot;umap&quot;</span>) <span class="op">+</span><span class="st"> </span><span class="kw">theme</span>(<span class="dt">aspect.ratio =</span> <span class="dv">1</span>)</span>
<span id="cb20-12"><a href="#cb20-12" aria-hidden="true"></a>p &lt;-<span class="st"> </span>cowplot<span class="op">::</span><span class="kw">plot_grid</span>(<span class="dt">plotlist =</span> list_p)</span></code></pre></div>
</div>
<div id="compare-gene-activity-and-tf-activity" class="section level3">
<h3>Compare gene activity and TF activity</h3>
<p>Plot side-by-side:</p>
<ul>
<li>A UMAP projection showing where cells along the B lineage are located</li>
<li>A UMAP projection showing <code>EBF1</code> gene activity per cell</li>
<li>A UMAP projection showing <code>EBF1</code> motif enrichment (i.e. TF activity) per cell</li>
</ul>
<div class="sourceCode" id="cb21"><pre class="sourceCode txt"><code class="sourceCode default"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true"></a>labels &lt;- c(</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true"></a>    &#39;HSC&#39;, </span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true"></a>    &#39;MPP&#39;,</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true"></a>    &#39;LMPP&#39;, </span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true"></a>    &#39;CLP&#39;, </span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true"></a>    &#39;Pro-B&#39;, </span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true"></a>    &#39;Pre-B&#39;, </span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true"></a>    &#39;Nai. B&#39;, </span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true"></a>    &#39;Mem. B&#39;, </span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true"></a>    &#39;Plasma&#39;</span>
<span id="cb21-11"><a href="#cb21-11" aria-hidden="true"></a>)</span>
<span id="cb21-12"><a href="#cb21-12" aria-hidden="true"></a>hemato$b_cells_annots &lt;- ifelse(hemato$renamed_clusters %in% labels, hemato$?...?, NA)</span>
<span id="cb21-13"><a href="#cb21-13" aria-hidden="true"></a>p1 &lt;- DimPlot(hemato, group.by = &#39;?...?&#39;, reduction = &quot;umap&quot;) + theme(aspect.ratio = 1) + scale_colour_discrete(na.value = &#39;#aaaaaa&#39;)</span>
<span id="cb21-14"><a href="#cb21-14" aria-hidden="true"></a>DefaultAssay(hemato) &lt;- &#39;SCT&#39;</span>
<span id="cb21-15"><a href="#cb21-15" aria-hidden="true"></a>p2 &lt;- FeaturePlot(hemato, features = ?...?, reduction = &quot;umap&quot;) + </span>
<span id="cb21-16"><a href="#cb21-16" aria-hidden="true"></a>        scale_colour_gradientn(</span>
<span id="cb21-17"><a href="#cb21-17" aria-hidden="true"></a>            colors = c(&#39;#dadad8&#39;, &#39;#f1eca3&#39;, &#39;#e0c067&#39;, &#39;#db8035&#39;, &#39;#a3160c&#39;),</span>
<span id="cb21-18"><a href="#cb21-18" aria-hidden="true"></a>            limits = c(0, 3), oob = scales::squish</span>
<span id="cb21-19"><a href="#cb21-19" aria-hidden="true"></a>        ) + </span>
<span id="cb21-20"><a href="#cb21-20" aria-hidden="true"></a>        theme(aspect.ratio = 1)</span>
<span id="cb21-21"><a href="#cb21-21" aria-hidden="true"></a>DefaultAssay(hemato) &lt;- &#39;MOTIF&#39;</span>
<span id="cb21-22"><a href="#cb21-22" aria-hidden="true"></a>p3 &lt;- FeaturePlot(hemato, features = ?...?, reduction = &quot;umap&quot;) + </span>
<span id="cb21-23"><a href="#cb21-23" aria-hidden="true"></a>        scale_colour_gradientn(</span>
<span id="cb21-24"><a href="#cb21-24" aria-hidden="true"></a>            colors = c(&#39;#190886&#39;, &#39;#6F07F8&#39;, &#39;#F954A5&#39;, &#39;#FF9D66&#39;, &#39;#edf118&#39;),</span>
<span id="cb21-25"><a href="#cb21-25" aria-hidden="true"></a>            limits = c(-5, 5), oob = scales::squish</span>
<span id="cb21-26"><a href="#cb21-26" aria-hidden="true"></a>        ) + </span>
<span id="cb21-27"><a href="#cb21-27" aria-hidden="true"></a>        theme(aspect.ratio = 1)</span>
<span id="cb21-28"><a href="#cb21-28" aria-hidden="true"></a>p &lt;- cowplot::plot_grid(p1, p2, p3, ncol = 3)</span></code></pre></div>
</div>
</div>


<footer class="footline">
	
</footer>

        
        </div>
        

      </div>

    <div id="navigation">
        
        

        
            
            
                
                    
                    
                
                

                    
                    
                        
                    
                    

                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        

        


	 
	 
		
			<a class="nav nav-prev" href="/scATACseq-workshop/homeworks/homework_2/" title="Homework 2"> <i class="fa fa-chevron-left"></i></a>
		
		
			<a class="nav nav-next" href="/scATACseq-workshop/extra_resources/" title="Extra resources" style="margin-right: 0px;"><i class="fa fa-chevron-right"></i></a>
		
	
    </div>

    </section>

    <div style="left: -1000px; overflow: scroll; position: absolute; top: -1000px; border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;">
      <div style="border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;"></div>
    </div>
    <script src="/scATACseq-workshop/js/clipboard.min.js?1628787916"></script>
    <script src="/scATACseq-workshop/js/perfect-scrollbar.min.js?1628787916"></script>
    <script src="/scATACseq-workshop/js/perfect-scrollbar.jquery.min.js?1628787916"></script>
    <script src="/scATACseq-workshop/js/jquery.sticky.js?1628787916"></script>
    <script src="/scATACseq-workshop/js/featherlight.min.js?1628787916"></script>
    <script src="/scATACseq-workshop/js/highlight.pack.js?1628787916"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script src="/scATACseq-workshop/js/modernizr.custom-3.6.0.js?1628787916"></script>
    <script src="/scATACseq-workshop/js/learn.js?1628787916"></script>
    <script src="/scATACseq-workshop/js/hugo-learn.js?1628787916"></script>
    
        
            <script src="/scATACseq-workshop/mermaid/mermaid.js?1628787916"></script>
        
        <script>
            mermaid.initialize({ startOnLoad: true });
        </script>
    
    <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-105947713-1', 'auto');
  ga('send', 'pageview');

</script>

  </body>
</html>

