<!DOCTYPE html>
<html lang="en" class="js csstransforms3d">
  <head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <meta name="generator" content="Hugo 0.81.0" />
    <meta name="description" content="Analysis of bulk and single-cell ATAC-seq datasets">
<meta name="author" content="Jacques Serizay">

    <link rel="icon" href="/scATACseq-workshop/images/favicon.png" type="image/png">

    <title>Homework 1 :: Analysis of bulk and single-cell ATAC-seq datasets</title>

    
    <link href="/scATACseq-workshop/css/nucleus.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/fontawesome-all.min.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/hybrid.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/featherlight.min.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/perfect-scrollbar.min.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/auto-complete.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/atom-one-dark-reasonable.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/theme.css?1630582900" rel="stylesheet">
    <link href="/scATACseq-workshop/css/hugo-theme.css?1630582900" rel="stylesheet">
    
    <link href="/scATACseq-workshop/css/theme-red.css?1630582900" rel="stylesheet">
    
    <link href="/scATACseq-workshop/css/custom_css.css?1630582900" rel="stylesheet">

    <script src="/scATACseq-workshop/js/jquery-3.3.1.min.js?1630582900"></script>

    <style>
      :root #header + #content > #left > #rlblock_left{
          display:none !important;
      }
      
        :not(pre) > code + span.copy-to-clipboard {
            display: none;
        }
      
    </style>
    
  </head>
  <body class="" data-url="/scATACseq-workshop/homeworks/homework_1/">
    <nav id="sidebar" class="">



  <div id="header-wrapper">
    <div id="header">
      <a id="logo">



</a>

    </div>
    
  </div>
  

    <div class="highlightable">
    <ul class="topics">

        
          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/introduction/" title=" Bulk and single-cell ATAC-seq analysis" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/introduction/">
           Bulk and single-cell ATAC-seq analysis
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/introduction/01_program/" title="1. Program" class="dd-item ">
        <a href="/scATACseq-workshop/introduction/01_program/">
        1. Program
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/introduction/02_prerequisites/" title="2. Prerequisites and required machine configuration" class="dd-item ">
        <a href="/scATACseq-workshop/introduction/02_prerequisites/">
        2. Prerequisites and required machine configuration
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/hands_on_sessions/" title="1. Hands-on sessions" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/hands_on_sessions/">
          1. Hands-on sessions
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/hands_on_sessions/day1/" title="Day 01" class="dd-item ">
        <a href="/scATACseq-workshop/hands_on_sessions/day1/">
        Day 01
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/hands_on_sessions/day2/" title="Day 02" class="dd-item ">
        <a href="/scATACseq-workshop/hands_on_sessions/day2/">
        Day 02
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/hands_on_sessions/day3/" title="Day 03" class="dd-item ">
        <a href="/scATACseq-workshop/hands_on_sessions/day3/">
        Day 03
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/homeworks/" title="2. Homeworks" class="dd-item
        parent
        
        
        ">
      <a href="/scATACseq-workshop/homeworks/">
          2. Homeworks
          
      </a>
      
      
        <ul>
          
          
          

        
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/homeworks/homework_1/" title="Homework 1" class="dd-item active">
        <a href="/scATACseq-workshop/homeworks/homework_1/">
        Homework 1
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/homeworks/homework_2/" title="Homework 2" class="dd-item ">
        <a href="/scATACseq-workshop/homeworks/homework_2/">
        Homework 2
        
        </a>
    </li>
     
  
 

            
          
            
            




 
  
    
      <li data-nav-id="/scATACseq-workshop/homeworks/homework_3/" title="Homework 3" class="dd-item ">
        <a href="/scATACseq-workshop/homeworks/homework_3/">
        Homework 3
        
        </a>
    </li>
     
  
 

            
          
        
        </ul>
      
    </li>
  
 

          
          




 
  
    
    <li data-nav-id="/scATACseq-workshop/extra_resources/" title="Extra resources" class="dd-item
        
        
        
        ">
      <a href="/scATACseq-workshop/extra_resources/">
          Extra resources
          
      </a>
      
      
    </li>
  
 

          
        
    </ul>

    
    

    
    <section id="footer">
      <center>

    <a href="https://github.com/js2264/scATACseq-workshop">Check this website source on Github</a>
    <p>Created by J. Serizay</p>
    <p>Copyright 2021</p>

</center>

<script async defer src="https://buttons.github.io/buttons.js"></script>

    </section>
  </div>
</nav>




        <section id="body">
        <div id="overlay"></div>
        <div class="padding highlightable">
        <div id="head-tags">
        
        </div>
        
        <div id="body-inner">
          
            <h1>
              
              Homework 1
            </h1>
          

        



<script src="/scATACseq-workshop/rmarkdown-libs/header-attrs/header-attrs.js"></script>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<div id="goals" class="section level2">
<h2>Goals</h2>
<ol style="list-style-type: decimal">
<li>Process bulk ATAC-seq for PBMCs (<code>GSE111013</code>), perform QCs and compare them to the bulk ATAC-seq QCs seen during demonstration session 1</li>
</ol>
<blockquote>
<p>We recommend focusing on few datasets from healthy donors, e.g. <code>SRR6762787</code>, <code>SRR6762790</code> and <code>SRR6762793</code>.</p>
</blockquote>
<ol start="2" style="list-style-type: decimal">
<li>Split bam file from human 5,000 PBMCs scATACseq into bams / cell type using <code>sinto</code></li>
</ol>
<blockquote>
<p>Processed data can be obtained directly from <a href="https://cf.10xgenomics.com/samples/cell-atac/2.0.0/atac_pbmc_5k_nextgem/atac_pbmc_5k_nextgem_web_summary.html">10X Genomics</a></p>
</blockquote>
<ol start="3" style="list-style-type: decimal">
<li><p>Make some celltype-specific pseudo-bulk bigwig tracks</p></li>
<li><p>Compare bulk and pseudo-bulk tracks in IGV; try to infer cell type for some of the cell clusters in scATACseq</p></li>
</ol>
</div>
<div id="process-bulk-atac-seq-from-human-b-cells-cd4-and-cd8-t-cells" class="section level2">
<h2>1. Process bulk ATAC-seq from human B cells, CD4+ and CD8+ T cells</h2>
<p>Data was generated in <a href="https://doi.org/10.1038/s41467-019-14081-6">this study: DOI 10.1038/s41467-019-14081-6</a>.</p>
<div id="download-data" class="section level3">
<h3>Download data</h3>
<p>First, we need to fetch to raw reads. You can go to the <a href="https://trace.ncbi.nlm.nih.gov/Traces/study/?acc=GSE111013&amp;o=acc_s%3Aa">SRA Run Selector</a>
and search for your <code>GSE</code> of interest. Try and download reads for samples <code>SRR6762787</code>, <code>SRR6762790</code> and <code>SRR6762793</code>.</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="ex">https</span>://trace.ncbi.nlm.nih.gov/Traces/study/?acc=GSE111013<span class="kw">&amp;</span><span class="va">o=</span>acc_s<span class="ex">%3Aa</span></span></code></pre></div>
<p>Alternatively, I like to use another web-based approach: the <a href="https://sra-explorer.info/#">sra-explorer</a> website.
It allows you to search for individual experiments using GEO ids, SRA ids, and much more, and reports download links for
your samples of interest.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="fu">mkdir</span> -p data/Homework_1/fastq</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true"></a><span class="ex">curl</span> -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR676/007/SRR6762787/SRR6762787.fastq.gz -o data/Homework_1/fastq/Bcells.fq.gz</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true"></a><span class="ex">curl</span> -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR676/000/SRR6762790/SRR6762790.fastq.gz -o data/Homework_1/fastq/CD4Tcells.fq.gz</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true"></a><span class="ex">curl</span> -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR676/003/SRR6762793/SRR6762793.fastq.gz -o data/Homework_1/fastq/CD8Tcells.fq.gz</span></code></pre></div>
<p>Yet another approach is to use the recent <code>ffq</code> tool. It’s quite effective, though it is not very fast and one
eventually has to manually download each link…</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a><span class="ex">conda</span> install -c conda-forge -c bioconda ffq</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a><span class="ex">ffq</span> SRR6762787 SRR6762790 SRR6762793 <span class="kw">|</span> <span class="fu">grep</span> <span class="st">&quot;url&quot;</span></span></code></pre></div>
</div>
<div id="map-raw-data-onto-grch38" class="section level3">
<h3>Map raw data onto GRCh38</h3>
<p>Note that in the next step (focusing on scATACseq), the reads were aligned to the <code>GRCh38</code> reference. For tracks to be comparable,
we will align bulk ATAC-seq data onto <code>GRCh38</code> reference as well.</p>
<blockquote>
<p>Build GRCh38 genome reference</p>
</blockquote>
<p>If you don’t have a human reference yet, you will need to build one for Bowtie2!
We recommend fetching assembly and annotations from Ensembl.
This step will take ~ 15 min at least, but you’ll only have to do it once!.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a><span class="fu">mkdir</span> ~/genomes/GRCh38/</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true"></a><span class="ex">curl</span> -L http://ftp.ensembl.org/pub/release-98/fasta/homo_sapiens/dna/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz -o ~/genomes/GRCh38/GRCh38.fa.gz</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true"></a><span class="fu">gunzip</span> ~/genomes/GRCh38/*</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true"></a><span class="ex">bowtie2-build</span> --large-index --threads 16 ~/genomes/GRCh38/GRCh38.fa ~/genomes/GRCh38/GRCh38.fa</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true"></a><span class="ex">samtools</span> faidx ~/genomes/GRCh38/GRCh38.fa</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true"></a><span class="fu">cut</span> -f1,2 ~/genomes/GRCh38/GRCh38.fa.fai <span class="op">&gt;</span> ~/genomes/GRCh38/GRCh38.sizes.genome</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true"></a><span class="fu">cat</span> ~/genomes/GRCh38/GRCh38.sizes.genome <span class="kw">\</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true"></a>    <span class="kw">|</span> <span class="fu">sed</span> -E <span class="st">&#39;s/^([0-9]+|[XY])/chr\1/&#39;</span> <span class="kw">\</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true"></a>    <span class="kw">|</span> <span class="fu">sed</span> -E <span class="st">&#39;s/^MT/chrM/&#39;</span> <span class="kw">\</span></span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true"></a>    <span class="op">&gt;</span> <span class="ex">~/genomes/GRCh38/GRCh38.sizes.genome.modified</span></span></code></pre></div>
<blockquote>
<p>Align reads to GRCh38 reference</p>
</blockquote>
<p>Note that in this study, the ATAC-seq experiments were sequenced in single-end mode.
This is not a major issue, though it limits QCs as fragment sizes are not available.</p>
<p>Try and adapt the different mapping steps used during the demonstration session
to map single-end reads. Commands for trimming, mapping, filtering and track
generation will all be affected !! Check the manual for each command if you are
unsure of some flags.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a><span class="fu">mkdir</span> data/Homework_1/bam</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true"></a><span class="fu">mkdir</span> data/Homework_1/bw</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true"></a><span class="va">GENOME=</span>~/genomes/GRCh38/GRCh38.fa</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true"></a><span class="va">CPU=</span>16</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true"></a></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true"></a><span class="kw">for</span> <span class="ex">SAMPLE</span> in Bcells CD4Tcells CD8Tcells</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true"></a><span class="kw">do</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true"></a></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true"></a>    <span class="va">READS=</span>data/Homework_1/fastq/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.fq.gz</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true"></a>    <span class="va">TRIMMEDREADS=</span>data/Homework_1/fastq/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>_trimmed.fq.gz</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true"></a>    <span class="va">SAMFILE=</span>data/Homework_1/bam/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.sam</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true"></a>    <span class="va">FILTEREDBAM=</span>data/Homework_1/bam/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.bam</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true"></a>    <span class="va">TRACK=</span>data/Homework_1/bw/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.bw</span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true"></a></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true"></a>    <span class="co">## -- Trim reads and run QC</span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true"></a>    <span class="ex">trim_galore</span> --fastqc --cores 4 --output_dir data/Homework_1/fastq/ <span class="st">&quot;</span><span class="va">${READS}</span><span class="st">&quot;</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true"></a></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true"></a>    <span class="co">## -- Map reads in single-end mode</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true"></a>    <span class="ex">bowtie2</span> --threads <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> -x <span class="st">&quot;</span><span class="va">${GENOME}</span><span class="st">&quot;</span> -U <span class="st">&quot;</span><span class="va">${TRIMMEDREADS}</span><span class="st">&quot;</span> <span class="op">&gt;</span> <span class="st">&quot;</span><span class="va">${SAMFILE}</span><span class="st">&quot;</span></span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true"></a></span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true"></a>    <span class="co">## -- Filter reads </span></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true"></a>    <span class="ex">samtools</span> sort -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> -T <span class="st">&quot;</span><span class="va">${SAMFILE}</span><span class="st">&quot;</span>_sorting <span class="st">&quot;</span><span class="va">${SAMFILE}</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true"></a>    <span class="ex">samtools</span> markdup -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> -r -T <span class="st">&quot;</span><span class="va">${SAMFILE}</span><span class="st">&quot;</span>_markdup - - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true"></a>    <span class="ex">samtools</span> view -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> -F 4 -q 10 -1 -b - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true"></a>    <span class="ex">samtools</span> sort -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -l 9 -T <span class="st">&quot;</span><span class="va">${SAMFILE}</span><span class="st">&quot;</span>_sorting2 -o <span class="st">&quot;</span><span class="va">${FILTEREDBAM}</span><span class="st">&quot;</span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true"></a>    <span class="ex">samtools</span> index -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> <span class="st">&quot;</span><span class="va">${FILTEREDBAM}</span><span class="st">&quot;</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true"></a></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true"></a><span class="kw">done</span></span></code></pre></div>
<p>Because we have single-end reads only, we recommend to rely on <code>macs2</code> to
predict the fragment size model, rather than <code>bamCoverage</code>. <code>macs2</code> can compute both
peaks and normalized coverage. Check <code>macs2 callpeak</code> help to see which flags are important to generate
a depth-normalized coverage track.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a><span class="kw">for</span> <span class="ex">SAMPLE</span> in CD4Tcells CD8Tcells</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a><span class="kw">do</span></span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true"></a>    <span class="co">## -- Generate track and peaks simultaneously with MACS2</span></span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true"></a>    <span class="ex">macs2</span> callpeak <span class="kw">\</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true"></a>        <span class="ex">-t</span> data/Homework_1/bam/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.bam <span class="kw">\</span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true"></a>        <span class="ex">--gsize</span> hs <span class="kw">\</span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true"></a>        <span class="ex">--outdir</span> data/Homework_1/bw <span class="kw">\</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true"></a>        <span class="ex">--name</span> <span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span> <span class="kw">\</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true"></a>        <span class="ex">--bdg</span> --SPMR </span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true"></a>    </span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true"></a>    <span class="co">## -- Convert macs2 bdg track to bigwig </span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true"></a>    <span class="fu">sed</span> -i <span class="st">&#39;s/,/./&#39;</span> data/Homework_1/bw/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>_treat_pileup.bdg</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true"></a>    <span class="ex">bedGraphToBigWig</span> data/Homework_1/bw/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>_treat_pileup.bdg ~/genomes/GRCh38/GRCh38.sizes.genome data/Homework_1/bw/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.bw</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true"></a></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true"></a><span class="kw">done</span></span></code></pre></div>
</div>
<div id="qc-the-3-different-samples-in-r" class="section level3">
<h3>QC the 3 different samples in R</h3>
<blockquote>
<p>Get human gene features</p>
</blockquote>
<p>We can define TSSs and promoters manually from gene annotations.
We recommend you fetch gene annotations using <code>AnnotationHub</code>.
Keep in mind that you obtained the <code>GRCh38</code> reference from Ensembl <code>v98</code>!!
Once the genebodies are defined, you can build <code>TSSs</code> and
promoters (e.g. -/+ 2000 bp centered around the <code>TSS</code>).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a><span class="kw">library</span>(AnnotationDbi)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true"></a><span class="kw">library</span>(plyranges)</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true"></a><span class="kw">library</span>(annotatr)</span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true"></a></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true"></a><span class="co">## -- Import genomic features</span></span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true"></a>AnnotationHub<span class="op">::</span><span class="kw">query</span>(AnnotationHub<span class="op">::</span><span class="kw">AnnotationHub</span>(), <span class="kw">c</span>(<span class="st">&#39;ensembl&#39;</span>, <span class="st">&#39;98&#39;</span>, <span class="st">&#39;GRCh38&#39;</span>, <span class="st">&#39;gtf&#39;</span>))</span>
<span id="cb7-8"><a href="#cb7-8" aria-hidden="true"></a>hg_genes &lt;-<span class="st"> </span>AnnotationHub<span class="op">::</span><span class="kw">AnnotationHub</span>()[[<span class="st">&#39;AH75393&#39;</span>]] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-9"><a href="#cb7-9" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(type <span class="op">==</span><span class="st"> &#39;gene&#39;</span>, gene_biotype <span class="op">==</span><span class="st"> &#39;protein_coding&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-10"><a href="#cb7-10" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(<span class="op">!</span>(<span class="kw">grepl</span>(<span class="st">&#39;GL|KI&#39;</span>, seqnames)))</span>
<span id="cb7-11"><a href="#cb7-11" aria-hidden="true"></a>hg_TSSs &lt;-<span class="st"> </span>hg_genes <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-12"><a href="#cb7-12" aria-hidden="true"></a><span class="st">    </span><span class="kw">anchor_start</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-13"><a href="#cb7-13" aria-hidden="true"></a><span class="st">    </span><span class="kw">resize</span>(<span class="dt">width =</span> <span class="dv">1</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-14"><a href="#cb7-14" aria-hidden="true"></a><span class="st">    </span><span class="kw">unanchor</span>()</span>
<span id="cb7-15"><a href="#cb7-15" aria-hidden="true"></a>hg_promoters &lt;-<span class="st"> </span>hg_TSSs <span class="op">%&gt;%</span></span>
<span id="cb7-16"><a href="#cb7-16" aria-hidden="true"></a><span class="st">    </span><span class="kw">flank_upstream</span>(<span class="dv">4000</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-17"><a href="#cb7-17" aria-hidden="true"></a><span class="st">    </span><span class="kw">shift_downstream</span>(<span class="dv">2000</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb7-18"><a href="#cb7-18" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">ID =</span> <span class="kw">paste0</span>(<span class="st">&#39;prom_&#39;</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">n</span>()))</span></code></pre></div>
<blockquote>
<p>Import the 3 bam files in R</p>
</blockquote>
<p><code>VplotR</code> provides the <code>importPEBamFiles()</code> function, which wraps
<code>GenomicAlignments::readGAlignmentPairs()</code> to import paired-end ATAC-seq data
as fragments and shift them.
In this study, since we are working with single-end data, we cannot rely on it.
Try and import reads “manually”. Search for the right function from <code>GenomicAlignments</code> package!</p>
<p>Try also to import <code>macs2</code> peaks in R. You can leverage the convenient <code>rtracklayer</code> package for this!</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true"></a><span class="kw">library</span>(GenomicAlignments)</span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true"></a><span class="kw">library</span>(rtracklayer)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true"></a>samples &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&#39;Bcells&#39;</span>, <span class="st">&#39;CD4Tcells&#39;</span>, <span class="st">&#39;CD8Tcells&#39;</span>)</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true"></a>bams &lt;-<span class="st"> </span><span class="kw">lapply</span>(samples, <span class="cf">function</span>(sample) {</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true"></a>    bfile &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;data/Homework_1/bam/&#39;</span>, sample, <span class="st">&#39;.bam&#39;</span>)</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true"></a>    ga &lt;-<span class="st"> </span><span class="kw">readGAlignments</span>(bfile)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true"></a>    gr &lt;-<span class="st"> </span><span class="kw">as</span>(ga, <span class="st">&#39;GRanges&#39;</span>)</span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true"></a>    <span class="kw">return</span>(gr)</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true"></a>})</span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true"></a><span class="kw">lengths</span>(bams)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true"></a>peaks &lt;-<span class="st"> </span><span class="kw">lapply</span>(samples, <span class="cf">function</span>(sample) {</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true"></a>    peaks &lt;-<span class="st"> </span><span class="kw">import</span>(<span class="kw">paste0</span>(<span class="st">&#39;data/Homework_1/bw/&#39;</span>, sample, <span class="st">&#39;_peaks.narrowPeak&#39;</span>))</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true"></a>    peak_summits &lt;-<span class="st"> </span><span class="kw">import</span>(<span class="kw">paste0</span>(<span class="st">&#39;data/Homework_1/bw/&#39;</span>, sample, <span class="st">&#39;_summits.bed&#39;</span>))</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true"></a>    peaks<span class="op">$</span>summit &lt;-<span class="st"> </span><span class="kw">start</span>(peak_summits)</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true"></a>    <span class="kw">return</span>(peaks)</span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true"></a>})</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true"></a><span class="kw">lengths</span>(peaks)</span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true"></a><span class="kw">names</span>(bams) &lt;-<span class="st"> </span><span class="kw">names</span>(peaks) &lt;-<span class="st"> </span>samples</span></code></pre></div>
<blockquote>
<p>Compute FRiP</p>
</blockquote>
<p>FRiP is the easiest metric to compute in ATAC QC: one just needs to identify
which reads are overlapping with annotated peaks. To see whether a <code>GRanges</code>
query overlaps with a <code>GRanges</code> subject, the <code>findOverlaps()</code> function is
very useful is relatively fast!</p>
<p>When counting the total number of reads overlapping peaks,
careful with reads which may overlap several peaks simultaneously!!</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true"></a>Bcells_reads &lt;-<span class="st"> </span>bams[[<span class="st">&#39;Bcells&#39;</span>]]</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true"></a><span class="co"># nInPeaks &lt;- length(queryHits(findOverlaps(Bcells_reads, peaks[[&#39;Bcells&#39;]])))</span></span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true"></a>nInPeaks &lt;-<span class="st"> </span><span class="kw">length</span>(<span class="kw">unique</span>(<span class="kw">queryHits</span>(<span class="kw">findOverlaps</span>(Bcells_reads, peaks[[<span class="st">&#39;Bcells&#39;</span>]]))))</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true"></a>nTot &lt;-<span class="st"> </span><span class="kw">length</span>(Bcells_reads)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true"></a>nInPeaks<span class="op">/</span>nTot</span></code></pre></div>
<blockquote>
<p>Check distance from Tn5 “cut” sites to TSS</p>
</blockquote>
<p>The “cut” site is at the first base of the single-end read.
We can calculate the distance from the start of each ATAC read to the nearest TSS,
and check the cumulative distribution of distances.</p>
<p>Start first by trying to manually compute the distance using the standard
<code>distanceToNearest()</code> function from <code>IRanges</code>, then build a data.frame and
plot results manually.</p>
<p>If you feel confident handling <code>GRanges</code> objects, try to leverage functions from
<code>plyranges</code> to perform as much of the computation in a tidy workflow and
building on top of the <code>GRanges</code> structure.</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a><span class="co">## -- Manual way</span></span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true"></a>Bcells_reads &lt;-<span class="st"> </span>bams[[<span class="st">&#39;Bcells&#39;</span>]]</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true"></a>Bcells_reads &lt;-<span class="st"> </span><span class="kw">resize</span>(Bcells_reads, <span class="dt">width =</span> <span class="dv">1</span>, <span class="dt">fix =</span> <span class="st">&#39;start&#39;</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true"></a>dists &lt;-<span class="st"> </span><span class="kw">distanceToNearest</span>(Bcells_reads, hg_TSSs, <span class="dt">ignore.strand =</span> <span class="ot">TRUE</span>)</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true"></a>dists &lt;-<span class="st"> </span><span class="kw">elementMetadata</span>(dists)<span class="op">$</span>distance</span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true"></a>dists &lt;-<span class="st"> </span><span class="kw">round</span>(dists<span class="op">/</span><span class="dv">10</span>, <span class="dv">0</span>)<span class="op">*</span><span class="dv">10</span></span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true"></a>dists &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">table</span>(dists))</span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true"></a>dists<span class="op">$</span>cumsum &lt;-<span class="st"> </span><span class="kw">cumsum</span>(dists<span class="op">$</span>Freq)</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true"></a>dists<span class="op">$</span>log10dists &lt;-<span class="st"> </span><span class="kw">log10</span>(<span class="kw">as.numeric</span>(<span class="kw">as.character</span>(dists<span class="op">$</span>dists)))</span>
<span id="cb10-10"><a href="#cb10-10" aria-hidden="true"></a><span class="kw">plot</span>(<span class="dt">x =</span> dists<span class="op">$</span>log10dists, <span class="dt">y =</span> dists<span class="op">$</span>cumsum)</span>
<span id="cb10-11"><a href="#cb10-11" aria-hidden="true"></a></span>
<span id="cb10-12"><a href="#cb10-12" aria-hidden="true"></a><span class="co">## -- &quot;Tidy&quot; way, using `plyranges` and `tidyverse` functions</span></span>
<span id="cb10-13"><a href="#cb10-13" aria-hidden="true"></a>df&lt;-bams[[<span class="st">&#39;Bcells&#39;</span>]] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb10-14"><a href="#cb10-14" aria-hidden="true"></a><span class="st">    </span><span class="kw">resize</span>(<span class="dt">width =</span> <span class="dv">1</span>, <span class="dt">fix =</span> <span class="st">&#39;start&#39;</span>) <span class="op">%&gt;%</span></span>
<span id="cb10-15"><a href="#cb10-15" aria-hidden="true"></a><span class="st">    </span><span class="kw">add_nearest_distance</span>(hg_TSSs) <span class="op">%&gt;%</span></span>
<span id="cb10-16"><a href="#cb10-16" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">distance =</span> <span class="kw">round</span>(distance<span class="op">/</span><span class="dv">10</span>, <span class="dv">0</span>)<span class="op">*</span><span class="dv">10</span>) <span class="op">%&gt;%</span></span>
<span id="cb10-17"><a href="#cb10-17" aria-hidden="true"></a><span class="st">    </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb10-18"><a href="#cb10-18" aria-hidden="true"></a><span class="st">    </span><span class="kw">count</span>(distance) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb10-19"><a href="#cb10-19" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">cumsum =</span> <span class="kw">cumsum</span>(n), <span class="dt">pct =</span> cumsum<span class="op">/</span><span class="kw">max</span>(cumsum))</span>
<span id="cb10-20"><a href="#cb10-20" aria-hidden="true"></a>p&lt;-<span class="st"> </span><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">x =</span> distance, <span class="dt">y =</span> pct)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb10-21"><a href="#cb10-21" aria-hidden="true"></a><span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb10-22"><a href="#cb10-22" aria-hidden="true"></a><span class="st">    </span><span class="kw">theme_minimal</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb10-23"><a href="#cb10-23" aria-hidden="true"></a><span class="st">    </span><span class="kw">scale_x_log10</span>(<span class="dt">limits =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="fl">1e7</span>)) <span class="op">+</span></span>
<span id="cb10-24"><a href="#cb10-24" aria-hidden="true"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&#39;none&#39;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb10-25"><a href="#cb10-25" aria-hidden="true"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&#39;Cumulative distribution of Bcells ATAC-seq reads&#39;</span>, <span class="dt">x =</span> <span class="st">&#39;Distance from TSS&#39;</span>, <span class="dt">y =</span> <span class="st">&#39;Cum. %&#39;</span>)</span></code></pre></div>
<blockquote>
<p>Compute promoter (TSS) Enrichment Score</p>
</blockquote>
<p>TSSES is another inmportant metric. Again, try both approaches: (1) using “dirty”
(but efficient!) successives functions, or (2) a “tidy” approach relying on <code>plyranges</code>.</p>
<p>A summaryzing plot is also typically shown to illustrate the “shape” of the TSS
enrichment.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true"></a><span class="co">## -- Manual way</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true"></a>Bcells_reads &lt;-<span class="st"> </span>bams[[<span class="st">&#39;Bcells&#39;</span>]]</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true"></a>Bcells_reads &lt;-<span class="st"> </span><span class="kw">resize</span>(Bcells_reads, <span class="dt">width =</span> <span class="dv">1</span>, <span class="dt">fix =</span> <span class="st">&#39;start&#39;</span>)</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true"></a>Bcells_reads_noMT &lt;-<span class="st"> </span>Bcells_reads[<span class="kw">seqnames</span>(Bcells_reads) <span class="op">!=</span><span class="st"> &quot;MT&quot;</span>]</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true"></a>ov &lt;-<span class="st"> </span><span class="kw">findOverlaps</span>(Bcells_reads_noMT, hg_promoters, <span class="dt">ignore.strand =</span> <span class="ot">TRUE</span>)</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true"></a>midprom &lt;-<span class="st"> </span><span class="kw">start</span>(hg_promoters[<span class="kw">subjectHits</span>(ov)]) <span class="op">+</span><span class="st"> </span>(<span class="kw">end</span>(hg_promoters[<span class="kw">subjectHits</span>(ov)]) <span class="op">-</span><span class="st"> </span><span class="kw">start</span>(hg_promoters[<span class="kw">subjectHits</span>(ov)])) <span class="op">/</span><span class="st"> </span><span class="dv">2</span></span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true"></a>distance &lt;-<span class="st"> </span><span class="kw">start</span>(Bcells_reads_noMT[<span class="kw">queryHits</span>(ov)]) <span class="op">-</span><span class="st"> </span>midprom</span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true"></a>binned_distance &lt;-<span class="st"> </span><span class="kw">round</span>(distance<span class="op">/</span><span class="dv">100</span>, <span class="dv">0</span>)<span class="op">*</span><span class="dv">100</span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true"></a>dists &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">table</span>(binned_distance))</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true"></a>bg &lt;-<span class="st"> </span><span class="kw">mean</span>(dists<span class="op">$</span>Freq[<span class="kw">abs</span>(<span class="kw">as.numeric</span>(<span class="kw">as.character</span>(dists<span class="op">$</span>binned_distance))) <span class="op">==</span><span class="st"> </span><span class="dv">2000</span>], <span class="dt">na.rm =</span> <span class="ot">TRUE</span>)</span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true"></a>dists<span class="op">$</span>enrich &lt;-<span class="st"> </span>dists<span class="op">$</span>Freq<span class="op">/</span>bg</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true"></a>TSSES &lt;-<span class="st"> </span><span class="kw">round</span>(dists<span class="op">$</span>enrich[<span class="kw">as.numeric</span>(<span class="kw">as.character</span>(dists<span class="op">$</span>binned_distance)) <span class="op">==</span><span class="st"> </span><span class="dv">0</span>], <span class="dv">2</span>)</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true"></a><span class="kw">plot</span>(<span class="dt">x =</span> <span class="kw">as.numeric</span>(<span class="kw">as.character</span>(dists<span class="op">$</span>binned_distance)), <span class="dt">y =</span> dists<span class="op">$</span>enrich, <span class="dt">type =</span> <span class="st">&#39;l&#39;</span>)</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true"></a></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true"></a><span class="co">## -- &quot;Tidy&quot; way, using `plyranges` and `tidyverse` functions</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true"></a>df&lt;-bams[[<span class="st">&#39;Bcells&#39;</span>]] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true"></a><span class="st">    </span><span class="kw">resize</span>(<span class="dt">width =</span> <span class="dv">1</span>, <span class="dt">fix =</span> <span class="st">&#39;start&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true"></a><span class="st">    </span><span class="kw">join_overlap_left</span>(hg_promoters) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true"></a><span class="st">    </span>plyranges<span class="op">::</span><span class="kw">select</span>(ID) <span class="op">%&gt;%</span></span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">is.na</span>(ID), seqnames <span class="op">!=</span><span class="st"> &#39;MT&#39;</span>) <span class="op">%&gt;%</span></span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true"></a><span class="st">    </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true"></a><span class="st">    </span><span class="kw">left_join</span>(<span class="kw">as_tibble</span>(hg_promoters) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(start, end, ID), <span class="dt">by =</span> <span class="st">&#39;ID&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(</span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true"></a>        <span class="dt">prom_mid =</span> start.y <span class="op">+</span><span class="st"> </span>(end.y <span class="op">-</span><span class="st"> </span>start.y) <span class="op">/</span><span class="st"> </span><span class="dv">2</span>, </span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true"></a>        <span class="dt">distance =</span> start.x <span class="op">-</span><span class="st"> </span>prom_mid, </span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true"></a>        <span class="dt">binned_distance =</span> <span class="kw">round</span>(distance<span class="op">/</span><span class="dv">100</span>, <span class="dv">0</span>)<span class="op">*</span><span class="dv">100</span></span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true"></a>    ) <span class="op">%&gt;%</span></span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true"></a><span class="st">    </span><span class="kw">count</span>(binned_distance) <span class="op">%&gt;%</span></span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(</span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true"></a>        <span class="dt">bg =</span> <span class="kw">sum</span>(.data<span class="op">$</span>n[<span class="kw">abs</span>(.data<span class="op">$</span>binned_distance) <span class="op">==</span><span class="st"> </span><span class="dv">2000</span>]), </span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true"></a>        <span class="dt">enrich_score =</span> n<span class="op">/</span>bg</span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true"></a>    )</span>
<span id="cb11-33"><a href="#cb11-33" aria-hidden="true"></a>TSSES &lt;-<span class="st"> </span><span class="kw">round</span>(df[df<span class="op">$</span>binned_distance <span class="op">==</span><span class="st"> </span><span class="dv">0</span>, <span class="st">&#39;enrich_score&#39;</span>], <span class="dv">2</span>)</span>
<span id="cb11-34"><a href="#cb11-34" aria-hidden="true"></a>p&lt;-<span class="st"> </span><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">x =</span> binned_distance, <span class="dt">y =</span> enrich_score)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb11-35"><a href="#cb11-35" aria-hidden="true"></a><span class="st">    </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb11-36"><a href="#cb11-36" aria-hidden="true"></a><span class="st">    </span><span class="kw">theme_minimal</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb11-37"><a href="#cb11-37" aria-hidden="true"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&#39;none&#39;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb11-38"><a href="#cb11-38" aria-hidden="true"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> glue<span class="op">::</span><span class="kw">glue</span>(<span class="st">&#39;TSS enrichment score (Signal-to-noise ratio): {TSSES}&#39;</span>), <span class="dt">x =</span> <span class="st">&#39;Distance from TSS&#39;</span>, <span class="dt">y =</span> <span class="st">&#39;Enrich. score&#39;</span>)</span></code></pre></div>
<p>How do you find the “tidy” approach, compared to the manual one without <code>plyranges</code>? What are the pros/cons of each approach?</p>
<blockquote>
<p>Can you wrap the 3 samples together?</p>
</blockquote>
<p>Try leveraging <code>BiocParallel</code> package. It provides <code>apply()</code>-like functions, and can run computation in
parallel using the <code>BPPARAM</code> argument.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true"></a><span class="kw">library</span>(BiocParallel)</span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true"></a>pl &lt;-<span class="st"> </span><span class="kw">bplapply</span>(<span class="dt">BPPARAM =</span> <span class="kw">MulticoreParam</span>(<span class="dt">workers =</span> <span class="dv">3</span>), samples, <span class="cf">function</span>(sample) {</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true"></a>    reads &lt;-<span class="st"> </span>bams[[sample]]</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true"></a>    reads_1bp &lt;-<span class="st"> </span><span class="kw">resize</span>(reads, <span class="dt">width =</span> <span class="dv">1</span>, <span class="dt">fix =</span> <span class="st">&#39;start&#39;</span>)</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true"></a>    peaks &lt;-<span class="st"> </span>peaks[[sample]]</span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true"></a>    <span class="co">## FRiP</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true"></a>    df&lt;-<span class="kw">add_nearest_distance</span>(reads, peaks) <span class="op">%&gt;%</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true"></a><span class="st">        </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true"></a><span class="st">        </span><span class="kw">mutate</span>(<span class="dt">isInPeak =</span> distance <span class="op">==</span><span class="st"> </span><span class="dv">0</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true"></a><span class="st">        </span><span class="kw">drop_na</span>(isInPeak)</span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true"></a>    pct &lt;-<span class="st"> </span><span class="kw">round</span>(<span class="kw">sum</span>(df<span class="op">$</span>isInPeak <span class="op">==</span><span class="st"> </span><span class="dv">1</span>) <span class="op">/</span><span class="st"> </span><span class="kw">nrow</span>(df) <span class="op">*</span><span class="st"> </span><span class="dv">100</span>, <span class="dv">2</span>)</span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true"></a>    p1&lt;-<span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">y =</span> isInPeak)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true"></a><span class="st">        </span><span class="kw">geom_bar</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true"></a><span class="st">        </span><span class="kw">theme_minimal</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true"></a><span class="st">        </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&#39;none&#39;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true"></a><span class="st">        </span><span class="kw">labs</span>(<span class="dt">title =</span> glue<span class="op">::</span><span class="kw">glue</span>(<span class="st">&#39;{sample}: Fraction of reads in peaks (FRiP): {pct}&#39;</span>), <span class="dt">x =</span> <span class="st">&#39;# of ATAC-seq reads&#39;</span>, <span class="dt">y =</span> <span class="st">&#39;Fragment within peak&#39;</span>)</span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true"></a>    <span class="co">## Distance from TSSs</span></span>
<span id="cb12-18"><a href="#cb12-18" aria-hidden="true"></a>    df&lt;-reads_1bp <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb12-19"><a href="#cb12-19" aria-hidden="true"></a><span class="st">        </span><span class="kw">add_nearest_distance</span>(hg_TSSs) <span class="op">%&gt;%</span></span>
<span id="cb12-20"><a href="#cb12-20" aria-hidden="true"></a><span class="st">        </span><span class="kw">mutate</span>(<span class="dt">distance =</span> <span class="kw">round</span>(distance<span class="op">/</span><span class="dv">10</span>, <span class="dv">0</span>)<span class="op">*</span><span class="dv">10</span>) <span class="op">%&gt;%</span></span>
<span id="cb12-21"><a href="#cb12-21" aria-hidden="true"></a><span class="st">        </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb12-22"><a href="#cb12-22" aria-hidden="true"></a><span class="st">        </span><span class="kw">count</span>(distance) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb12-23"><a href="#cb12-23" aria-hidden="true"></a><span class="st">        </span><span class="kw">mutate</span>(<span class="dt">cumsum =</span> <span class="kw">cumsum</span>(n), <span class="dt">pct =</span> cumsum<span class="op">/</span><span class="kw">max</span>(cumsum))</span>
<span id="cb12-24"><a href="#cb12-24" aria-hidden="true"></a>    p2&lt;-<span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">x =</span> distance, <span class="dt">y =</span> pct)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb12-25"><a href="#cb12-25" aria-hidden="true"></a><span class="st">        </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb12-26"><a href="#cb12-26" aria-hidden="true"></a><span class="st">        </span><span class="kw">theme_minimal</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb12-27"><a href="#cb12-27" aria-hidden="true"></a><span class="st">        </span><span class="kw">scale_x_log10</span>(<span class="dt">limits =</span> <span class="kw">c</span>(<span class="dv">1</span>, <span class="fl">1e7</span>)) <span class="op">+</span></span>
<span id="cb12-28"><a href="#cb12-28" aria-hidden="true"></a><span class="st">        </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&#39;none&#39;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb12-29"><a href="#cb12-29" aria-hidden="true"></a><span class="st">        </span><span class="kw">labs</span>(<span class="dt">title =</span> glue<span class="op">::</span><span class="kw">glue</span>(<span class="st">&#39;{sample}: Cumulative distribution of ATAC-seq reads&#39;</span>), <span class="dt">x =</span> <span class="st">&#39;Distance from TSS&#39;</span>, <span class="dt">y =</span> <span class="st">&#39;Cum. %&#39;</span>)</span>
<span id="cb12-30"><a href="#cb12-30" aria-hidden="true"></a>    <span class="co">## TSSES</span></span>
<span id="cb12-31"><a href="#cb12-31" aria-hidden="true"></a>    df&lt;-reads_1bp <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb12-32"><a href="#cb12-32" aria-hidden="true"></a><span class="st">        </span><span class="kw">join_overlap_left</span>(hg_promoters) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb12-33"><a href="#cb12-33" aria-hidden="true"></a><span class="st">        </span>plyranges<span class="op">::</span><span class="kw">select</span>(ID) <span class="op">%&gt;%</span></span>
<span id="cb12-34"><a href="#cb12-34" aria-hidden="true"></a><span class="st">        </span><span class="kw">filter</span>(<span class="op">!</span><span class="kw">is.na</span>(ID), seqnames <span class="op">!=</span><span class="st"> &#39;MT&#39;</span>) <span class="op">%&gt;%</span></span>
<span id="cb12-35"><a href="#cb12-35" aria-hidden="true"></a><span class="st">        </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb12-36"><a href="#cb12-36" aria-hidden="true"></a><span class="st">        </span><span class="kw">left_join</span>(<span class="kw">as_tibble</span>(hg_promoters) <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(start, end, ID), <span class="dt">by =</span> <span class="st">&#39;ID&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb12-37"><a href="#cb12-37" aria-hidden="true"></a><span class="st">        </span><span class="kw">mutate</span>(</span>
<span id="cb12-38"><a href="#cb12-38" aria-hidden="true"></a>            <span class="dt">prom_mid =</span> start.y <span class="op">+</span><span class="st"> </span>(end.y <span class="op">-</span><span class="st"> </span>start.y) <span class="op">/</span><span class="st"> </span><span class="dv">2</span>, </span>
<span id="cb12-39"><a href="#cb12-39" aria-hidden="true"></a>            <span class="dt">distance =</span> start.x <span class="op">-</span><span class="st"> </span>prom_mid, </span>
<span id="cb12-40"><a href="#cb12-40" aria-hidden="true"></a>            <span class="dt">binned_distance =</span> <span class="kw">round</span>(distance<span class="op">/</span><span class="dv">100</span>, <span class="dv">0</span>)<span class="op">*</span><span class="dv">100</span></span>
<span id="cb12-41"><a href="#cb12-41" aria-hidden="true"></a>        ) <span class="op">%&gt;%</span></span>
<span id="cb12-42"><a href="#cb12-42" aria-hidden="true"></a><span class="st">        </span><span class="kw">count</span>(binned_distance) <span class="op">%&gt;%</span></span>
<span id="cb12-43"><a href="#cb12-43" aria-hidden="true"></a><span class="st">        </span><span class="kw">mutate</span>(</span>
<span id="cb12-44"><a href="#cb12-44" aria-hidden="true"></a>            <span class="dt">bg =</span> <span class="kw">sum</span>(.data<span class="op">$</span>n[<span class="kw">abs</span>(.data<span class="op">$</span>binned_distance) <span class="op">==</span><span class="st"> </span><span class="dv">2000</span>]), </span>
<span id="cb12-45"><a href="#cb12-45" aria-hidden="true"></a>            <span class="dt">enrich_score =</span> n<span class="op">/</span>bg</span>
<span id="cb12-46"><a href="#cb12-46" aria-hidden="true"></a>        )</span>
<span id="cb12-47"><a href="#cb12-47" aria-hidden="true"></a>    TSSES &lt;-<span class="st"> </span><span class="kw">round</span>(df[df<span class="op">$</span>binned_distance <span class="op">==</span><span class="st"> </span><span class="dv">0</span>, <span class="st">&#39;enrich_score&#39;</span>], <span class="dv">2</span>)</span>
<span id="cb12-48"><a href="#cb12-48" aria-hidden="true"></a>    p3&lt;-<span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">x =</span> binned_distance, <span class="dt">y =</span> enrich_score)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb12-49"><a href="#cb12-49" aria-hidden="true"></a><span class="st">        </span><span class="kw">geom_line</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb12-50"><a href="#cb12-50" aria-hidden="true"></a><span class="st">        </span><span class="kw">theme_minimal</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb12-51"><a href="#cb12-51" aria-hidden="true"></a><span class="st">        </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&#39;none&#39;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb12-52"><a href="#cb12-52" aria-hidden="true"></a><span class="st">        </span><span class="kw">labs</span>(<span class="dt">title =</span> glue<span class="op">::</span><span class="kw">glue</span>(<span class="st">&#39;{sample}: TSS enrichment score (Signal-to-noise ratio): {TSSES}&#39;</span>), <span class="dt">x =</span> <span class="st">&#39;Distance from TSS&#39;</span>, <span class="dt">y =</span> <span class="st">&#39;Enrich. score&#39;</span>)</span>
<span id="cb12-53"><a href="#cb12-53" aria-hidden="true"></a>    cowplot<span class="op">::</span><span class="kw">plot_grid</span>(p1, p2, p3, <span class="dt">nrow =</span> <span class="dv">1</span>)</span>
<span id="cb12-54"><a href="#cb12-54" aria-hidden="true"></a>})</span>
<span id="cb12-55"><a href="#cb12-55" aria-hidden="true"></a>cowplot<span class="op">::</span><span class="kw">plot_grid</span>(<span class="dt">plotlist =</span> pl, <span class="dt">nrow =</span> <span class="dv">3</span>)</span></code></pre></div>
<p>Comment the differences. Can you visually appreciate whether some samples are
a better quality than others, when loading the tracks in IGV?</p>
</div>
</div>
<div id="split-bam-file-from-human-5000-pbmcs-scatacseq-into-bams-cell-type" class="section level2">
<h2>2. Split bam file from human 5,000 PBMCs scATACseq into bams / cell type</h2>
<div id="download-data-1" class="section level3">
<h3>Download data</h3>
<p>scATACseq for human 5K PBMCs can be downloaded from
<a href="https://support.10xgenomics.com/single-cell-atac/datasets/2.0.0/atac_pbmc_5k_nextgem">10X Genomics</a>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true"></a><span class="fu">mkdir</span> -p data/Homework_1/scATAC</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/2.0.0/atac_pbmc_5k_nextgem/atac_pbmc_5k_nextgem_analysis.tar.gz -o data/Homework_1/scATAC/atac_pbmc_5k_nextgem_analysis.tar.gz</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/2.0.0/atac_pbmc_5k_nextgem/atac_pbmc_5k_nextgem_possorted_bam.bam -o data/Homework_1/scATAC/atac_pbmc_5k_nextgem_possorted_bam.bam</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/2.0.0/atac_pbmc_5k_nextgem/atac_pbmc_5k_nextgem_possorted_bam.bam.bai -o data/Homework_1/scATAC/atac_pbmc_5k_nextgem_possorted_bam.bam.bai</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/2.0.0/atac_pbmc_5k_nextgem/atac_pbmc_5k_nextgem_web_summary.html -o data/Homework_1/scATAC/atac_pbmc_5k_nextgem_web_summary.html</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true"></a><span class="ex">curl</span> https://cf.10xgenomics.com/samples/cell-atac/2.0.0/atac_pbmc_5k_nextgem/atac_pbmc_5k_nextgem_cloupe.cloupe -o data/Homework_1/scATAC/atac_pbmc_5k_nextgem_cloupe.cloupe</span></code></pre></div>
</div>
<div id="check-the-clusters-found-by-cellranger" class="section level3">
<h3>Check the clusters found by cellranger</h3>
<p>Different cluster strategies are ran by <code>cellranger-atac</code>.
Check the HTML report: by default, the graph-based clustering results are shown. What do you think of them?
Load the <code>cloupe</code> file in <code>Loupe</code> to check other clustering results.
Is there another clustering approach that may look better-suited?</p>
</div>
<div id="split-bam-file-into-cluster-specific-bam-files-using-sinto" class="section level3">
<h3>Split bam file into cluster-specific bam files using sinto</h3>
<p>Now that you have chosen which cluster strategy you are going to use by inspecting data in Loupe,
can you find where the different clustering results are stored in the <code>scATAC</code> directory?
Can you locate the clustering results from <code>kmeans</code> clustering with <code>k = 5</code>? How many cells are in each cluster?</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true"></a><span class="fu">sed</span> <span class="st">&#39;1d&#39;</span> data/Homework_1/scATAC/analysis/clustering/kmeans_5_clusters/clusters.csv <span class="kw">|</span> <span class="fu">cut</span> -f2 -d, <span class="kw">|</span> <span class="fu">sort</span> <span class="kw">|</span> <span class="fu">uniq</span> -c</span></code></pre></div>
<p>Read <a href="https://timoast.github.io/sinto/basic_usage.html"><code>sinto</code></a> documentation. Which subcommand
is going to be useful to split a single scATACseq bam file into cluster-specific bam files?
Do we have the required files for <code>sinto</code> to work? What about the sorting of the original bam file?
Which bam tag is storing the cell barcode information, in the bam file generated by <code>cellranger-atac count</code>?</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true"></a><span class="ex">samtools</span> view -h data/Homework_1/scATAC/atac_pbmc_5k_nextgem_possorted_bam.bam <span class="kw">|</span> <span class="fu">head</span> -n 220 <span class="kw">|</span> <span class="fu">grep</span> -Pv <span class="st">&quot;^@SQ&quot;</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true"></a><span class="ex">samtools</span> view data/Homework_1/scATAC/atac_pbmc_5k_nextgem_possorted_bam.bam <span class="kw">|</span> <span class="fu">head</span> -n 1 </span></code></pre></div>
<p>Try and separate the <code>bam</code> file into individual bigwig files using <code>sinto</code>.
Watch out the space needed by <code>sinto</code>! <code>sinto</code> does not provide any way to
indicate the temporary/destination directory to use, it simply creates the bam files
according to barcodes in the working directory. Since the original <code>bam</code> file is ~15Gb,
the generated data is going to take at least this much disk space in the current directory!</p>
<p><code>sinto</code> will take about 15 min to run for the 5K cells dataset.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true"></a><span class="va">currdir=</span><span class="kw">`</span><span class="bu">pwd</span><span class="kw">`</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true"></a><span class="bu">cd</span> data/Homework_1/scATAC</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true"></a><span class="fu">sed</span> <span class="st">&#39;1d&#39;</span> analysis/clustering/kmeans_5_clusters/clusters.csv <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s/,/\t/&#39;</span> <span class="op">&gt;</span> clusters.tsv</span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true"></a><span class="ex">sinto</span> filterbarcodes <span class="kw">\</span></span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true"></a>    <span class="ex">-b</span> atac_pbmc_5k_nextgem_possorted_bam.bam <span class="kw">\</span></span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true"></a>    <span class="ex">--cells</span> clusters.tsv <span class="kw">\</span></span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true"></a>    <span class="ex">-p</span> 12 </span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true"></a><span class="bu">cd</span> <span class="va">$currdir</span></span></code></pre></div>
</div>
</div>
<div id="make-pseudo-bulk-bigwig-tracks" class="section level2">
<h2>3. Make pseudo-bulk bigwig tracks</h2>
<p>Are the generated bam files ready to be converted into coverage track?
Are there any processing step we should take care of (e.g. mapping, filtering)?
Which method can be used at this stage, to generate a depth-normalized track from
bam files?</p>
<blockquote>
<p>With macs2</p>
</blockquote>
<p>Watch out, this file is now paired-end !</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true"></a><span class="fu">mkdir</span> -p data/Homework_1/scATAC/peaks</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true"></a><span class="fu">mkdir</span> -p data/Homework_1/scATAC/bw</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true"></a><span class="va">CLUSTER=</span>5</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true"></a><span class="ex">macs2</span> callpeak <span class="kw">\</span></span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true"></a>    <span class="ex">-t</span> data/Homework_1/scATAC/<span class="st">&quot;</span><span class="va">${CLUSTER}</span><span class="st">&quot;</span>.bam <span class="kw">\</span></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true"></a>    <span class="ex">--format</span> BAMPE <span class="kw">\</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true"></a>    <span class="ex">--gsize</span> hs <span class="kw">\</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true"></a>    <span class="ex">--outdir</span> data/Homework_1/scATAC/peaks/ <span class="kw">\</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true"></a>    <span class="ex">--name</span> <span class="st">&quot;</span><span class="va">${CLUSTER}</span><span class="st">&quot;</span> <span class="kw">\</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true"></a>    <span class="ex">--bdg</span> --SPMR </span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true"></a><span class="fu">sed</span> -i <span class="st">&#39;s/,/./&#39;</span> data/Homework_1/scATAC/peaks/<span class="st">&quot;</span><span class="va">${CLUSTER}</span><span class="st">&quot;</span>_treat_pileup.bdg</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true"></a><span class="ex">bedGraphToBigWig</span> data/Homework_1/scATAC/peaks/<span class="st">&quot;</span><span class="va">${CLUSTER}</span><span class="st">&quot;</span>_treat_pileup.bdg ~/genomes/GRCh38/GRCh38.sizes.genome.modified data/Homework_1/scATAC/bw/<span class="st">&quot;</span><span class="va">${CLUSTER}</span><span class="st">&quot;</span>_macs2.bw</span></code></pre></div>
<blockquote>
<p>With bedtools</p>
</blockquote>
<div class="sourceCode" id="cb18"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true"></a><span class="va">CLUSTER=</span>5</span>
<span id="cb18-2"><a href="#cb18-2" aria-hidden="true"></a><span class="ex">genomeCoverageBed</span> -ibam -d -bg -pc -i data/Homework_1/scATAC/<span class="st">&quot;</span><span class="va">${CLUSTER}</span><span class="st">&quot;</span>.bam <span class="kw">|</span> <span class="fu">sort</span> -k1,1 -k2,2n <span class="op">&gt;</span> tmp.bdg</span>
<span id="cb18-3"><a href="#cb18-3" aria-hidden="true"></a><span class="ex">bedGraphToBigWig</span> tmp.bdg ~/genomes/GRCh38/GRCh38.sizes.genome.modified data/Homework_1/scATAC/bw/<span class="st">&quot;</span><span class="va">${CLUSTER}</span><span class="st">&quot;</span>_bedtools.bw</span></code></pre></div>
<blockquote>
<p>With deeptools bamCoverage</p>
</blockquote>
<div class="sourceCode" id="cb19"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true"></a><span class="va">CLUSTER=</span>5</span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true"></a><span class="ex">samtools</span> index data/Homework_1/scATAC/<span class="st">&quot;</span><span class="va">${CLUSTER}</span><span class="st">&quot;</span>.bam</span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true"></a><span class="ex">bamCoverage</span> <span class="kw">\</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true"></a>    <span class="ex">--bam</span> data/Homework_1/scATAC/<span class="st">&quot;</span><span class="va">${CLUSTER}</span><span class="st">&quot;</span>.bam <span class="kw">\</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true"></a>    <span class="ex">--outFileName</span> data/Homework_1/scATAC/bw/<span class="st">&quot;</span><span class="va">${CLUSTER}</span><span class="st">&quot;</span>_deeptools.bw <span class="kw">\</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true"></a>    <span class="ex">--binSize</span> 1 <span class="kw">\</span></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true"></a>    <span class="ex">--numberOfProcessors</span> 16 <span class="kw">\</span></span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true"></a>    <span class="ex">--extendReads</span> <span class="kw">\</span></span>
<span id="cb19-9"><a href="#cb19-9" aria-hidden="true"></a>    <span class="ex">--ignoreDuplicates</span></span></code></pre></div>
<blockquote>
<p>With tracklayer in R</p>
</blockquote>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true"></a><span class="kw">library</span>(GenomicAlignments)</span>
<span id="cb20-2"><a href="#cb20-2" aria-hidden="true"></a><span class="cf">for</span> (sample <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>) {</span>
<span id="cb20-3"><a href="#cb20-3" aria-hidden="true"></a>    bam &lt;-<span class="st"> </span>glue<span class="op">::</span><span class="kw">glue</span>(<span class="st">&#39;data/Homework_1/scATAC/{sample}.bam&#39;</span>)</span>
<span id="cb20-4"><a href="#cb20-4" aria-hidden="true"></a>    track &lt;-<span class="st"> </span>glue<span class="op">::</span><span class="kw">glue</span>(<span class="st">&#39;data/Homework_1/scATAC/bw/{sample}_R.bw&#39;</span>)</span>
<span id="cb20-5"><a href="#cb20-5" aria-hidden="true"></a>    ncells &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="kw">read.table</span>(<span class="st">&#39;data/Homework_1/scATAC/clusters.tsv&#39;</span>)<span class="op">$</span>V2 <span class="op">==</span><span class="st"> </span>sample)</span>
<span id="cb20-6"><a href="#cb20-6" aria-hidden="true"></a>    ga &lt;-<span class="st"> </span><span class="kw">readGAlignments</span>(bam)</span>
<span id="cb20-7"><a href="#cb20-7" aria-hidden="true"></a>    gr &lt;-<span class="st"> </span><span class="kw">as</span>(ga, <span class="st">&#39;GRanges&#39;</span>)</span>
<span id="cb20-8"><a href="#cb20-8" aria-hidden="true"></a>    gr.cov &lt;-<span class="st"> </span>IRanges<span class="op">::</span><span class="kw">coverage</span>(gr)</span>
<span id="cb20-9"><a href="#cb20-9" aria-hidden="true"></a>    rtracklayer<span class="op">::</span><span class="kw">export.bw</span>(gr.cov, track)</span>
<span id="cb20-10"><a href="#cb20-10" aria-hidden="true"></a>}</span></code></pre></div>
<p>Should the sequencing depth be taken into account here?
How can we normalize the coverage signal in this context?
Using the <code>R</code> method, try to normalize by the number of cells in each cluster.</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" aria-hidden="true"></a><span class="kw">library</span>(GenomicAlignments)</span>
<span id="cb21-2"><a href="#cb21-2" aria-hidden="true"></a><span class="cf">for</span> (sample <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>) {</span>
<span id="cb21-3"><a href="#cb21-3" aria-hidden="true"></a>    bam &lt;-<span class="st"> </span>glue<span class="op">::</span><span class="kw">glue</span>(<span class="st">&#39;data/Homework_1/scATAC/{sample}.bam&#39;</span>)</span>
<span id="cb21-4"><a href="#cb21-4" aria-hidden="true"></a>    track &lt;-<span class="st"> </span>glue<span class="op">::</span><span class="kw">glue</span>(<span class="st">&#39;data/Homework_1/scATAC/bw/{sample}_R_cellnb-normalized.bw&#39;</span>)</span>
<span id="cb21-5"><a href="#cb21-5" aria-hidden="true"></a>    ncells &lt;-<span class="st"> </span><span class="kw">sum</span>(<span class="kw">read.table</span>(<span class="st">&#39;data/Homework_1/scATAC/clusters.tsv&#39;</span>)<span class="op">$</span>V2 <span class="op">==</span><span class="st"> </span>sample)</span>
<span id="cb21-6"><a href="#cb21-6" aria-hidden="true"></a>    ga &lt;-<span class="st"> </span><span class="kw">readGAlignments</span>(bam)</span>
<span id="cb21-7"><a href="#cb21-7" aria-hidden="true"></a>    gr &lt;-<span class="st"> </span><span class="kw">as</span>(ga, <span class="st">&#39;GRanges&#39;</span>)</span>
<span id="cb21-8"><a href="#cb21-8" aria-hidden="true"></a>    gr.cov &lt;-<span class="st"> </span>IRanges<span class="op">::</span><span class="kw">coverage</span>(gr) <span class="op">/</span><span class="st"> </span>(ncells<span class="op">*</span><span class="dv">2</span>) <span class="op">*</span><span class="st"> </span><span class="dv">100</span></span>
<span id="cb21-9"><a href="#cb21-9" aria-hidden="true"></a>    rtracklayer<span class="op">::</span><span class="kw">export.bw</span>(gr.cov, track)</span>
<span id="cb21-10"><a href="#cb21-10" aria-hidden="true"></a>}</span></code></pre></div>
</div>
<div id="compare-bulk-and-pseudo-bulk-tracks" class="section level2">
<h2>4. Compare bulk and pseudo-bulk tracks</h2>
<p>Load tracks from bulk ATAC-seq (B cells, CD4+ T cells and CD8+ T cells) in IGV,
along with the 5 pseudo-bulk tracks from scATACseq.
Look at relevant loci (<code>CD4</code>, <code>CD8A</code>, <code>CD8B</code>, …). Which cluster could correspond
to B cells? To T cells?</p>
</div>


<footer class="footline">
	
</footer>

        
        </div>
        

      </div>

    <div id="navigation">
        
        

        
            
            
                
                    
                    
                
                

                    
                    
                        
                    
                    

                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
                        
            
            
                
                    
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                        
                        
                    
                
                

                    
                    
                    

                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        
                    
                        
            
            
                
                    
                
                

                    
                    
                    

                    
            
        
                    
            
        

        


	 
	 
		
			<a class="nav nav-prev" href="/scATACseq-workshop/homeworks/" title="2. Homeworks"> <i class="fa fa-chevron-left"></i></a>
		
		
			<a class="nav nav-next" href="/scATACseq-workshop/homeworks/homework_2/" title="Homework 2" style="margin-right: 0px;"><i class="fa fa-chevron-right"></i></a>
		
	
    </div>

    </section>

    <div style="left: -1000px; overflow: scroll; position: absolute; top: -1000px; border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;">
      <div style="border: none; box-sizing: content-box; height: 200px; margin: 0px; padding: 0px; width: 200px;"></div>
    </div>
    <script src="/scATACseq-workshop/js/clipboard.min.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/perfect-scrollbar.min.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/perfect-scrollbar.jquery.min.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/jquery.sticky.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/featherlight.min.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/highlight.pack.js?1630582900"></script>
    <script>hljs.initHighlightingOnLoad();</script>
    <script src="/scATACseq-workshop/js/modernizr.custom-3.6.0.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/learn.js?1630582900"></script>
    <script src="/scATACseq-workshop/js/hugo-learn.js?1630582900"></script>
    
        
            <script src="/scATACseq-workshop/mermaid/mermaid.js?1630582900"></script>
        
        <script>
            mermaid.initialize({ startOnLoad: true });
        </script>
    
    <script>
  (function(i,s,o,g,r,a,m){i['GoogleAnalyticsObject']=r;i[r]=i[r]||function(){
  (i[r].q=i[r].q||[]).push(arguments)},i[r].l=1*new Date();a=s.createElement(o),
  m=s.getElementsByTagName(o)[0];a.async=1;a.src=g;m.parentNode.insertBefore(a,m)
  })(window,document,'script','https://www.google-analytics.com/analytics.js','ga');

  ga('create', 'UA-105947713-1', 'auto');
  ga('send', 'pageview');

</script>

  </body>
</html>

