---
title: "Day 02"
output:
  blogdown::html_page:
    highlight: tango
---

<script src="/scATACseq-workshop/rmarkdown-libs/header-attrs/header-attrs.js"></script>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<div id="hands-on-session-3-analysing-scatacseq-with-signac" class="section level2">
<h2>Hands-on session 3: analysing scATACseq with Signac</h2>
<p>Here, we will focus on an example dataset generated by 10X Genomics: a set of ~5,000 cells from an mouse adult brain.
Processed data (generated by <code>cellranger-atac count</code>) can be downladed from here:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a><span class="fu">mkdir</span> -p data/MouseBrain</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_analysis.tar.gz -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_analysis.tar.gz</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_filtered_peak_bc_matrix.tar.gz -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_filtered_peak_bc_matrix.tar.gz</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_filtered_peak_bc_matrix.h5 -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_filtered_peak_bc_matrix.h5</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_filtered_tf_bc_matrix.tar.gz -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_filtered_tf_bc_matrix.tar.gz</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_filtered_tf_bc_matrix.h5 -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_filtered_tf_bc_matrix.h5</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_fragments.tsv.gz -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_fragments.tsv.gz</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_fragments.tsv.gz.tbi -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_fragments.tsv.gz.tbi</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_peak_annotation.tsv -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_peak_annotation.tsv</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_peaks.bed -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_peaks.bed</span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true"></a><span class="fu">wget</span> https://cg.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_possorted_bam.bam -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_possorted_bam.bam</span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_possorted_bam.bam.bai -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_possorted_bam.bam.bai</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_raw_peak_bc_matrix.tar.gz -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_raw_peak_bc_matrix.tar.gz</span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_raw_peak_bc_matrix.h5 -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_raw_peak_bc_matrix.h5</span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_singlecell.csv -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_singlecell.csv</span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_summary.csv -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_summary.csv</span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_summary.json -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_summary.json</span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_web_summary.html -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_web_summary.html</span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true"></a><span class="fu">wget</span> https://cf.10xgenomics.com/samples/cell-atac/1.2.0/atac_v1_adult_brain_fresh_5k/atac_v1_adult_brain_fresh_5k_cloupe.cloupe -O data/MouseBrain/atac_v1_adult_brain_fresh_5k_cloupe.cloupe</span></code></pre></div>
<div id="importing-data" class="section level3">
<h3>Importing data</h3>
<p>scATACseq raw data processing can be very different from bulk ATAC-seq, and downstream analysis is also quite specific to single-cell assays.
As of today, <code>Signac</code> is one of the few solutions developed to investigate scATACseq data in R, and is the most widely used approach. <code>Signac</code> is
an “extension” of <code>Seurat</code>, which originally focused on single-cell RNA-seq data analysis. It uses the same type of object structure.
A slight difference regarding the features is that they are actually chromatin accessible loci, rather than genes. Because peaks are frequently
studied in relation to neighbouring genes, the <code>Seurat</code> object can conventiently store gene annotations (or any type of genomic annotations)
in its <code>annotation</code> slot.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="kw">library</span>(Signac)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true"></a><span class="kw">library</span>(Seurat)</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true"></a><span class="kw">library</span>(BiocParallel)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true"></a><span class="kw">library</span>(plyranges)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true"></a></span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true"></a><span class="co">## -- Read brain ATAC-seq counts</span></span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true"></a>brain_assay &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">CreateChromatinAssay</span>(</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true"></a>    <span class="dt">counts =</span> Seurat<span class="op">::</span><span class="kw">Read10X_h5</span>(<span class="st">&quot;data/MouseBrain/atac_v1_adult_brain_fresh_5k_filtered_peak_bc_matrix.h5&quot;</span>),</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true"></a>    <span class="dt">sep =</span> <span class="kw">c</span>(<span class="st">&quot;:&quot;</span>, <span class="st">&quot;-&quot;</span>),</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true"></a>    <span class="dt">genome =</span> <span class="st">&quot;mm10&quot;</span>,</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true"></a>    <span class="dt">fragments =</span> <span class="st">&#39;data/MouseBrain/atac_v1_adult_brain_fresh_5k_fragments.tsv.gz&#39;</span>,</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true"></a>    <span class="dt">min.cells =</span> <span class="dv">1</span></span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true"></a>)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true"></a><span class="kw">class</span>(brain_assay)</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true"></a></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true"></a><span class="co">## -- Create Seurat object</span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true"></a>metadata &lt;-<span class="st"> </span>vroom<span class="op">::</span><span class="kw">vroom</span>(</span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true"></a>    <span class="dt">file =</span> <span class="st">&quot;data/MouseBrain/atac_v1_adult_brain_fresh_5k_singlecell.csv&quot;</span>,</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true"></a>    <span class="dt">col_names =</span> <span class="ot">TRUE</span></span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true"></a>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(barcode <span class="op">!=</span><span class="st"> &#39;NO_BARCODE&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">cell =</span> barcode) <span class="op">%&gt;%</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true"></a><span class="st">    </span><span class="kw">column_to_rownames</span>(<span class="st">&#39;barcode&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true"></a><span class="st">    </span><span class="kw">as.data.frame</span>()</span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">CreateSeuratObject</span>(</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true"></a>    <span class="dt">counts =</span> brain_assay,</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true"></a>    <span class="dt">assay =</span> <span class="st">&#39;peaks&#39;</span>,</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true"></a>    <span class="dt">project =</span> <span class="st">&#39;ATAC&#39;</span>,</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true"></a>    <span class="dt">meta.data =</span> metadata</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true"></a>)</span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true"></a></span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true"></a><span class="co">## -- Add genome annotations</span></span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true"></a><span class="kw">Annotation</span>(brain) &lt;-<span class="st"> </span>AnnotationHub<span class="op">::</span><span class="kw">AnnotationHub</span>()[[<span class="st">&#39;AH49547&#39;</span>]] <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(gene_type <span class="op">==</span><span class="st"> &#39;protein_coding&#39;</span>, type <span class="op">==</span><span class="st"> &#39;gene&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">gene_biotype =</span> gene_type)</span></code></pre></div>
</div>
<div id="accessing-data" class="section level3">
<h3>Accessing data</h3>
<p>Accessing the data stored in a <code>Seurat</code> object is (generally) relatively easy, using the <code>getAssayData()</code> function.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a><span class="co"># Data slots available in the Seurat object</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a><span class="kw">Assays</span>(brain)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a>slots &lt;-<span class="st"> </span><span class="kw">slotNames</span>(brain<span class="op">@</span>assays[[brain<span class="op">@</span>active.assay]])</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true"></a></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true"></a><span class="co"># Find peaks </span></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true"></a>peaks &lt;-<span class="st"> </span><span class="kw">GetAssayData</span>(brain, <span class="dt">slot =</span> <span class="st">&#39;ranges&#39;</span>)</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true"></a></span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true"></a><span class="co"># Find counts </span></span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true"></a>counts &lt;-<span class="st"> </span><span class="kw">GetAssayData</span>(brain, <span class="dt">slot =</span> <span class="st">&quot;counts&quot;</span>)</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true"></a>data &lt;-<span class="st"> </span><span class="kw">GetAssayData</span>(brain, <span class="dt">slot =</span> <span class="st">&quot;data&quot;</span>)</span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true"></a></span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true"></a><span class="co"># Find annotations</span></span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true"></a>annots &lt;-<span class="st"> </span><span class="kw">GetAssayData</span>(brain, <span class="dt">slot =</span> <span class="st">&quot;annotation&quot;</span>)</span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true"></a></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true"></a><span class="co"># Find seqinfo </span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true"></a>seqinfo &lt;-<span class="st"> </span><span class="kw">GetAssayData</span>(brain, <span class="dt">slot =</span> <span class="st">&quot;seqinfo&quot;</span>)</span></code></pre></div>
</div>
<div id="annotate-peaks" class="section level3">
<h3>Annotate peaks</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a>peaks &lt;-<span class="st"> </span><span class="kw">granges</span>(brain)</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true"></a>annots &lt;-<span class="st"> </span><span class="kw">Annotation</span>(brain) </span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true"></a>annotated_peaks &lt;-<span class="st"> </span><span class="kw">join_nearest</span>(peaks, annots) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true"></a><span class="st">    </span><span class="kw">add_nearest_distance</span>(annots) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true"></a>        <span class="dt">peak =</span> <span class="kw">paste</span>(seqnames, start, end, <span class="dt">sep =</span> <span class="st">&#39;-&#39;</span>),</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true"></a>        <span class="dt">inGene =</span> distance <span class="op">==</span><span class="st"> </span><span class="dv">0</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true"></a>    )</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true"></a><span class="kw">table</span>(annotated_peaks<span class="op">$</span>inGene)</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true"></a></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true"></a><span class="co"># - Distance to nearest promoter </span></span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true"></a>mm_promoters &lt;-<span class="st"> </span><span class="kw">Annotation</span>(brain) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true"></a><span class="st">    </span><span class="kw">resize</span>(<span class="dv">1</span>, <span class="dt">fix =</span> <span class="st">&#39;start&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true"></a><span class="st">    </span><span class="kw">resize</span>(<span class="dv">4000</span>, <span class="dt">fix =</span> <span class="st">&#39;center&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">prom_id =</span> <span class="kw">paste0</span>(<span class="st">&#39;prom_&#39;</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">n</span>()))</span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true"></a><span class="kw">add_nearest_distance</span>(peaks, mm_promoters) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true"></a><span class="st">    </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true"></a><span class="st">    </span><span class="kw">pull</span>(distance) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true"></a><span class="st">    </span><span class="kw">quantile</span>(<span class="dt">probs =</span> <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">1</span>, <span class="fl">0.1</span>))</span></code></pre></div>
</div>
<div id="qcing-data" class="section level3">
<h3>QCing data</h3>
<p>QC of scATACseq data stored in <code>Seurat</code> is simplified, using <code>Signac</code>-provided functions. On top of these utilities,
scATACseq fragments can always be imported (remember: they are stored in the <code>fragments.tsv.gz</code> generated by
<code>cellranger-atac count</code>, and they also contain information re: the cell-of-origin, for each fragment),
if one wants to manually check fragment distribution, size, …</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a><span class="co">## -- Nucleosome signal </span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">NucleosomeSignal</span>(<span class="dt">object =</span> brain)</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true"></a>brain<span class="op">$</span>nucleosome_group &lt;-<span class="st"> </span><span class="kw">ifelse</span>(brain<span class="op">$</span>nucleosome_signal <span class="op">&gt;</span><span class="st"> </span><span class="dv">4</span>, <span class="st">&#39;NS &gt; 4&#39;</span>, <span class="st">&#39;NS &lt; 4&#39;</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true"></a><span class="kw">range</span>(brain<span class="op">$</span>nucleosome_group)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true"></a><span class="kw">table</span>(brain<span class="op">$</span>nucleosome_group)</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true"></a></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true"></a><span class="co">## -- Frag. size distribution</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true"></a>frags &lt;-<span class="st"> </span>vroom<span class="op">::</span><span class="kw">vroom</span>(<span class="st">&#39;/data/20210804_scATACseq-workshop/data/MouseBrain/atac_v1_adult_brain_fresh_5k_fragments.tsv.gz&#39;</span>, <span class="dt">col_names =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true"></a><span class="st">    </span><span class="kw">setNames</span>(<span class="kw">c</span>(<span class="st">&#39;chr&#39;</span>, <span class="st">&#39;start&#39;</span>, <span class="st">&#39;stop&#39;</span>, <span class="st">&#39;cell&#39;</span>, <span class="st">&#39;cluster&#39;</span>))</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true"></a>frags_chr12 &lt;-<span class="st"> </span><span class="kw">filter</span>(frags, chr <span class="op">==</span><span class="st"> &#39;chr12&#39;</span>, start <span class="op">&lt;</span><span class="st"> </span><span class="dv">50000000</span>, cluster <span class="op">%in%</span><span class="st"> </span><span class="kw">c</span>(<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>))</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true"></a>x &lt;-<span class="st"> </span><span class="kw">left_join</span>(frags_chr12, dplyr<span class="op">::</span><span class="kw">select</span>(brain<span class="op">@</span>meta.data, cell, nucleosome_group)) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">width =</span> stop <span class="op">-</span><span class="st"> </span>start) </span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true"></a>p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(x, <span class="kw">aes</span>(<span class="dt">x =</span> width)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true"></a><span class="st">    </span><span class="kw">geom_bar</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true"></a><span class="st">    </span><span class="kw">facet_wrap</span>(<span class="op">~</span>nucleosome_group<span class="op">+</span>cluster, <span class="dt">scales =</span> <span class="st">&#39;free&#39;</span>, <span class="dt">ncol =</span> <span class="dv">10</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true"></a><span class="st">    </span><span class="kw">xlim</span>(<span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">800</span>))</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true"></a></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true"></a><span class="co">## -- TSS enrichment</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">TSSEnrichment</span>(brain, <span class="dt">fast =</span> <span class="ot">TRUE</span>)</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true"></a>brain<span class="op">$</span>high.tss &lt;-<span class="st"> </span><span class="kw">ifelse</span>(brain<span class="op">$</span>TSS.enrichment <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>, <span class="st">&#39;High&#39;</span>, <span class="st">&#39;Low&#39;</span>)</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true"></a></span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true"></a>df &lt;-<span class="st"> </span>frags_chr12 <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true"></a><span class="st">    </span><span class="kw">left_join</span>(dplyr<span class="op">::</span><span class="kw">select</span>(brain<span class="op">@</span>meta.data, cell, high.tss)) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># For each fragment, recover cell TSS enrich. status</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true"></a><span class="st">    </span>dplyr<span class="op">::</span><span class="kw">rename</span>(<span class="kw">c</span>(<span class="st">&#39;seqnames&#39;</span> =<span class="st"> &#39;chr&#39;</span>, <span class="st">&#39;end&#39;</span> =<span class="st"> &#39;stop&#39;</span>)) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true"></a><span class="st">    </span><span class="kw">as_granges</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true"></a><span class="st">    </span><span class="kw">join_overlap_left</span>(plyranges<span class="op">::</span><span class="kw">select</span>(mm_promoters, prom_id)) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># Link each fragment to overlapping promoter</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true"></a><span class="st">    </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true"></a><span class="st">    </span><span class="kw">drop_na</span>(prom_id) <span class="op">%&gt;%</span><span class="st"> </span><span class="co"># Remove fragments which are not overlapping any promoter</span></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true"></a><span class="st">    </span><span class="kw">left_join</span>(</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true"></a>        <span class="kw">as_tibble</span>(mm_promoters), <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&#39;prom_id&#39;</span>)</span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true"></a>    ) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(</span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true"></a>        <span class="dt">midprom =</span> start.y <span class="op">+</span><span class="st"> </span>(end.y <span class="op">-</span><span class="st"> </span>start.y)<span class="op">/</span><span class="dv">2</span>, </span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true"></a>        <span class="dt">distance =</span> midprom <span class="op">-</span><span class="st"> </span>start.x</span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true"></a>    ) <span class="op">%&gt;%</span></span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(high.tss <span class="op">==</span><span class="st"> &#39;High&#39;</span>) <span class="op">%&gt;%</span></span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true"></a><span class="st">    </span><span class="kw">count</span>(distance)</span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true"></a>p&lt;-<span class="st"> </span><span class="kw">ggplot</span>(df, <span class="kw">aes</span>(<span class="dt">x =</span> distance, <span class="dt">y =</span> n)) <span class="op">+</span><span class="st"> </span></span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true"></a><span class="st">    </span><span class="kw">geom_col</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true"></a><span class="st">    </span><span class="kw">geom_smooth</span>(<span class="dt">method =</span> <span class="st">&#39;loess&#39;</span>, <span class="dt">span =</span> <span class="fl">0.05</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true"></a><span class="st">    </span><span class="kw">theme_minimal</span>() <span class="op">+</span><span class="st"> </span></span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true"></a><span class="st">    </span><span class="kw">theme</span>(<span class="dt">legend.position =</span> <span class="st">&#39;none&#39;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true"></a><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="st">&#39;Fragment &quot;cut&quot; sites&#39;</span>, <span class="dt">x =</span> <span class="st">&#39;Distance from TSS&#39;</span>, <span class="dt">y =</span> <span class="st">&#39;# of cut sites&#39;</span>) <span class="op">+</span><span class="st"> </span></span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true"></a><span class="st">    </span><span class="kw">xlim</span>(<span class="kw">c</span>(<span class="op">-</span><span class="dv">1000</span>, <span class="dv">1000</span>))</span></code></pre></div>
</div>
<div id="subsetting-data" class="section level3">
<h3>Subsetting data</h3>
<p>Cell selection can be done based on some QC metrics (e.g. FRiP, FRiBl).
Feature selection can be done based on a count threshold.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a>brain<span class="op">$</span>FRiP &lt;-<span class="st"> </span>brain<span class="op">$</span>peak_region_fragments <span class="op">/</span><span class="st"> </span>brain<span class="op">$</span>passed_filters <span class="op">*</span><span class="st"> </span><span class="dv">100</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a>brain<span class="op">$</span>FRiBl &lt;-<span class="st"> </span>brain<span class="op">$</span>blacklist_region_fragments <span class="op">/</span><span class="st"> </span>brain<span class="op">$</span>peak_region_fragments</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a><span class="co"># Keep cells with high number of fragments in peaks</span></span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">subset</span>(brain, peak_region_fragments <span class="op">&gt;</span><span class="st"> </span><span class="dv">3000</span>) </span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true"></a><span class="co"># Remove cells with number of fragments in peaks too high</span></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">subset</span>(brain, peak_region_fragments <span class="op">&lt;</span><span class="st"> </span><span class="dv">100000</span>) </span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true"></a><span class="co"># Keep cells with high FRiP </span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">subset</span>(brain, FRiP <span class="op">&gt;</span><span class="st"> </span><span class="dv">50</span>) </span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true"></a><span class="co"># Keep cells with low FRiBl</span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">subset</span>(brain, FRiBl <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>) </span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true"></a><span class="co"># Remove cells with high nucleosome signal</span></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">subset</span>(brain, nucleosome_signal <span class="op">&lt;</span><span class="st"> </span><span class="dv">4</span>) </span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true"></a><span class="co"># Keep cells with good TSSES</span></span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">subset</span>(brain, TSS.enrichment <span class="op">&gt;</span><span class="st"> </span><span class="dv">2</span>) </span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true"></a><span class="co"># Remove peaks with less than 10 counts overall</span></span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>brain[<span class="kw">rowSums</span>(<span class="kw">GetAssayData</span>(brain, <span class="dt">slot =</span> <span class="st">&quot;counts&quot;</span>)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>, ]</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true"></a><span class="co"># Remove peaks detected in les than 10 cells overall</span></span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>brain[<span class="kw">rowSums</span>(<span class="kw">GetAssayData</span>(brain, <span class="dt">slot =</span> <span class="st">&quot;counts&quot;</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>, ]</span></code></pre></div>
</div>
<div id="normalizing-and-reduce-dimensionality-data" class="section level3">
<h3>Normalizing and reduce dimensionality data</h3>
<p>Data normalization and dimensionality reduction is done differently in scATACseq, compared to scRNAseq. This is
mostly due to the fact that scATACseq data is overly sparse, compared to (already quite sparse) scRNAseq data.
For more info, read <code>Signac</code> pre-print: (Multimodal single-cell chromatin analysis with Signac)[<a href="https://www.biorxiv.org/content/10.1101/2020.11.09.373613v1.full" class="uri">https://www.biorxiv.org/content/10.1101/2020.11.09.373613v1.full</a>]
To alleviate normalization problems linked to massive sparsity, an alternative normalization approach based on a “term frequency-inverse document frequency” approach.
TF-IDF transformation yields metrics reflecting how important a “word” is to a “document” in a “collection”
(in this case, a “word” is an accessible chromatin locus, a “document” is a cell, and the “collection” is the entire set of cells).</p>
<p>Dimensions of the normalized data matrix generated by this transformation are then reduced using singular value decomposition (SVD).</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a><span class="co"># Normalize data</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">RunTFIDF</span>(brain) </span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true"></a><span class="co"># Find variable features</span></span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">FindTopFeatures</span>(brain, <span class="dt">min.cutoff =</span> <span class="st">&#39;q0&#39;</span>) </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true"></a><span class="co"># Perform dimensionality reduction</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true"></a>brain &lt;-<span class="st"> </span>Signac<span class="op">::</span><span class="kw">RunSVD</span>(<span class="dt">object =</span> brain) </span></code></pre></div>
</div>
<div id="clustering" class="section level3">
<h3>Clustering</h3>
<p>Now that cells are embedded in a lower dimensional space, they can be graph-clustered just like “regular” cells from scRNAseq.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true"></a><span class="co"># Find neighboring cells</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">FindNeighbors</span>(<span class="dt">object =</span> brain, <span class="dt">reduction =</span> <span class="st">&#39;lsi&#39;</span>, <span class="dt">dims =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">30</span>)</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true"></a><span class="co"># Cluster cells</span></span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">FindClusters</span>(<span class="dt">object =</span> brain, <span class="dt">algorithm =</span> <span class="dv">3</span>, <span class="dt">resolution =</span> <span class="fl">1.2</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true"></a><span class="co"># Embbed cells in lower dimensional space</span></span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true"></a>brain &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>(brain, <span class="dt">reduction =</span> <span class="st">&#39;lsi&#39;</span>, <span class="dt">dims =</span> <span class="dv">2</span><span class="op">:</span><span class="dv">30</span>)</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true"></a><span class="kw">saveRDS</span>(brain, <span class="st">&#39;data/MouseBrain/brain.rds&#39;</span>)</span></code></pre></div>
</div>
<div id="da-with-seurat" class="section level3">
<h3>DA with Seurat</h3>
<p>Differential accessibility analysis of chromatin loci in scATACseq is a difficult exercise, which is still very experimental and rapidly-evolving.
One “easy” way to perform DA analysis within a scATACseq workflow is to rely on the already-existing <code>FindMarkers()</code> function from <code>Seurat</code>.
This approach has not been very thoroughly used yet, but allows quick-n-dirty exploratory analysis of the chromatin loci differently accessible
between pairs of (goups of) clusters.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true"></a><span class="co">## - Find markers between 2 groups of clusters </span></span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true"></a>da_peaks &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">FindMarkers</span>(</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true"></a>    <span class="dt">object =</span> brain,</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true"></a>    <span class="dt">ident.1 =</span> <span class="kw">c</span>(<span class="dv">10</span>, <span class="dv">16</span>),</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true"></a>    <span class="dt">ident.2 =</span> <span class="kw">c</span>(<span class="dv">4</span>, <span class="dv">5</span>),</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true"></a>    <span class="dt">min.pct =</span> <span class="fl">0.05</span></span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true"></a>)</span></code></pre></div>
</div>
<div id="exercises-session-3" class="section level3">
<h3>Exercises session 3</h3>
<ul>
<li>Inspect genes closest to the most DA peaks (in DA analysis between clusters 10+16 and 4+5). Propose a function for cells in these groups.</li>
</ul>
<div class="sourceCode" id="cb10"><pre class="sourceCode txt"><code class="sourceCode default"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a>da_peaks &lt;- readRDS(&#39;data/MouseBrain/da_peaks_10-16_vs_4-5.rds&#39;)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true"></a>peaks &lt;- rownames(da_peaks)[da_peaks$?...? &lt;= 0.01 &amp; da_peaks$?...? &gt;= 1.5]</span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true"></a>peaks &lt;- GenomicRanges::GRanges(str_replace(peaks, &#39;-&#39;, &#39;:&#39;))</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true"></a>Annotation(brain)[nearest(?...?, Annotation(?...?))]$?...?</span></code></pre></div>
</div>
</div>
<div id="hands-on-session-4-peaks-differential-accessibility-analysis" class="section level2">
<h2>Hands-on session 4: Peaks differential accessibility analysis</h2>
<div id="processing-raw-atac-data" class="section level3">
<h3>Processing raw ATAC data</h3>
<p>In this section, we will rely on worm bulk ATACseq data obtained from isolated tissues, to study differential accessibility analysis.</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true"></a><span class="co">## -- Download raw data </span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true"></a><span class="fu">mkdir</span> data/ATAC_worm/fastq/</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true"></a><span class="ex">curl</span> -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR105/079/SRR10564679/SRR10564679_1.fastq.gz -o data/ATAC_worm/fastq/SRR10564679_GSM4197239_Neurons_YA_ATAC_pe_rep1_Caenorhabditis_elegans_ATAC-seq_1.fastq.gz</span>
<span id="cb11-4"><a href="#cb11-4" aria-hidden="true"></a><span class="ex">curl</span> -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR105/079/SRR10564679/SRR10564679_2.fastq.gz -o data/ATAC_worm/fastq/SRR10564679_GSM4197239_Neurons_YA_ATAC_pe_rep1_Caenorhabditis_elegans_ATAC-seq_2.fastq.gz</span>
<span id="cb11-5"><a href="#cb11-5" aria-hidden="true"></a><span class="ex">curl</span> -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR105/086/SRR10564686/SRR10564686_1.fastq.gz -o data/ATAC_worm/fastq/SRR10564686_GSM4197244_Neurons_YA_ATAC_pe_rep2_Caenorhabditis_elegans_ATAC-seq_1.fastq.gz</span>
<span id="cb11-6"><a href="#cb11-6" aria-hidden="true"></a><span class="ex">curl</span> -L ftp://ftp.sra.ebi.ac.uk/vol1/fastq/SRR105/086/SRR10564686/SRR10564686_2.fastq.gz -o data/ATAC_worm/fastq/SRR10564686_GSM4197244_Neurons_YA_ATAC_pe_rep2_Caenorhabditis_elegans_ATAC-seq_2.fastq.gz</span>
<span id="cb11-7"><a href="#cb11-7" aria-hidden="true"></a></span>
<span id="cb11-8"><a href="#cb11-8" aria-hidden="true"></a><span class="co">## -- Process data </span></span>
<span id="cb11-9"><a href="#cb11-9" aria-hidden="true"></a><span class="va">GENOME=</span>~/genomes/ce11/ce11.fa</span>
<span id="cb11-10"><a href="#cb11-10" aria-hidden="true"></a><span class="va">CPU=</span>16 </span>
<span id="cb11-11"><a href="#cb11-11" aria-hidden="true"></a><span class="va">SAMPLE=</span>Neurons_rep1</span>
<span id="cb11-12"><a href="#cb11-12" aria-hidden="true"></a><span class="va">R1=</span>data/ATAC_worm/fastq/SRR10564679_GSM4197239_Neurons_YA_ATAC_pe_rep1_Caenorhabditis_elegans_ATAC-seq_1.fastq.gz</span>
<span id="cb11-13"><a href="#cb11-13" aria-hidden="true"></a><span class="va">R2=</span>data/ATAC_worm/fastq/SRR10564679_GSM4197239_Neurons_YA_ATAC_pe_rep1_Caenorhabditis_elegans_ATAC-seq_2.fastq.gz</span>
<span id="cb11-14"><a href="#cb11-14" aria-hidden="true"></a><span class="ex">bowtie2</span> --threads <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> -x <span class="st">&quot;</span><span class="va">${GENOME}</span><span class="st">&quot;</span> --maxins 1000 -1 <span class="st">&quot;</span><span class="va">${R1}</span><span class="st">&quot;</span> -2 <span class="st">&quot;</span><span class="va">${R2}</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb11-15"><a href="#cb11-15" aria-hidden="true"></a><span class="ex">samtools</span> fixmate -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -m - - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb11-16"><a href="#cb11-16" aria-hidden="true"></a><span class="ex">samtools</span> sort -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -T data/ATAC_worm/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.sam_sorting - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb11-17"><a href="#cb11-17" aria-hidden="true"></a><span class="ex">samtools</span> markdup -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -r -T data/ATAC_worm/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.sam_markdup - - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb11-18"><a href="#cb11-18" aria-hidden="true"></a><span class="ex">samtools</span> view -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -f 2 -q 10 -1 -b - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb11-19"><a href="#cb11-19" aria-hidden="true"></a><span class="ex">samtools</span> sort -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -l 9 -T data/ATAC_worm/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.sam_sorting2 -o data/ATAC_worm/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>_filtered.bam</span>
<span id="cb11-20"><a href="#cb11-20" aria-hidden="true"></a><span class="ex">samtools</span> index -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> data/ATAC_worm/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>_filtered.bam</span>
<span id="cb11-21"><a href="#cb11-21" aria-hidden="true"></a><span class="va">SAMPLE=</span>Neurons_rep2</span>
<span id="cb11-22"><a href="#cb11-22" aria-hidden="true"></a><span class="va">R1=</span>data/ATAC_worm/fastq/SRR10564686_GSM4197244_Neurons_YA_ATAC_pe_rep2_Caenorhabditis_elegans_ATAC-seq_1.fastq.gz</span>
<span id="cb11-23"><a href="#cb11-23" aria-hidden="true"></a><span class="va">R2=</span>data/ATAC_worm/fastq/SRR10564686_GSM4197244_Neurons_YA_ATAC_pe_rep2_Caenorhabditis_elegans_ATAC-seq_2.fastq.gz</span>
<span id="cb11-24"><a href="#cb11-24" aria-hidden="true"></a><span class="ex">bowtie2</span> --threads <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> -x <span class="st">&quot;</span><span class="va">${GENOME}</span><span class="st">&quot;</span> --maxins 1000 -1 <span class="st">&quot;</span><span class="va">${R1}</span><span class="st">&quot;</span> -2 <span class="st">&quot;</span><span class="va">${R2}</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb11-25"><a href="#cb11-25" aria-hidden="true"></a><span class="ex">samtools</span> fixmate -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -m - - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb11-26"><a href="#cb11-26" aria-hidden="true"></a><span class="ex">samtools</span> sort -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -T data/ATAC_worm/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.sam_sorting - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb11-27"><a href="#cb11-27" aria-hidden="true"></a><span class="ex">samtools</span> markdup -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -r -T data/ATAC_worm/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.sam_markdup - - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb11-28"><a href="#cb11-28" aria-hidden="true"></a><span class="ex">samtools</span> view -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -f 2 -q 10 -1 -b - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb11-29"><a href="#cb11-29" aria-hidden="true"></a><span class="ex">samtools</span> sort -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -l 9 -T data/ATAC_worm/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.sam_sorting2 -o data/ATAC_worm/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>_filtered.bam</span>
<span id="cb11-30"><a href="#cb11-30" aria-hidden="true"></a><span class="ex">samtools</span> index -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> data/ATAC_worm/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>_filtered.bam</span>
<span id="cb11-31"><a href="#cb11-31" aria-hidden="true"></a></span>
<span id="cb11-32"><a href="#cb11-32" aria-hidden="true"></a><span class="co"># This was repeated for 4 other tissues</span></span></code></pre></div>
</div>
<div id="defining-a-unified-set-of-peaks-from-a-bunch-of-atac-seq-experiments" class="section level3">
<h3>Defining a unified set of peaks from a bunch of ATAC-seq experiments</h3>
<p>When dealing with bulk ATAC-seq samples (or “pseudo-bulk” samples from scATACseq), another approach is to cluster accessible chromatin loci
based on their levels of accessibility in each sample. This can leverage a traditional <code>DESeq2</code> workflow, but first, one needs to identify a
set of chromatin loci overlapping <em>all</em> the accessible sites of each sample.</p>
<p>An approach to do this (if not using <code>cellranger-atac count</code>-identified peaks) can be to use <code>yapc</code> (yet another peak caller) command-line tool.
<code>yapc</code> relies on the ATAC-seq coverage “shape” of duplicates to identify “concave” coverage regions (i.e. “peaks”).
So we need bigwig tracks of each ATAC-seq experiment to map peaks with <code>yapc</code>.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb12-1"><a href="#cb12-1" aria-hidden="true"></a><span class="co">## -- Downloading tracks </span></span>
<span id="cb12-2"><a href="#cb12-2" aria-hidden="true"></a><span class="fu">mkdir</span> data/ATAC_worm</span>
<span id="cb12-3"><a href="#cb12-3" aria-hidden="true"></a><span class="fu">wget</span> https://ftp.ncbi.nlm.nih.gov/geo/series/GSE141nnn/GSE141210/suppl//GSE141210_RAW.tar?tool=geoquery -O data/ATAC_worm/ATACseq_tracks.tar.gz</span>
<span id="cb12-4"><a href="#cb12-4" aria-hidden="true"></a><span class="fu">tar</span> -xvf data/ATAC_worm/ATACseq_tracks.tar.gz -C data/ATAC_worm/</span>
<span id="cb12-5"><a href="#cb12-5" aria-hidden="true"></a></span>
<span id="cb12-6"><a href="#cb12-6" aria-hidden="true"></a><span class="co">## -- Find peaks with yapc</span></span>
<span id="cb12-7"><a href="#cb12-7" aria-hidden="true"></a><span class="ex">yapc</span> data/ATAC_worm/yapc <span class="kw">\</span></span>
<span id="cb12-8"><a href="#cb12-8" aria-hidden="true"></a>    <span class="ex">Germline</span> data/ATAC_worm/GSM4197238_Germline_YA_ATAC_pe_rep1.bw data/ATAC_worm/GSM4197243_Germline_YA_ATAC_pe_rep2.bw <span class="kw">\</span></span>
<span id="cb12-9"><a href="#cb12-9" aria-hidden="true"></a>    <span class="ex">Neurons</span> data/ATAC_worm/GSM4197239_Neurons_YA_ATAC_pe_rep1.bw data/ATAC_worm/GSM4197244_Neurons_YA_ATAC_pe_rep2.bw <span class="kw">\</span></span>
<span id="cb12-10"><a href="#cb12-10" aria-hidden="true"></a>    <span class="ex">Muscle</span> data/ATAC_worm/GSM4197240_Muscle_YA_ATAC_pe_rep1.bw data/ATAC_worm/GSM4197245_Muscle_YA_ATAC_pe_rep2.bw <span class="kw">\</span></span>
<span id="cb12-11"><a href="#cb12-11" aria-hidden="true"></a>    <span class="ex">Hypodermis</span> data/ATAC_worm/GSM4197241_Hypodermis_YA_ATAC_pe_rep1.bw data/ATAC_worm/GSM4197246_Hypodermis_YA_ATAC_pe_rep2.bw <span class="kw">\</span></span>
<span id="cb12-12"><a href="#cb12-12" aria-hidden="true"></a>    <span class="ex">Intestine</span> data/ATAC_worm/GSM4197242_Intestine_YA_ATAC_pe_rep1.bw data/ATAC_worm/GSM4197247_Intestine_YA_ATAC_pe_rep2.bw</span>
<span id="cb12-13"><a href="#cb12-13" aria-hidden="true"></a></span>
<span id="cb12-14"><a href="#cb12-14" aria-hidden="true"></a><span class="co">## -- Further processing depending on what you intend to do </span></span>
<span id="cb12-15"><a href="#cb12-15" aria-hidden="true"></a><span class="co"># .....</span></span>
<span id="cb12-16"><a href="#cb12-16" aria-hidden="true"></a><span class="co"># Details of how the identified peaks were further processed are available in Serizay et al., Genome Research 2020. </span></span>
<span id="cb12-17"><a href="#cb12-17" aria-hidden="true"></a><span class="co"># Eventually, ~15,000 tissue-specific promoters were identified in this study, and available from VplotR::ce11_proms</span></span></code></pre></div>
</div>
<div id="counting-reads-over-peaks" class="section level3">
<h3>Counting reads over peaks</h3>
<p>From the ~ 45,000 identified peaks with <code>yapc</code>, we annotated and filtered down ~ 17,000 promoters.
We can compute the coverage of each promoter in each of the 10 ATAC-seq samples in <code>R</code> using
<code>featureCounts()</code> function from <code>R</code> wrapper of <code>subread</code>.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true"></a>files &lt;-<span class="st"> </span><span class="kw">c</span>(</span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true"></a>    <span class="st">&#39;data/ATAC_worm/Germline_rep1_filtered.bam&#39;</span>,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true"></a>    <span class="st">&#39;data/ATAC_worm/Germline_rep2_filtered.bam&#39;</span>,</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true"></a>    <span class="st">&#39;data/ATAC_worm/Neurons_rep1_filtered.bam&#39;</span>,</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true"></a>    <span class="st">&#39;data/ATAC_worm/Neurons_rep2_filtered.bam&#39;</span>,</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true"></a>    <span class="st">&#39;data/ATAC_worm/Muscle_rep1_filtered.bam&#39;</span>,</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true"></a>    <span class="st">&#39;data/ATAC_worm/Muscle_rep2_filtered.bam&#39;</span>,</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true"></a>    <span class="st">&#39;data/ATAC_worm/Hypodermis_rep1_filtered.bam&#39;</span>,</span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true"></a>    <span class="st">&#39;data/ATAC_worm/Hypodermis_rep2_filtered.bam&#39;</span>,</span>
<span id="cb13-10"><a href="#cb13-10" aria-hidden="true"></a>    <span class="st">&#39;data/ATAC_worm/Intestine_rep1_filtered.bam&#39;</span>,</span>
<span id="cb13-11"><a href="#cb13-11" aria-hidden="true"></a>    <span class="st">&#39;data/ATAC_worm/Intestine_rep2_filtered.bam&#39;</span></span>
<span id="cb13-12"><a href="#cb13-12" aria-hidden="true"></a>)</span>
<span id="cb13-13"><a href="#cb13-13" aria-hidden="true"></a>gtf &lt;-<span class="st"> &#39;data/ATAC_worm/ce11_proms.gtf&#39;</span></span>
<span id="cb13-14"><a href="#cb13-14" aria-hidden="true"></a>cnts &lt;-<span class="st"> </span>Rsubread<span class="op">::</span><span class="kw">featureCounts</span>(</span>
<span id="cb13-15"><a href="#cb13-15" aria-hidden="true"></a>    <span class="dt">files =</span> files, </span>
<span id="cb13-16"><a href="#cb13-16" aria-hidden="true"></a>    <span class="dt">annot.ext =</span> <span class="st">&#39;data/ATAC_worm/ce11_proms.gtf&#39;</span>, </span>
<span id="cb13-17"><a href="#cb13-17" aria-hidden="true"></a>    <span class="dt">isGTFAnnotationFile =</span> <span class="ot">TRUE</span>, </span>
<span id="cb13-18"><a href="#cb13-18" aria-hidden="true"></a>    <span class="dt">GTF.featureType =</span> <span class="st">&#39;sequence_feature&#39;</span>, </span>
<span id="cb13-19"><a href="#cb13-19" aria-hidden="true"></a>    <span class="dt">GTF.attrType =</span> <span class="st">&#39;id&#39;</span>, </span>
<span id="cb13-20"><a href="#cb13-20" aria-hidden="true"></a>    <span class="dt">isPairedEnd =</span> <span class="ot">TRUE</span>, </span>
<span id="cb13-21"><a href="#cb13-21" aria-hidden="true"></a>    <span class="dt">nthreads =</span> <span class="dv">16</span></span>
<span id="cb13-22"><a href="#cb13-22" aria-hidden="true"></a>)</span></code></pre></div>
</div>
<div id="performing-da-tests" class="section level3">
<h3>Performing DA tests</h3>
<p><code>DESeq2</code> time-tested workflow can be followed, starting from the counts obtained by <code>featureCounts</code>.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true"></a><span class="kw">library</span>(SummarizedExperiment)</span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true"></a><span class="kw">library</span>(DESeq2)</span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true"></a></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true"></a><span class="kw">data</span>(ce11_proms, <span class="st">&#39;VplotR&#39;</span>)</span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true"></a>counts &lt;-<span class="st"> </span><span class="kw">SummarizedExperiment</span>(</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true"></a>    <span class="dt">assays =</span> <span class="kw">list</span>(<span class="st">&#39;counts&#39;</span> =<span class="st"> </span>cnts<span class="op">$</span>counts), </span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true"></a>    <span class="dt">rowRanges =</span> ce11_proms,</span>
<span id="cb14-9"><a href="#cb14-9" aria-hidden="true"></a>    <span class="dt">colData =</span> S4Vectors<span class="op">::</span><span class="kw">DataFrame</span>(</span>
<span id="cb14-10"><a href="#cb14-10" aria-hidden="true"></a>        <span class="dt">sample =</span> cnts<span class="op">$</span>targets, </span>
<span id="cb14-11"><a href="#cb14-11" aria-hidden="true"></a>        <span class="dt">tissue =</span> <span class="kw">factor</span>(<span class="kw">str_replace_all</span>(cnts<span class="op">$</span>targets, <span class="st">&#39;_.*&#39;</span>, <span class="st">&#39;&#39;</span>), <span class="kw">c</span>(<span class="st">&#39;Germline&#39;</span>, <span class="st">&#39;Neurons&#39;</span>, <span class="st">&#39;Muscle&#39;</span>, <span class="st">&#39;Hypodermis&#39;</span>, <span class="st">&#39;Intestine&#39;</span>)), </span>
<span id="cb14-12"><a href="#cb14-12" aria-hidden="true"></a>        <span class="dt">rep =</span> <span class="kw">str_replace_all</span>(cnts<span class="op">$</span>targets, <span class="st">&#39;_filtered.bam|.*rep&#39;</span>, <span class="st">&#39;&#39;</span>)</span>
<span id="cb14-13"><a href="#cb14-13" aria-hidden="true"></a>    )</span>
<span id="cb14-14"><a href="#cb14-14" aria-hidden="true"></a>)</span>
<span id="cb14-15"><a href="#cb14-15" aria-hidden="true"></a></span>
<span id="cb14-16"><a href="#cb14-16" aria-hidden="true"></a>dds &lt;-<span class="st"> </span><span class="kw">DESeqDataSet</span>(counts, <span class="dt">design =</span> <span class="op">~</span><span class="st"> </span>tissue)</span>
<span id="cb14-17"><a href="#cb14-17" aria-hidden="true"></a>dds &lt;-<span class="st"> </span><span class="kw">DESeq</span>(dds)</span>
<span id="cb14-18"><a href="#cb14-18" aria-hidden="true"></a></span>
<span id="cb14-19"><a href="#cb14-19" aria-hidden="true"></a>contrasts &lt;-<span class="st"> </span><span class="kw">levels</span>(counts<span class="op">$</span>tissue) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb14-20"><a href="#cb14-20" aria-hidden="true"></a><span class="st">    </span><span class="kw">combn</span>(<span class="dv">2</span>, <span class="dt">simplify =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb14-21"><a href="#cb14-21" aria-hidden="true"></a><span class="st">    </span><span class="kw">lapply</span>(<span class="cf">function</span>(x) {<span class="kw">c</span>(<span class="st">&quot;tissue&quot;</span>, x[<span class="dv">2</span>], x[<span class="dv">1</span>])})</span>
<span id="cb14-22"><a href="#cb14-22" aria-hidden="true"></a>contrasts_names &lt;-<span class="st"> </span><span class="kw">sapply</span>(contrasts, <span class="cf">function</span>(x) <span class="kw">paste0</span>(x[<span class="dv">2</span>], <span class="st">&quot;_vs_&quot;</span>, x[<span class="dv">3</span>]))</span>
<span id="cb14-23"><a href="#cb14-23" aria-hidden="true"></a>out.l2fc &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">matrix</span>(<span class="dt">nrow=</span><span class="kw">length</span>(ce11_proms), <span class="dt">ncol=</span><span class="kw">length</span>(contrasts))) ; <span class="kw">colnames</span>(out.l2fc) &lt;-<span class="st"> </span><span class="kw">paste0</span>(contrasts_names, <span class="st">&#39;.l2FC&#39;</span>)</span>
<span id="cb14-24"><a href="#cb14-24" aria-hidden="true"></a>out.padj &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="kw">matrix</span>(<span class="dt">nrow=</span><span class="kw">length</span>(ce11_proms), <span class="dt">ncol=</span><span class="kw">length</span>(contrasts))) ; <span class="kw">colnames</span>(out.padj) &lt;-<span class="st"> </span><span class="kw">paste0</span>(contrasts_names, <span class="st">&#39;.padj&#39;</span>)</span>
<span id="cb14-25"><a href="#cb14-25" aria-hidden="true"></a><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(contrasts)) {</span>
<span id="cb14-26"><a href="#cb14-26" aria-hidden="true"></a>    res &lt;-<span class="st"> </span><span class="kw">results</span>(dds, <span class="dt">contrast =</span> contrasts[[i]])</span>
<span id="cb14-27"><a href="#cb14-27" aria-hidden="true"></a>    <span class="kw">metadata</span>(res)<span class="op">$</span>alpha &lt;-<span class="st"> </span><span class="fl">0.01</span></span>
<span id="cb14-28"><a href="#cb14-28" aria-hidden="true"></a>    out &lt;-<span class="st"> </span><span class="kw">as.data.frame</span>(res)</span>
<span id="cb14-29"><a href="#cb14-29" aria-hidden="true"></a>    out.l2fc[,i] &lt;-<span class="st"> </span>out[,<span class="dv">2</span>]</span>
<span id="cb14-30"><a href="#cb14-30" aria-hidden="true"></a>    out.padj[,i] &lt;-<span class="st"> </span>out[,<span class="dv">6</span>]</span>
<span id="cb14-31"><a href="#cb14-31" aria-hidden="true"></a>}</span>
<span id="cb14-32"><a href="#cb14-32" aria-hidden="true"></a></span>
<span id="cb14-33"><a href="#cb14-33" aria-hidden="true"></a>FPM &lt;-<span class="st"> </span><span class="kw">fpm</span>(dds, <span class="dt">robust =</span> <span class="ot">FALSE</span>)</span>
<span id="cb14-34"><a href="#cb14-34" aria-hidden="true"></a><span class="kw">colnames</span>(FPM) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&#39;FPM_&#39;</span>, counts<span class="op">$</span>sample) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">str_replace_all</span>(<span class="st">&#39;_filtered.bam&#39;</span>, <span class="st">&#39;&#39;</span>)</span>
<span id="cb14-35"><a href="#cb14-35" aria-hidden="true"></a></span>
<span id="cb14-36"><a href="#cb14-36" aria-hidden="true"></a>results &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb14-37"><a href="#cb14-37" aria-hidden="true"></a>    <span class="kw">as.data.frame</span>(ce11_proms),</span>
<span id="cb14-38"><a href="#cb14-38" aria-hidden="true"></a>    FPM,</span>
<span id="cb14-39"><a href="#cb14-39" aria-hidden="true"></a>    out.l2fc,</span>
<span id="cb14-40"><a href="#cb14-40" aria-hidden="true"></a>    out.padj</span>
<span id="cb14-41"><a href="#cb14-41" aria-hidden="true"></a>)</span>
<span id="cb14-42"><a href="#cb14-42" aria-hidden="true"></a>results<span class="op">$</span>sign.l2FC &lt;-<span class="st"> </span><span class="kw">apply</span>(results[,<span class="kw">grepl</span>(<span class="st">&quot;l2FC&quot;</span>, <span class="kw">colnames</span>(results))], <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="kw">any</span>(x <span class="op">&lt;</span><span class="st"> </span><span class="op">-</span><span class="kw">log2</span>(<span class="dv">2</span>) <span class="op">|</span><span class="st"> </span>x <span class="op">&gt;</span><span class="st"> </span><span class="kw">log2</span>(<span class="dv">2</span>)))</span>
<span id="cb14-43"><a href="#cb14-43" aria-hidden="true"></a>results<span class="op">$</span>sign.padj &lt;-<span class="st"> </span><span class="kw">apply</span>(results[,<span class="kw">grepl</span>(<span class="st">&quot;padj&quot;</span>, <span class="kw">colnames</span>(results))], <span class="dv">1</span>, <span class="cf">function</span>(x) <span class="kw">any</span>(x <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.01</span>))</span>
<span id="cb14-44"><a href="#cb14-44" aria-hidden="true"></a><span class="kw">write.table</span>(results, <span class="st">&#39;data/ATAC_worm/quantif.txt&#39;</span>, <span class="dt">quote =</span> <span class="ot">FALSE</span>, <span class="dt">col.names =</span> <span class="ot">TRUE</span>, <span class="dt">row.names =</span> <span class="ot">FALSE</span>, <span class="dt">sep =</span> <span class="st">&#39;</span><span class="ch">\t</span><span class="st">&#39;</span>)</span>
<span id="cb14-45"><a href="#cb14-45" aria-hidden="true"></a></span>
<span id="cb14-46"><a href="#cb14-46" aria-hidden="true"></a><span class="kw">rowData</span>(dds) &lt;-<span class="st"> </span><span class="kw">cbind</span>(</span>
<span id="cb14-47"><a href="#cb14-47" aria-hidden="true"></a>    <span class="kw">rowData</span>(dds), </span>
<span id="cb14-48"><a href="#cb14-48" aria-hidden="true"></a>    FPM,</span>
<span id="cb14-49"><a href="#cb14-49" aria-hidden="true"></a>    out.l2fc,</span>
<span id="cb14-50"><a href="#cb14-50" aria-hidden="true"></a>    out.padj</span>
<span id="cb14-51"><a href="#cb14-51" aria-hidden="true"></a>)</span>
<span id="cb14-52"><a href="#cb14-52" aria-hidden="true"></a><span class="kw">saveRDS</span>(dds, <span class="st">&#39;data/ATAC_worm/quantif.rds&#39;</span>)</span></code></pre></div>
</div>
<div id="clustering-peaks" class="section level3">
<h3>Clustering peaks</h3>
<p>Regularized log values can be averaged across replicates to quantify chromatin
accessibility at each peak in each sample.</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true"></a>rlogs &lt;-<span class="st"> </span>DESeq2<span class="op">::</span><span class="kw">rlog</span>(dds, <span class="dt">blind =</span> <span class="ot">FALSE</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true"></a><span class="st">    </span><span class="kw">assay</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true"></a><span class="st">    </span><span class="kw">as_tibble</span>() <span class="op">%&gt;%</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true"></a><span class="st">    </span><span class="kw">setNames</span>(<span class="kw">paste0</span>(<span class="st">&#39;rlog_&#39;</span>, dds<span class="op">$</span>sample) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">str_replace_all</span>(<span class="st">&#39;_filtered.bam&#39;</span>, <span class="st">&#39;&#39;</span>)) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">id =</span> <span class="kw">paste0</span>(<span class="st">&#39;prom_&#39;</span>, <span class="dv">1</span><span class="op">:</span><span class="kw">n</span>())) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true"></a><span class="st">    </span><span class="kw">pivot_longer</span>(<span class="op">-</span>id, <span class="dt">names_to =</span> <span class="st">&#39;class&#39;</span>, <span class="dt">values_to =</span> <span class="st">&#39;rlog&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true"></a><span class="st">    </span><span class="kw">separate</span>(class, <span class="dt">into =</span> <span class="kw">c</span>(<span class="ot">NA</span>, <span class="st">&#39;tissue&#39;</span>, <span class="st">&#39;rep&#39;</span>), <span class="dt">sep =</span> <span class="st">&#39;_&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true"></a><span class="st">    </span><span class="kw">group_by</span>(id, tissue) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true"></a><span class="st">    </span><span class="kw">summarize</span>(<span class="dt">rlog =</span> <span class="kw">mean</span>(rlog)) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true"></a><span class="st">    </span><span class="kw">pivot_wider</span>(<span class="dt">names_from =</span> tissue, <span class="dt">values_from =</span> rlog) <span class="op">%&gt;%</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true"></a><span class="st">    </span><span class="kw">ungroup</span>() <span class="op">%&gt;%</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true"></a><span class="st">    </span><span class="kw">filter</span>(<span class="kw">rowSums</span>(<span class="kw">select</span>(., <span class="op">-</span>id)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>)</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true"></a><span class="kw">saveRDS</span>(rlogs, <span class="st">&#39;data/ATAC_worm/rlogs.rds&#39;</span>)</span></code></pre></div>
<p>Peaks can then be clustered based on these rlog values, using an unsupervised approach
such as <code>k-means</code> or <code>k-medoids</code> (through the <code>pam()</code> function).</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true"></a>clusts &lt;-<span class="st"> </span>cluster<span class="op">::</span><span class="kw">pam</span>(<span class="kw">select</span>(rlogs, <span class="op">-</span>id), <span class="dt">k =</span> <span class="dv">9</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true"></a>p &lt;-<span class="st"> </span>pheatmap<span class="op">::</span><span class="kw">pheatmap</span>(</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true"></a>    <span class="kw">select</span>(rlogs, <span class="op">-</span>id)[<span class="kw">order</span>(clusts<span class="op">$</span>clustering), ], </span>
<span id="cb16-4"><a href="#cb16-4" aria-hidden="true"></a>    <span class="dt">cluster_rows =</span> <span class="ot">FALSE</span>, </span>
<span id="cb16-5"><a href="#cb16-5" aria-hidden="true"></a>    <span class="dt">cluster_cols =</span> <span class="ot">FALSE</span>, </span>
<span id="cb16-6"><a href="#cb16-6" aria-hidden="true"></a>    <span class="dt">breaks =</span> <span class="kw">seq</span>(<span class="dv">0</span>, <span class="dv">12</span>, <span class="dt">length.out =</span> <span class="dv">100</span>), </span>
<span id="cb16-7"><a href="#cb16-7" aria-hidden="true"></a>    <span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">&quot;#ffffff&quot;</span>, <span class="st">&quot;#ffffff&quot;</span>, <span class="st">&quot;#dda937&quot;</span>, <span class="st">&quot;#E31A1C&quot;</span>, <span class="st">&quot;#a51618&quot;</span>))(<span class="dv">100</span>)</span>
<span id="cb16-8"><a href="#cb16-8" aria-hidden="true"></a>)</span></code></pre></div>
</div>
<div id="exercises-session-4" class="section level3">
<h3>Exercises session 4</h3>
<ul>
<li>Compare clustering to provided cell type activity for ce11 promoters. Comment.</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="sourceCode txt"><code class="sourceCode default"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true"></a>dds &lt;- readRDS(&#39;data/ATAC_worm/quantif.rds&#39;)</span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true"></a>rlogs &lt;- readRDS(&#39;?...?&#39;)</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true"></a>clusts &lt;- cluster::pam(select(rlogs, -?...?), k = 9)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true"></a>table(rowData(dds)$which.tissues[!rowData(dds)$allZero], clusts$?...?)</span></code></pre></div>
</div>
</div>
