---
title: "Day 03"
output:
  blogdown::html_page:
    highlight: tango
---

<script src="/scATACseq-workshop/rmarkdown-libs/header-attrs/header-attrs.js"></script>
<style type="text/css">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  background-color: #f8f8f8; }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ef2929; } /* Alert */
code span.an { color: #8f5902; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #c4a000; } /* Attribute */
code span.bn { color: #0000cf; } /* BaseN */
code span.cf { color: #204a87; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4e9a06; } /* Char */
code span.cn { color: #000000; } /* Constant */
code span.co { color: #8f5902; font-style: italic; } /* Comment */
code span.cv { color: #8f5902; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #8f5902; font-weight: bold; font-style: italic; } /* Documentation */
code span.dt { color: #204a87; } /* DataType */
code span.dv { color: #0000cf; } /* DecVal */
code span.er { color: #a40000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #0000cf; } /* Float */
code span.fu { color: #000000; } /* Function */
code span.im { } /* Import */
code span.in { color: #8f5902; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #204a87; font-weight: bold; } /* Keyword */
code span.op { color: #ce5c00; font-weight: bold; } /* Operator */
code span.ot { color: #8f5902; } /* Other */
code span.pp { color: #8f5902; font-style: italic; } /* Preprocessor */
code span.sc { color: #000000; } /* SpecialChar */
code span.ss { color: #4e9a06; } /* SpecialString */
code span.st { color: #4e9a06; } /* String */
code span.va { color: #000000; } /* Variable */
code span.vs { color: #4e9a06; } /* VerbatimString */
code span.wa { color: #8f5902; font-weight: bold; font-style: italic; } /* Warning */
</style>


<div id="hands-on-session-5-motif-and-tf-analysis-of-peak-clusters" class="section level2">
<h2>Hands-on session 5: Motif and TF analysis of peak clusters</h2>
<div id="scan-genome-for-jaspar-motifs" class="section level3">
<h3>Scan genome for JASPAR motifs</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true"></a>motifs &lt;-<span class="st"> </span>TFBSTools<span class="op">::</span><span class="kw">getMatrixSet</span>(</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true"></a>    JASPAR2020<span class="op">::</span>JASPAR2020,</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true"></a>    <span class="kw">list</span>(<span class="dt">species =</span> <span class="dv">6239</span>, <span class="dt">all_versions =</span> <span class="ot">FALSE</span>)</span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true"></a>)</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true"></a>ce11_seq &lt;-<span class="st"> </span>Biostrings<span class="op">::</span><span class="kw">getSeq</span>(BSgenome.Celegans.UCSC.ce11<span class="op">::</span>BSgenome.Celegans.UCSC.ce11)</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true"></a>motifs_hits &lt;-<span class="st"> </span>BiocParallel<span class="op">::</span><span class="kw">bplapply</span>(motifs, <span class="cf">function</span>(motif) {</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true"></a>    TF &lt;-<span class="st"> </span>TFBSTools<span class="op">::</span><span class="kw">name</span>(motif)</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true"></a>    <span class="kw">message</span>(TF)</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true"></a>    TFBSTools<span class="op">::</span><span class="kw">searchSeq</span>(<span class="kw">as</span>(motif, <span class="st">&#39;PWMatrix&#39;</span>), ce11_seq, <span class="dt">min.score =</span> <span class="fl">0.95</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true"></a><span class="st">        </span><span class="kw">lapply</span>(as, <span class="st">&#39;GRanges&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true"></a><span class="st">        </span>GenomicRanges<span class="op">::</span><span class="kw">GRangesList</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true"></a><span class="st">        </span><span class="kw">unlist</span>()</span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true"></a>}) <span class="op">%&gt;%</span><span class="st"> </span>purrr<span class="op">::</span><span class="kw">set_names</span>(TFBSTools<span class="op">::</span><span class="kw">name</span>(motifs)) <span class="op">%&gt;%</span><span class="st"> </span>GenomicRanges<span class="op">::</span><span class="kw">GRangesList</span>()</span></code></pre></div>
</div>
<div id="for-each-set-of-promoters-check-the-enrichment-of-each-tf-motif-in-it" class="section level3">
<h3>For each set of promoters, check the enrichment of each TF motif in it</h3>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true"></a><span class="kw">library</span>(plyranges)</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true"></a>ce11_proms &lt;-<span class="st"> </span>SummarizedExperiment<span class="op">::</span><span class="kw">rowRanges</span>(<span class="kw">readRDS</span>(<span class="st">&#39;data/ATAC_worm/quantif.rds&#39;</span>))</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true"></a>enrichments &lt;-<span class="st"> </span>BiocParallel<span class="op">::</span><span class="kw">bplapply</span>(<span class="kw">names</span>(motifs_hits), <span class="cf">function</span>(TF) {</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true"></a>    tibble<span class="op">::</span><span class="kw">tibble</span>(</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true"></a>        <span class="dt">isprom =</span> <span class="ot">TRUE</span>,</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true"></a>        <span class="dt">overlap =</span> ce11_proms <span class="op">%over%</span><span class="st"> </span>motifs_hits[[TF]], </span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true"></a>        <span class="dt">tissue =</span> ce11_proms<span class="op">$</span>which.tissues</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true"></a>    ) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true"></a><span class="st">        </span><span class="kw">group_by</span>(tissue) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true"></a><span class="st">        </span><span class="kw">summarize</span>(</span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true"></a>            <span class="dt">TF =</span> TF,</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true"></a>            <span class="dt">nProms =</span> <span class="kw">sum</span>(isprom), </span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true"></a>            <span class="dt">nTF =</span> <span class="kw">sum</span>(overlap), </span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true"></a>            <span class="dt">nProms_tot =</span> <span class="kw">length</span>(ce11_proms)</span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true"></a>        )</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true"></a>}) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true"></a><span class="st">    </span><span class="kw">bind_rows</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true"></a><span class="st">    </span><span class="kw">left_join</span>(<span class="kw">tibble</span>(<span class="dt">TF =</span> <span class="kw">names</span>(motifs_hits), <span class="dt">nTF_tot =</span> <span class="kw">sapply</span>(motifs_hits, <span class="cf">function</span>(x) <span class="kw">sum</span>(x <span class="op">%over%</span><span class="st"> </span>ce11_proms)))) <span class="op">%&gt;%</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(</span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true"></a>        <span class="dt">a =</span> nTF,</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true"></a>        <span class="dt">b =</span> nTF_tot <span class="op">-</span><span class="st"> </span>nTF,</span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true"></a>        <span class="dt">c =</span> nProms <span class="op">-</span><span class="st"> </span>nTF,</span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true"></a>        <span class="dt">d =</span> nProms_tot <span class="op">-</span><span class="st"> </span>nTF_tot <span class="op">-</span><span class="st"> </span>nProms <span class="op">+</span><span class="st"> </span>nTF</span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true"></a>    ) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true"></a><span class="st">    </span><span class="kw">rowwise</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(</span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true"></a>        <span class="dt">pval =</span> <span class="kw">fisher.test</span>(<span class="kw">matrix</span>(<span class="kw">c</span>(a, b, c, d), <span class="dt">ncol =</span> <span class="dv">2</span>))<span class="op">$</span>p.value,</span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true"></a>        <span class="dt">odd =</span> <span class="kw">fisher.test</span>(<span class="kw">matrix</span>(<span class="kw">c</span>(a, b, c, d), <span class="dt">ncol =</span> <span class="dv">2</span>))<span class="op">$</span>estimate</span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true"></a>    ) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="op">-</span>a, <span class="op">-</span>b, <span class="op">-</span>c, <span class="op">-</span>d)</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true"></a>enrichments <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(pval <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.01</span>, odd <span class="op">&gt;</span><span class="st"> </span><span class="dv">3</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(tissue)</span></code></pre></div>
</div>
<div id="check-promoter-groups-against-a-list-of-tf-binding-locations-by-chip" class="section level3">
<h3>Check promoter groups against a list of TF binding locations (by ChIP)</h3>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true"></a>beds &lt;-<span class="st"> </span><span class="kw">list.files</span>(<span class="st">&#39;data/ATAC_worm/ChIP&#39;</span>, <span class="dt">full.names =</span> <span class="ot">TRUE</span>)</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true"></a>TF_peaks &lt;-<span class="st"> </span><span class="kw">lapply</span>(beds, rtracklayer<span class="op">::</span>import)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true"></a><span class="kw">names</span>(TF_peaks) &lt;-<span class="st"> </span><span class="kw">basename</span>(beds) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">str_replace_all</span>(<span class="st">&#39;.bed|&#39;</span>, <span class="st">&#39;&#39;</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true"></a>enrichments &lt;-<span class="st"> </span>BiocParallel<span class="op">::</span><span class="kw">bplapply</span>(<span class="kw">names</span>(TF_peaks), <span class="cf">function</span>(TF) {</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true"></a>    tibble<span class="op">::</span><span class="kw">tibble</span>(</span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true"></a>        <span class="dt">isprom =</span> <span class="ot">TRUE</span>,</span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true"></a>        <span class="dt">overlap =</span> ce11_proms <span class="op">%over%</span><span class="st"> </span>TF_peaks[[TF]], </span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true"></a>        <span class="dt">tissue =</span> ce11_proms<span class="op">$</span>which.tissues</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true"></a>    ) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true"></a><span class="st">        </span><span class="kw">group_by</span>(tissue) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true"></a><span class="st">        </span><span class="kw">summarize</span>(</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true"></a>            <span class="dt">TF =</span> TF,</span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true"></a>            <span class="dt">nProms =</span> <span class="kw">sum</span>(isprom), </span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true"></a>            <span class="dt">nTF =</span> <span class="kw">sum</span>(overlap), </span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true"></a>            <span class="dt">nProms_tot =</span> <span class="kw">length</span>(ce11_proms)</span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true"></a>        )</span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true"></a>}) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true"></a><span class="st">    </span><span class="kw">bind_rows</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true"></a><span class="st">    </span><span class="kw">left_join</span>(<span class="kw">tibble</span>(<span class="dt">TF =</span> <span class="kw">names</span>(TF_peaks), <span class="dt">nTF_tot =</span> <span class="kw">sapply</span>(TF_peaks, <span class="cf">function</span>(x) <span class="kw">sum</span>(x <span class="op">%over%</span><span class="st"> </span>ce11_proms)))) <span class="op">%&gt;%</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(</span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true"></a>        <span class="dt">a =</span> nTF,</span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true"></a>        <span class="dt">b =</span> nTF_tot <span class="op">-</span><span class="st"> </span>nTF,</span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true"></a>        <span class="dt">c =</span> nProms <span class="op">-</span><span class="st"> </span>nTF,</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true"></a>        <span class="dt">d =</span> nProms_tot <span class="op">-</span><span class="st"> </span>nTF_tot <span class="op">-</span><span class="st"> </span>nProms <span class="op">+</span><span class="st"> </span>nTF</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true"></a>    ) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true"></a><span class="st">    </span><span class="kw">rowwise</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(</span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true"></a>        <span class="dt">pval =</span> <span class="kw">fisher.test</span>(<span class="kw">matrix</span>(<span class="kw">c</span>(a, b, c, d), <span class="dt">ncol =</span> <span class="dv">2</span>))<span class="op">$</span>p.value,</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true"></a>        <span class="dt">odd =</span> <span class="kw">fisher.test</span>(<span class="kw">matrix</span>(<span class="kw">c</span>(a, b, c, d), <span class="dt">ncol =</span> <span class="dv">2</span>))<span class="op">$</span>estimate</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true"></a>    ) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">select</span>(<span class="op">-</span>a, <span class="op">-</span>b, <span class="op">-</span>c, <span class="op">-</span>d)</span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true"></a>enrichments <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">filter</span>(pval <span class="op">&lt;</span><span class="st"> </span><span class="fl">1e-5</span>, odd <span class="op">&gt;</span><span class="st"> </span><span class="dv">3</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(tissue) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">print</span>(<span class="dt">n =</span> <span class="st">&#39;all&#39;</span>)</span></code></pre></div>
</div>
</div>
<div id="hands-on-session-6-chromvar-on-single-cell-atac-seq-from-human-hematopoiesis" class="section level2">
<h2>Hands-on session 6: chromVAR on single-cell ATAC-seq from human hematopoiesis</h2>
<p>Data is obtained from <code>Corces et al., Nat. Genet. 2018</code> (DOI: <a href="https://doi.org/10.1038/ng.3646">10.1038/ng.3646</a>)</p>
<div id="download-raw-data" class="section level3">
<h3>Download raw data</h3>
<div class="sourceCode" id="cb4"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true"></a><span class="co">## -- Downloading tracks </span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true"></a><span class="fu">mkdir</span> data/scATAC_LSC-LMPP-mono</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true"></a><span class="bu">cd</span> data/scATAC_LSC-LMPP-mono</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true"></a><span class="fu">cat</span> download.txt <span class="kw">|</span> <span class="ex">parallel</span> --bar -j 16 {}</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true"></a><span class="fu">mv</span> SRR* fastq/</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true"></a><span class="bu">cd</span> ../..</span></code></pre></div>
</div>
<div id="process-raw-data" class="section level3">
<h3>Process raw data</h3>
<div class="sourceCode" id="cb5"><pre class="sourceCode sh"><code class="sourceCode bash"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true"></a><span class="co">## -- Process data </span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true"></a><span class="va">GENOME=</span>~/genomes/hg38/hg38.fa</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true"></a><span class="va">CPU=</span>16</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true"></a></span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true"></a><span class="kw">for</span> <span class="ex">SAMPLE</span> in <span class="kw">`</span><span class="fu">ls</span> data/scATAC_LSC-LMPP-mono/fastq/SRR* <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s,_Homo_.*,,&#39;</span> <span class="kw">|</span> <span class="fu">sed</span> <span class="st">&#39;s,data/scATAC_LSC-LMPP-mono/fastq/,,&#39;</span> <span class="kw">|</span> <span class="fu">uniq</span><span class="kw">`</span></span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true"></a><span class="kw">do</span></span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true"></a>    <span class="va">R1=</span>data/scATAC_LSC-LMPP-mono/fastq/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>_Homo_sapiens_OTHER_1.fastq.gz</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true"></a>    <span class="va">R2=</span>data/scATAC_LSC-LMPP-mono/fastq/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>_Homo_sapiens_OTHER_2.fastq.gz</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true"></a>    <span class="ex">bowtie2</span> --threads <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> -x <span class="st">&quot;</span><span class="va">${GENOME}</span><span class="st">&quot;</span> --maxins 1000 -1 <span class="st">&quot;</span><span class="va">${R1}</span><span class="st">&quot;</span> -2 <span class="st">&quot;</span><span class="va">${R2}</span><span class="st">&quot;</span> <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true"></a>        <span class="ex">samtools</span> fixmate -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -m - - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true"></a>        <span class="ex">samtools</span> sort -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -T data/scATAC_LSC-LMPP-mono/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.sam_sorting - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true"></a>        <span class="ex">samtools</span> markdup -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -r -T data/scATAC_LSC-LMPP-mono/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.sam_markdup - - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true"></a>        <span class="ex">samtools</span> view -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -f 2 -q 10 -1 -b - <span class="kw">|</span> <span class="kw">\</span></span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true"></a>        <span class="ex">samtools</span> sort -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> --output-fmt bam -l 9 -T data/scATAC_LSC-LMPP-mono/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>.sam_sorting2 -o data/scATAC_LSC-LMPP-mono/bam/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>_filtered.bam</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true"></a>    <span class="ex">samtools</span> index -@ <span class="st">&quot;</span><span class="va">${CPU}</span><span class="st">&quot;</span> data/scATAC_LSC-LMPP-mono/bam/<span class="st">&quot;</span><span class="va">${SAMPLE}</span><span class="st">&quot;</span>_filtered.bam</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true"></a><span class="kw">done</span></span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true"></a></span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true"></a><span class="co">## -- Merge all bams</span></span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true"></a><span class="ex">samtools</span> merge -@ 12 -r -p <span class="kw">`</span><span class="fu">ls</span> data/scATAC_LSC-LMPP-mono/bam/*bam<span class="kw">`</span> -o data/scATAC_LSC-LMPP-mono/384_cells.bam</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true"></a><span class="ex">samtools</span> sort -@ 12 --output-fmt bam -l 9 -T data/scATAC_LSC-LMPP-mono/384_cells_sorting -o data/scATAC_LSC-LMPP-mono/384_cells_sorted.bam data/scATAC_LSC-LMPP-mono/384_cells.bam</span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true"></a><span class="ex">samtools</span> index data/scATAC_LSC-LMPP-mono/384_cells_sorted.bam</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true"></a></span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true"></a><span class="co">## -- Call peaks</span></span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true"></a><span class="ex">macs2</span> callpeak -t data/scATAC_LSC-LMPP-mono/384_cells_sorted.bam --format BAMPE --gsize 2900000000 --outdir data/scATAC_LSC-LMPP-mono/peaks --name 384_cells</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true"></a></span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true"></a><span class="co">## -- Make track</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true"></a><span class="ex">bamCoverage</span> <span class="kw">\</span></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true"></a>    <span class="ex">--bam</span> data/scATAC_LSC-LMPP-mono/384_cells_sorted.bam <span class="kw">\</span></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true"></a>    <span class="ex">--outFileName</span> data/scATAC_LSC-LMPP-mono/384_cells_sorted.bw <span class="kw">\</span></span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true"></a>    <span class="ex">--binSize</span> 10 <span class="kw">\</span></span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true"></a>    <span class="ex">--numberOfProcessors</span> 12 <span class="kw">\</span></span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true"></a>    <span class="ex">--normalizeUsing</span> CPM <span class="kw">\</span></span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true"></a>    <span class="ex">--skipNonCoveredRegions</span> <span class="kw">\</span></span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true"></a>    <span class="ex">--extendReads</span></span></code></pre></div>
</div>
<div id="get-a-summarizedexperiment-of-scatacseq-counts-over-dhs-sites" class="section level3">
<h3>Get a SummarizedExperiment of scATACseq counts over DHS sites</h3>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true"></a><span class="kw">library</span>(tidyverse)</span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true"></a><span class="kw">library</span>(SummarizedExperiment)</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true"></a><span class="kw">library</span>(plyranges)</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true"></a>BiocParallel<span class="op">::</span><span class="kw">register</span>(BiocParallel<span class="op">::</span><span class="kw">MulticoreParam</span>(<span class="dv">16</span>, <span class="dt">progressbar =</span> <span class="ot">TRUE</span>))</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true"></a></span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true"></a><span class="co">## -- Get peaks </span></span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true"></a>peaks &lt;-<span class="st"> </span>rtracklayer<span class="op">::</span><span class="kw">import</span>(<span class="st">&#39;data/scATAC_LSC-LMPP-mono/peaks/384_cells_peaks.narrowPeak&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(<span class="dt">start =</span> start <span class="op">+</span><span class="st"> </span>peak) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true"></a><span class="st">    </span><span class="kw">resize</span>(<span class="dt">width =</span> <span class="dv">1</span>, <span class="dt">fix =</span> <span class="st">&#39;start&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true"></a><span class="st">    </span><span class="kw">resize</span>(<span class="dt">width =</span> <span class="dv">200</span>, <span class="dt">fix =</span> <span class="st">&#39;center&#39;</span>)</span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true"></a></span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true"></a><span class="co">## -- Import scATACseq fragments </span></span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true"></a>ga &lt;-<span class="st"> </span>GenomicAlignments<span class="op">::</span><span class="kw">readGAlignmentPairs</span>(</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true"></a>    <span class="st">&#39;data/scATAC_LSC-LMPP-mono/384_cells_sorted.bam&#39;</span>, </span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true"></a>    <span class="dt">use.names =</span> <span class="ot">TRUE</span>, </span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true"></a>    <span class="dt">param =</span> Rsamtools<span class="op">::</span><span class="kw">ScanBamParam</span>(<span class="dt">tag =</span> <span class="st">&quot;RG&quot;</span>)</span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true"></a>)</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true"></a><span class="kw">mcols</span>(ga)<span class="op">$</span>RG &lt;-<span class="st"> </span><span class="kw">GRanges</span>(<span class="kw">first</span>(ga))<span class="op">$</span>RG</span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true"></a>ga &lt;-<span class="st"> </span>ga <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true"></a><span class="st">    </span><span class="kw">GRanges</span>() <span class="op">%&gt;%</span></span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true"></a><span class="st">    </span><span class="kw">mutate</span>(</span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true"></a>        <span class="dt">cellid =</span> <span class="kw">as_tibble</span>(.) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">separate</span>(RG, <span class="dt">into =</span> <span class="kw">c</span>(<span class="ot">NA</span>, <span class="ot">NA</span>, <span class="st">&#39;cellID&#39;</span>, <span class="ot">NA</span>), <span class="dt">sep =</span> <span class="st">&#39;_&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">pull</span>(cellID)</span>
<span id="cb6-23"><a href="#cb6-23" aria-hidden="true"></a>    )</span>
<span id="cb6-24"><a href="#cb6-24" aria-hidden="true"></a>   </span>
<span id="cb6-25"><a href="#cb6-25" aria-hidden="true"></a><span class="co">## -- Find overlaps between fragments and peaks</span></span>
<span id="cb6-26"><a href="#cb6-26" aria-hidden="true"></a>ovPEAK &lt;-<span class="st"> </span>GenomicRanges<span class="op">::</span><span class="kw">findOverlaps</span>(peaks, ga)</span>
<span id="cb6-27"><a href="#cb6-27" aria-hidden="true"></a></span>
<span id="cb6-28"><a href="#cb6-28" aria-hidden="true"></a><span class="co">## -- List cell barcodes</span></span>
<span id="cb6-29"><a href="#cb6-29" aria-hidden="true"></a>uniqueBarcodes &lt;-<span class="st"> </span><span class="kw">unique</span>(ga<span class="op">$</span>cellid)</span>
<span id="cb6-30"><a href="#cb6-30" aria-hidden="true"></a>id &lt;-<span class="st"> </span><span class="kw">factor</span>(ga<span class="op">$</span>cellid, <span class="dt">levels =</span> uniqueBarcodes)</span>
<span id="cb6-31"><a href="#cb6-31" aria-hidden="true"></a></span>
<span id="cb6-32"><a href="#cb6-32" aria-hidden="true"></a><span class="co">## -- Assemble counts as a Sparse matrix</span></span>
<span id="cb6-33"><a href="#cb6-33" aria-hidden="true"></a>countdf &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb6-34"><a href="#cb6-34" aria-hidden="true"></a>    <span class="dt">peaks =</span> <span class="kw">queryHits</span>(ovPEAK),</span>
<span id="cb6-35"><a href="#cb6-35" aria-hidden="true"></a>    <span class="dt">sample =</span> id[<span class="kw">subjectHits</span>(ovPEAK)],</span>
<span id="cb6-36"><a href="#cb6-36" aria-hidden="true"></a>    <span class="dt">read =</span> <span class="kw">names</span>(ga)[<span class="kw">subjectHits</span>(ovPEAK)]</span>
<span id="cb6-37"><a href="#cb6-37" aria-hidden="true"></a>) <span class="op">%&gt;%</span></span>
<span id="cb6-38"><a href="#cb6-38" aria-hidden="true"></a><span class="st">    </span><span class="kw">distinct</span>() <span class="op">%&gt;%</span></span>
<span id="cb6-39"><a href="#cb6-39" aria-hidden="true"></a><span class="st">    </span><span class="kw">select</span>(<span class="op">-</span><span class="kw">one_of</span>(<span class="st">&quot;read&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb6-40"><a href="#cb6-40" aria-hidden="true"></a><span class="st">    </span><span class="kw">group_by</span>(peaks, sample) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb6-41"><a href="#cb6-41" aria-hidden="true"></a><span class="st">    </span><span class="kw">tally</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb6-42"><a href="#cb6-42" aria-hidden="true"></a><span class="st">    </span><span class="kw">data.matrix</span>()</span>
<span id="cb6-43"><a href="#cb6-43" aria-hidden="true"></a>cnts &lt;-<span class="st"> </span>Matrix<span class="op">::</span><span class="kw">sparseMatrix</span>(</span>
<span id="cb6-44"><a href="#cb6-44" aria-hidden="true"></a>    <span class="dt">i =</span> <span class="kw">c</span>(countdf[,<span class="dv">1</span>], <span class="kw">length</span>(peaks)),</span>
<span id="cb6-45"><a href="#cb6-45" aria-hidden="true"></a>    <span class="dt">j =</span> <span class="kw">c</span>(countdf[,<span class="dv">2</span>], <span class="kw">length</span>(uniqueBarcodes)),</span>
<span id="cb6-46"><a href="#cb6-46" aria-hidden="true"></a>    <span class="dt">x =</span> <span class="kw">c</span>(countdf[,<span class="dv">3</span>],<span class="dv">0</span>)</span>
<span id="cb6-47"><a href="#cb6-47" aria-hidden="true"></a>)</span>
<span id="cb6-48"><a href="#cb6-48" aria-hidden="true"></a><span class="kw">colnames</span>(cnts) &lt;-<span class="st"> </span>uniqueBarcodes</span>
<span id="cb6-49"><a href="#cb6-49" aria-hidden="true"></a></span>
<span id="cb6-50"><a href="#cb6-50" aria-hidden="true"></a><span class="co">## -- Generate colData</span></span>
<span id="cb6-51"><a href="#cb6-51" aria-hidden="true"></a>depth &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb6-52"><a href="#cb6-52" aria-hidden="true"></a>    <span class="dt">sample =</span> <span class="kw">as.numeric</span>(id),</span>
<span id="cb6-53"><a href="#cb6-53" aria-hidden="true"></a>    <span class="dt">read =</span> <span class="kw">names</span>(ga)</span>
<span id="cb6-54"><a href="#cb6-54" aria-hidden="true"></a>) <span class="op">%&gt;%</span></span>
<span id="cb6-55"><a href="#cb6-55" aria-hidden="true"></a><span class="st">    </span><span class="kw">distinct</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb6-56"><a href="#cb6-56" aria-hidden="true"></a><span class="st">    </span><span class="kw">group_by</span>(sample) <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb6-57"><a href="#cb6-57" aria-hidden="true"></a><span class="st">    </span><span class="kw">tally</span>() <span class="op">%&gt;%</span><span class="st"> </span></span>
<span id="cb6-58"><a href="#cb6-58" aria-hidden="true"></a><span class="st">    </span><span class="kw">data.matrix</span>()</span>
<span id="cb6-59"><a href="#cb6-59" aria-hidden="true"></a>colData &lt;-<span class="st"> </span><span class="kw">data.frame</span>(</span>
<span id="cb6-60"><a href="#cb6-60" aria-hidden="true"></a>    <span class="dt">sample =</span> uniqueBarcodes,</span>
<span id="cb6-61"><a href="#cb6-61" aria-hidden="true"></a>    <span class="dt">depth =</span> depth[,<span class="dv">2</span>],</span>
<span id="cb6-62"><a href="#cb6-62" aria-hidden="true"></a>    <span class="dt">FRIP =</span> Matrix<span class="op">::</span><span class="kw">colSums</span>(cnts)<span class="op">/</span>depth[,<span class="dv">2</span>], </span>
<span id="cb6-63"><a href="#cb6-63" aria-hidden="true"></a>    <span class="dt">celltype =</span> <span class="kw">str_replace_all</span>(uniqueBarcodes, <span class="st">&quot;singles-SU070-|singles-SU353-|singles-PB1022-|BM1077-&quot;</span>, <span class="st">&#39;&#39;</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">str_replace_all</span>(<span class="st">&#39;-.*&#39;</span>, <span class="st">&#39;&#39;</span>)</span>
<span id="cb6-64"><a href="#cb6-64" aria-hidden="true"></a>)</span>
<span id="cb6-65"><a href="#cb6-65" aria-hidden="true"></a></span>
<span id="cb6-66"><a href="#cb6-66" aria-hidden="true"></a><span class="co">## -- Make summarized Experiment</span></span>
<span id="cb6-67"><a href="#cb6-67" aria-hidden="true"></a>sumExp &lt;-<span class="st"> </span>SummarizedExperiment<span class="op">::</span><span class="kw">SummarizedExperiment</span>(</span>
<span id="cb6-68"><a href="#cb6-68" aria-hidden="true"></a>    <span class="dt">rowRanges =</span> peaks, </span>
<span id="cb6-69"><a href="#cb6-69" aria-hidden="true"></a>    <span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> cnts),</span>
<span id="cb6-70"><a href="#cb6-70" aria-hidden="true"></a>    <span class="dt">colData =</span> colData</span>
<span id="cb6-71"><a href="#cb6-71" aria-hidden="true"></a>)</span>
<span id="cb6-72"><a href="#cb6-72" aria-hidden="true"></a><span class="kw">saveRDS</span>(sumExp, <span class="st">&#39;data/scATAC_LSC-LMPP-mono/sumExp.rds&#39;</span>)</span></code></pre></div>
</div>
<div id="filter-cells-and-promoters" class="section level3">
<h3>Filter cells and promoters</h3>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true"></a>sumExp &lt;-<span class="st"> </span>chromVAR<span class="op">::</span><span class="kw">filterSamples</span>(</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true"></a>    sumExp,</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true"></a>    <span class="dt">min_depth =</span> <span class="dv">1000</span>, </span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true"></a>    <span class="dt">min_in_peaks =</span> <span class="fl">0.10</span>, </span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true"></a>    <span class="dt">shiny =</span> <span class="ot">FALSE</span></span>
<span id="cb7-6"><a href="#cb7-6" aria-hidden="true"></a>)</span>
<span id="cb7-7"><a href="#cb7-7" aria-hidden="true"></a>sumExp &lt;-<span class="st"> </span>sumExp[<span class="kw">rowSums</span>(<span class="kw">assay</span>(sumExp)) <span class="op">&gt;</span><span class="st"> </span><span class="dv">10</span>, ]</span></code></pre></div>
</div>
<div id="add-gc-bias-to-promoters" class="section level3">
<h3>Add GC bias to promoters</h3>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true"></a>sumExp &lt;-<span class="st"> </span>chromVAR<span class="op">::</span><span class="kw">addGCBias</span>(sumExp, <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg38<span class="op">::</span>BSgenome.Hsapiens.UCSC.hg38)</span></code></pre></div>
</div>
<div id="search-for-motifs-with-high-deviation-from-background-distribution" class="section level3">
<h3>Search for motifs with high deviation from background distribution</h3>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true"></a>motifs &lt;-<span class="st"> </span>TFBSTools<span class="op">::</span><span class="kw">getMatrixSet</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true"></a>    JASPAR2020<span class="op">::</span>JASPAR2020,</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true"></a>    <span class="kw">list</span>(<span class="dt">species =</span> <span class="dv">9606</span>, <span class="dt">all_versions =</span> <span class="ot">FALSE</span>)</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true"></a>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true"></a>motif_ix &lt;-<span class="st"> </span>motifmatchr<span class="op">::</span><span class="kw">matchMotifs</span>(motifs, sumExp, <span class="dt">genome =</span> BSgenome.Hsapiens.UCSC.hg38<span class="op">::</span>BSgenome.Hsapiens.UCSC.hg38)</span>
<span id="cb9-6"><a href="#cb9-6" aria-hidden="true"></a>bg &lt;-<span class="st"> </span>chromVAR<span class="op">::</span><span class="kw">getBackgroundPeaks</span>(<span class="dt">object =</span> sumExp)</span>
<span id="cb9-7"><a href="#cb9-7" aria-hidden="true"></a>dev &lt;-<span class="st"> </span>chromVAR<span class="op">::</span><span class="kw">computeDeviations</span>(<span class="dt">object =</span> sumExp, <span class="dt">annotations =</span> motif_ix, <span class="dt">background_peaks =</span> bg)</span>
<span id="cb9-8"><a href="#cb9-8" aria-hidden="true"></a><span class="kw">saveRDS</span>(dev, <span class="st">&#39;data/scATAC_LSC-LMPP-mono/dev.rds&#39;</span>)</span></code></pre></div>
</div>
<div id="check-tf-enrichment-in-different-cell-types" class="section level3">
<h3>Check TF enrichment in different cell types</h3>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true"></a>variability &lt;-<span class="st"> </span>chromVAR<span class="op">::</span><span class="kw">computeVariability</span>(dev)</span>
<span id="cb10-2"><a href="#cb10-2" aria-hidden="true"></a>p &lt;-<span class="st"> </span>chromVAR<span class="op">::</span><span class="kw">plotVariability</span>(variability, <span class="dt">use_plotly =</span> <span class="ot">FALSE</span>) </span>
<span id="cb10-3"><a href="#cb10-3" aria-hidden="true"></a>tsne_results &lt;-<span class="st"> </span>chromVAR<span class="op">::</span><span class="kw">deviationsTsne</span>(dev, <span class="dt">threshold =</span> <span class="fl">1.5</span>)</span>
<span id="cb10-4"><a href="#cb10-4" aria-hidden="true"></a>p &lt;-<span class="st"> </span>cowplot<span class="op">::</span><span class="kw">plot_grid</span>(</span>
<span id="cb10-5"><a href="#cb10-5" aria-hidden="true"></a>    chromVAR<span class="op">::</span><span class="kw">plotDeviationsTsne</span>(dev, tsne_results, <span class="dt">sample_column =</span> <span class="st">&quot;celltype&quot;</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)[[<span class="dv">1</span>]], </span>
<span id="cb10-6"><a href="#cb10-6" aria-hidden="true"></a>    chromVAR<span class="op">::</span><span class="kw">plotDeviationsTsne</span>(dev, tsne_results, <span class="dt">annotation_name =</span> <span class="st">&quot;FOS&quot;</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)[[<span class="dv">1</span>]], </span>
<span id="cb10-7"><a href="#cb10-7" aria-hidden="true"></a>    chromVAR<span class="op">::</span><span class="kw">plotDeviationsTsne</span>(dev, tsne_results, <span class="dt">annotation_name =</span> <span class="st">&quot;CEBPA&quot;</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)[[<span class="dv">1</span>]], </span>
<span id="cb10-8"><a href="#cb10-8" aria-hidden="true"></a>    chromVAR<span class="op">::</span><span class="kw">plotDeviationsTsne</span>(dev, tsne_results, <span class="dt">annotation_name =</span> <span class="st">&quot;HIF1A&quot;</span>, <span class="dt">shiny =</span> <span class="ot">FALSE</span>)[[<span class="dv">1</span>]]</span>
<span id="cb10-9"><a href="#cb10-9" aria-hidden="true"></a>)</span></code></pre></div>
</div>
</div>
