---
title: "1. Program"
output:
  blogdown::html_page:
    highlight: tango
---

<script src="/scATACseq-workshop/rmarkdown-libs/header-attrs/header-attrs.js"></script>


<div id="session-1-processing-bulk-and-single-cell-atac-seq-reads" class="section level2">
<h2>Session 1: Processing bulk and single-cell ATAC-seq reads</h2>
<div id="lecture-1-bulk-atac-seq-experimental-approaches" class="section level3">
<h3>Lecture 1: Bulk ATAC-seq experimental approaches</h3>
<blockquote>
<p>15’</p>
</blockquote>
<ul>
<li>Introduction to bulk ATAC-seq methods: “classic”, omniATAC-seq, low input ATAC-seq, …</li>
</ul>
</div>
<div id="hands-on-1-processing-bulk-atac-seq-data" class="section level3">
<h3>Hands-on 1: Processing bulk ATAC-seq data</h3>
<blockquote>
<p>35’</p>
</blockquote>
<ul>
<li>Sequencing QC</li>
<li>Mapping reads and correcting</li>
<li>Visualizing and QC-ing the results</li>
</ul>
</div>
<div id="exercise-1-inspecting-bulk-atac-seq-outputs" class="section level3">
<h3>Exercise 1: Inspecting bulk ATAC-seq outputs</h3>
<blockquote>
<p>10’</p>
</blockquote>
<ul>
<li>Visually inspect the generated tracks and the identifed peaks</li>
<li>Get the distribution of fragment sizes over ATAC peaks.</li>
<li>Compare this distribution to the previous one. Discuss.</li>
</ul>
<hr />
</div>
<div id="lecture-2-single-cell-atac-seq-experimental-approaches" class="section level3">
<h3>Lecture 2: Single-cell ATAC-seq experimental approaches</h3>
<blockquote>
<p>20’</p>
</blockquote>
<ul>
<li>Introduction to “single cell” ATAC-seq methods: droplet-based scATACseq, indexed scATACseq, joint scRNAseq and scATACseq</li>
</ul>
</div>
<div id="hands-on-2-processing-single-cell-atac-seq-data" class="section level3">
<h3>Hands-on 2: Processing single-cell ATAC-seq data</h3>
<blockquote>
<p>30’</p>
</blockquote>
<ul>
<li>10X scATACseq: from reads to matrix counts with <code>cellranger</code></li>
<li>QC-ing with <code>Loupe</code></li>
</ul>
</div>
<div id="exercise-2-bulk-versus-single-cell-atac-seq" class="section level3">
<h3>Exercise 2: bulk versus single-cell ATAC-seq</h3>
<blockquote>
<p>10’</p>
</blockquote>
<ul>
<li>Compare aggregated signal obtained from single-cell ATAC-seq to that of bulk ATAC-seq</li>
</ul>
<hr />
</div>
<div id="homework-download-raw-data-from-gse-and-process-it" class="section level3">
<h3>Homework: Download raw data from GSE and process it</h3>
<ul>
<li>Install <code>cellranger</code> and mapping tools</li>
<li>Download reads from GSE129785 and GSE87646</li>
</ul>
</div>
</div>
<div id="session-2-peak-centered-differential-accessibility-analysis" class="section level2">
<h2>Session 2: Peak-centered differential accessibility analysis</h2>
<div id="lecture-3-clustering-single-cell-data" class="section level3">
<h3>Lecture 3: Clustering single-cell data</h3>
<blockquote>
<p>20’</p>
</blockquote>
<ul>
<li>Normalizing scATACseq data</li>
<li>Clustering approaches</li>
<li>Annotating cell types</li>
</ul>
</div>
<div id="hands-on-3-downstream-analysis-of-scatacseq-data" class="section level3">
<h3>Hands-on 3: Downstream analysis of scATACseq data</h3>
<blockquote>
<p>30’</p>
</blockquote>
<ul>
<li>Reading <code>cellranger</code> output in <code>R</code></li>
<li>Normalizing scATACseq counts</li>
<li>Clustering cells by their accessible loci</li>
</ul>
</div>
<div id="exercise-3-recovering-peaks-associated-with" class="section level3">
<h3>Exercise 3: Recovering peaks associated with</h3>
<blockquote>
<p>10’</p>
</blockquote>
<ul>
<li>Finding cluster-specific peaks</li>
</ul>
<hr />
</div>
<div id="lecture-4-atac-seq-and-accessibility-peaks" class="section level3">
<h3>Lecture 4: ATAC-seq and accessibility peaks</h3>
<blockquote>
<p>20’</p>
</blockquote>
<ul>
<li>Peaks in bulk ATAC-seq vs single-cell ATAC-seq</li>
<li>Integrating peak sets from different experiments</li>
<li>Finding enriched sequences</li>
</ul>
</div>
<div id="hands-on-4-comparing-atac-seq-peaks" class="section level3">
<h3>Hands-on 4: Comparing ATAC-seq peaks</h3>
<blockquote>
<p>30’</p>
</blockquote>
<ul>
<li>Quantifying accessibility at ATAC peaks</li>
<li>Differential accessibility assay</li>
<li>Clustering peak sets</li>
</ul>
</div>
<div id="exercise-4-peak-location-vs-genomic-features" class="section level3">
<h3>Exercise 4: Peak location vs genomic features</h3>
<blockquote>
<p>10’</p>
</blockquote>
<ul>
<li>Find which types of chromatin stats are enriched in accessible loci</li>
</ul>
<hr />
</div>
<div id="homework" class="section level3">
<h3>Homework</h3>
<ul>
<li>Group peaks by accessibility scores in different samples / cells</li>
<li>Perform gene ontology over-representation analyses (GOA)</li>
<li>Perform gene Set Enrichment Analyses (GSEA)</li>
</ul>
</div>
</div>
<div id="session-3-advanced-topics-and-multi-omics-integration" class="section level2">
<h2>Session 3: Advanced topics and multi-omics integration</h2>
<div id="lecture-5-motif-enrichment-analysis" class="section level3">
<h3>Lecture 5: Motif enrichment analysis</h3>
<blockquote>
<p>20’</p>
</blockquote>
<ul>
<li>The good, the bad and the ugly: Ref. sequence, ref. annotations and ref. version!</li>
<li>Tools to perform motif enrichment analysis</li>
<li>Public databases for transcription factor binding sites (TFBS)</li>
</ul>
</div>
<div id="hands-on-5-find-motifs-enriched-in-peak-sets" class="section level3">
<h3>Hands-on 5: Find motifs enriched in peak sets</h3>
<blockquote>
<p>30’</p>
</blockquote>
<ul>
<li>Use web-based or command-line <code>meme</code> software to find de novo motifs enriched in a set of peaks</li>
<li>Scan sets of peaks for known TF binding sites</li>
</ul>
</div>
<div id="hands-on-5-find-motifs-enriched-in-peak-sets-1" class="section level3">
<h3>Hands-on 5: Find motifs enriched in peak sets</h3>
<blockquote>
<p>10’</p>
</blockquote>
<ul>
<li>Compare de novo found motifs to public databases with <code>fimo</code></li>
</ul>
<hr />
</div>
<div id="topic-6-inferring-regulatory-networks-from-scatacseq" class="section level3">
<h3>Topic 6: Inferring regulatory networks from scATACseq</h3>
<blockquote>
<p>1h’</p>
</blockquote>
<ul>
<li>CICERO</li>
</ul>
<hr />
</div>
<div id="homework-1" class="section level3">
<h3>Homework</h3>
<ul>
<li>Find TF motifs enriched in peaks with increased accessibility in a specific cluster</li>
<li>Run chromVAR on a single-cell dataset</li>
<li>Leverage the motifmatchr package to replicate the ugly nested apply functions in Hands-on # 5</li>
</ul>
</div>
</div>
